---
title: "Post analysis of gene sets from differential expression data (mRNA, methylation and miRNA)"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load libraries
library(synapseClient)
library(plyr)
library(dplyr)
library(data.table)
library(stringr)
library(tidyr)
library(reshape2)
library(knitr)
library(biomaRt)
library(tools)
library(knit2synapse)
library(igraph)

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
# knitToFolderEntity(file = "./AnalyseDiffExp.Rmd", 
#                    parentId = 'syn4640410',
#                    entityName = 'Integrated Differential Expression Analysis (Differentiation state specific)', 
#                    overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = FALSE)
```
### Download differential expression results
```{r download.comp}
# Get all comparison names from synapse
compNames <- synTableQuery("SELECT * FROM syn4483642")@values
usedIds = "syn4483642"
compNames = rbind(compNames %>%
                    mutate(direction = 'up',
                           comparisonName = str_c(comparison,direction,sep='__')),
                  compNames %>%
                    mutate(direction = 'down',
                           comparisonName = str_c(comparison,direction,sep='__')))

# Diff states in a sensible, quasi time-based ordering
orderedDiffState <- c("SC", "DE", "MESO-5", "ECTO", "MESO-15", "MESO-30", "EB")

# Download TFs from Enrichr genesets
load(synGet('syn4867851')@filePath)
usedIds = c(usedIds,'syn4867851')
TFs = unique(c(sapply(names(GeneSets$ChEA), function(x){str_split(x, '-')[[1]][1]}),
               sapply(names(GeneSets$TRANSFAC_and_JASPAR_PWMs), function(x){str_split(x, ' ')[[1]][1]}),
               sapply(names(GeneSets$"ENCODE_TF_ChIP-seq_2015"), function(x){str_split(x, '_')[[1]][1]})))

# Set thresholds for mRNA, miRNA and methylation
thresh.p.mrna <- 0.0001
thresh.lfc.mrna <- 3

thresh.p.mirna <- 0.0001
thresh.lfc.mirna <- 2

thresh.p.methylation <- 0.05
thresh.lfc.methylation <- 0.4
usedIds <- c()
```
#### Download mRNA data
```{r download.mrna}
mrna.obj <- synGet("syn4484232")
usedIds <- c(usedIds, mrna.obj@properties$id)

d.mrna <- fread(getFileLocation(mrna.obj), 
                data.table=FALSE, 
                showProgress = FALSE) %>% 
  mutate(assay='mRNA', comparisonName = Comparison) %>%
  separate(Comparison, into = c('dataRestrictionShort','comparison.short','direction'), sep='__') %>%
#   separate(comparison.short, into = c('variable1Short','vs','variable2Short'), sep='_') %>%
  merge(dplyr::select(compNames, class, comparisonName), by = 'comparisonName')

# Filter out the comparisons we're interested in
# And significance levels as well
d2.mrna <- d.mrna %>%
  dplyr::rename(feature=GeneSymbol) %>%
  filter(abs(logFC) > thresh.lfc.mrna,
         adj.P.value < thresh.p.mrna,
         dataRestrictionShort == 'All') %>%
  filter(class == "Diffname_short") %>%
  unique %>%
  mutate(target = feature)
```
#### Download miRNA data
```{r download.mirna}
mirna.obj <- synGet("syn4609631")
usedIds <- c(usedIds, mirna.obj@properties$id)

d.mirna <- fread(getFileLocation(mirna.obj), data.table=FALSE, 
                 showProgress = FALSE) %>%
  mutate(assay='miRNA', comparisonName = Comparison) %>%
  separate(Comparison, into = c('dataRestrictionShort','comparison.short','direction'), sep='__') %>%
#   separate(comparison.short, into = c('variable1Short','vs','variable2Short'), sep='_') %>%
  merge(dplyr::select(compNames, class, comparisonName), by = 'comparisonName')

# Filter out the comparisons we're interested in
# And significance levels as well
d2.mirna <- d.mirna %>%
  dplyr::rename(feature=GeneSymbol) %>%
  filter(abs(logFC) > thresh.lfc.mirna,
         adj.P.value < thresh.p.mirna,
         dataRestrictionShort == 'All') %>%
  filter(class == "Diffname_short") %>%
  unique

# Get miRNA mapping files
DMAP.ID = 'syn3461627'
usedIds <- c(usedIds, DMAP.ID)
DMAP = fread(synGet(DMAP.ID)@filePath, data.table=F, header=F)
setnames(DMAP, c('V1','V2'), c('hsa_id','ensembl_gene_id'))

DMAP <- filter(DMAP, hsa_id %in% d2.mirna$feature)

# Get human related mapping
Hs = useMart("ENSEMBL_MART_ENSEMBL", host="www.ensembl.org") # use this one when biomart.org is down
Hs = useDataset("hsapiens_gene_ensembl", Hs)
human_ensg2symbol = getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                          filters = "ensembl_gene_id",                         
                          values = unique(DMAP$ensembl_gene_id),
                          mart = Hs)

DMAP <- left_join(DMAP, human_ensg2symbol, by = 'ensembl_gene_id') %>%
  dplyr::rename(feature=hsa_id) %>% 
  dplyr::select(feature, hgnc_symbol)
d2.mirna <- d2.mirna %>% left_join(DMAP) %>% dplyr::rename(target = hgnc_symbol)
```
#### Download methylation data
```{r download.methyl}
methyl.obj <- synGet("syn4527629")
usedIds <- c(usedIds, methyl.obj@properties$id)

d.methyl <- fread(getFileLocation(methyl.obj), data.table=FALSE, 
                  showProgress = FALSE) %>%
  mutate(assay='methyl', comparisonName = Comparison) %>%
  separate(Comparison, into = c('dataRestrictionShort','comparison.short','direction'), sep='__') %>%
#   separate(comparison.short, into = c('variable1Short','vs','variable2Short'), sep='_') %>%
  merge(dplyr::select(compNames, class, comparisonName), by = 'comparisonName') %>%
  dplyr::select(-addressA, -addressB, -channel, -platform, -percentGC, -sourceSeq, -strand, -seqnames, -start,
                -probeType, -probeStart, -probeEnd, -probeTarget, -distanceToTSS, -nearestTSS, -end, -width,
                -nearestTSS.ID, -distanceToTx, -nearestTx.ID)

# Filter out the comparisons we're interested in
# And significance levels as well
d2.methyl <- d.methyl %>%
  dplyr::rename(feature=methProbeIDs, target = nearestTx) %>%
  filter(abs(logFC) > thresh.lfc.methylation,
         adj.P.value < thresh.p.methylation,
         dataRestrictionShort == 'All') %>%
  filter(class == "Diffname_short") %>%
  unique
  
```
#### Unique features from all three assays that are specific to each diff state
Filtered based on those features that are specific to each cellular transitions in a given differntiaiton stage 
```{r extract.feature}
# Add TF tags
isTF <- function(x){TF = ifelse(x %in% TFs, 1, 0); return(TF)}
allSigResults = rbind(d2.mrna, d2.mirna, d2.methyl) %>% mutate(TF = target) %>% mutate_each(funs(isTF), TF)

# Prune miRNA and methylation that has opposite differential expressions in mRNA
tmp = lapply(unique(allSigResults$comparison.short), function(comparison, allSigResults){
  mRNA = filter(allSigResults, comparison.short == comparison & assay == 'mRNA')
  miRNA = filter(allSigResults, comparison.short == comparison & assay == 'miRNA')
  methyl = filter(allSigResults, comparison.short == comparison & assay == 'methyl')
  
  tmp3 = merge(miRNA, mRNA, by = 'target') %>% filter(direction.x != direction.y)
  miRNA = tmp3[, c(1,grep('.x', colnames(tmp3)))]
  colnames(miRNA) = gsub('.x','', colnames(miRNA))
  
  tmp3 = merge(methyl, mRNA, by = 'target') %>% filter(direction.x != direction.y)
  methyl = tmp3[, c(1,grep('.x', colnames(tmp3)))]
  colnames(methyl) = gsub('.x','', colnames(methyl))
  
  return(rbind(mRNA, miRNA, methyl))
}, allSigResults)
allSigResults = ldply(tmp)

# Segregate results for all diff states
diffExp = lapply(c("SC", "DE", "MESO5", "ECTO", "MESO15", "MESO30", "EB"), function(diffstate, allSigResults){
  compName = unique(grep(diffstate, allSigResults$comparison.short, value = T))
  tmp = dplyr::filter(allSigResults, comparison.short %in% compName)
}, allSigResults)
names(diffExp) = c("SC", "DE", "MESO5", "ECTO", "MESO15", "MESO30", "EB")

diffExp = lapply(diffExp, function(DE.diffState){
  # Get unique features for each transitions
  comparisons = unique(DE.diffState$comparison.short)
  
  # Get unique features that are specific to this comparison
  tmp = lapply(comparisons, function(comparison, DE.diffState){
    featureExSet = dplyr::select(dplyr::filter(DE.diffState, comparison.short != comparison), feature) %>% unlist
    DE.diffState = filter(DE.diffState, !(feature %in% featureExSet) & comparison.short == comparison)
  }, DE.diffState)
  DE.diffState = ldply(tmp)
})
diffExp = ldply(diffExp) %>% plyr::rename(c(".id" = "DiffState"))

# Get miRNA and methylation mappings on mRNAs for each transitions
diffExp = split(diffExp, diffExp$comparison.short)
diffExp = lapply(diffExp, function(diffExp.comp){
  diffExp.comp = diffExp.comp %>% mutate(miRNA = 0, methyl = 0)
  diffExp.comp = split(diffExp.comp, diffExp.comp$assay)
  
  miRNA.targets = diffExp.comp$miRNA$target %>% unique %>% unlist
  diffExp.comp$mRNA$miRNA[diffExp.comp$mRNA$feature %in% miRNA.targets] = 1
  
  methyl.targets = diffExp.comp$methyl$target %>% unique %>% unlist
  diffExp.comp$mRNA$methyl[diffExp.comp$mRNA$feature %in% methyl.targets] = 1
  
  tmp = ldply(diffExp.comp) %>% dplyr::select(-(.id))
  return(tmp)
})
diffExp = ldply(diffExp, .id = NULL)

tmp <- diffExp %>% 
  dplyr::select(DiffState, assay, feature) %>% 
  unique %>% 
  group_by(DiffState, assay) %>% 
  summarise(count = n()) %>% 
  dcast(assay~DiffState, value.var = "count")

tmp[is.na(tmp)] = 0
tmp$All = rowSums(tmp[,-(1)])
kable(tmp)
```

```{r print.individuals}
tmp <- diffExp %>%
  dplyr::select(DiffState, assay, comparison.short, feature) %>%
  unique %>%
  group_by(DiffState, assay, comparison.short) %>% 
  summarise(count = n())
tmp <- split(tmp, tmp$DiffState)

for( name in names(tmp)){
  writeLines(name)
  tmp1 <- tmp[[name]] %>% dcast(assay~comparison.short, value.var = "count")
  tmp1[is.na(tmp1)] = 0
  tmp1$All = rowSums(tmp1[,-(1)])
  print(kable(tmp1))
}
```
#### TF only analysis
```{r analyse.TFs}
tmp1 <- diffExp %>%
  filter(assay == "mRNA") %>%
  dplyr::select(feature, TF) %>%
  dplyr::filter(TF == 1) %>%
  group_by(feature) %>%
  summarise(count = n()) %>%
  arrange(desc(count))
```
##### In all six comparisons #####
BCL11A - down regulated in ECTO (by miRNAs hsa-miR-92b-3p,  hsa-miR-448, hsa-miR-486-3p (vs MESO15), hsa-miR-363-3p, hsa-miR-23-3p (vs MESO30), hsa-miR-23b-3p (vs SC))
FOXC1 - down regulated in SC (no miRNA and methyl regulation)
GATA3  - down regualted in SC (no miRNA and methyl regualtion)
HOXA2 - upregulated in EB (miRNA regaulted transition from ECTO - hsa-miR-548ah-5p)
HOXA5 - upregulated in EB (miRNA regulated transition from MESO5 - hsa-miR-320d)
HOXD9 - upregulated in EB (miRNA regulated transition from SC - hsa-miR-18a-3p)
NR5A1 - up regulated in ECTO by both methylation and miRNA  (probeID: cg14143574 for all four diff states, hsa-miR-1910-5p from EB, MESO30 and SC, hsa-miR-24-3p from MESO30, hsa-miR-326 from MESO15)

##### In any five comparisons #####
KLF5 - upregulated in SC except ECTO (hsa-miR-18a-3p regulation in EB and MESO15, hsa-miR-1305 and hsa-miR-32-3p regulation in MESO30, no regulation of miRNA and methylation against MESO5 and DE)
NFIB - downregulated in ECTO except MESO15 (hsa-miR-129-5p regulation in DE, EB, MESO30, MESO5, SC and hsa-miR-17-3p regulation in MESO30)
NR3C1 - downregulated in ECTO except SC (mainly by miRNAs hsa-miR-548ah-5p, hsa-miR-204-5p,hsa-miR-548p, hsa-miR-129-2-3p, hsa-miR-548ah-3p, hsa-miR-500a-5p, hsa-miR-429, hsa-miR-589-3p, hsa-miR-18b-5p, hsa-miR-18a-5p, hsa-miR-378i)

### Plot expression
Subset of genes that participate in differential expression (including targets from miRNA and methylation)
```{r plot.exp, fig.height=24, fig.width=15}
diffExp.mRNA = dplyr::filter(diffExp, assay == "mRNA")
exp = fread(synGet('syn4483934')@filePath, data.table=F, header=T)

exp1 = filter(exp, GeneName %in% unique(diffExp.mRNA$target))
rownames(exp1) = exp1$GeneName
exp1 = exp1[,-(1)]

cov = fread(synGet('syn4483940')@filePath, data.table=F, header=T)
cov = dplyr::arrange(cov, Diffname_short)
exp1 = exp1[,cov$UID]

ha = HeatmapAnnotation(df = cov[,'Diffname_short',drop=F], which = "column")

annotate = as.data.frame(matrix(0,dim(exp1)[1],3))
rownames(annotate) = rownames(exp1)
colnames(annotate) = c("TF", "miRNA", "methyl")
annotate$TF = (rownames(annotate) %in% TFs)*1
annotate$miRNA = (rownames(annotate) %in% (filter(diffExp, assay == "miRNA") %>% dplyr::select(target) %>% unique %>% unlist))*1
annotate$methyl = (rownames(annotate) %in% (filter(diffExp, assay == "methyl") %>% dplyr::select(target) %>% unique %>% unlist))*1
ha1 = HeatmapAnnotation(df = annotate, which = "row")

Heatmap(exp1, row_title = 'Genes', column_title = 'Samples', cluster_columns = F, 
        show_row_names = F, show_column_names = F, top_annotation = ha) + ha1
```
#### Store results in synapse
```{r synapse.store}
obj = Folder(name = 'Integrated Differential Expression Analysis (Differentiation state specific)', parentId = 'syn4640410')
obj = synStore(obj)

write.table(diffExp, file = "uniqueSign.tsv", sep = "\t", row.names=F, quote=F)
objFile = File("uniqueSign.tsv", name = "Integrated Signatures (Differentiation state specific)", parentId = obj$properties$id)
objFile = synStore(objFile, used = usedIds, activityName = "Integrated Analysis of Differential Expression")
```