---
title: "Modularising mRNA, miRNA, DNA methylation probes and splicing junctions"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'AnalyseDiffExp_All.Rmd',
                                 parentId = "syn5194922",
                                 entityName = 'Differential Analysis Features')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
library(matrixStats)

library(fpc)
library(WGCNA)
library(flashClust)

library(circlize)
colfunc <- colorRampPalette(c("white", "red"))

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapseStore.params}
parentId = "syn5194922"
ALL_USED_IDs = c()

ActivityName <- 'Analyse differential analysis results'

# Get the latest commit of this code from github
thisFileName <- 'AnalyseDiffExp_All.Rmd'

thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='diff_exp')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Populate wiki with results
CODE <- Folder(name = 'Differential Analysis Features',parentId = parentId)
CODE <- synStore(CODE)
```

```{r set.thresholds}
pval.th <- 0.01

mrna.lfc.th <- log2(2)
mirna.lfc.th <- log2(2)
methyl.changeBeta.th <- 0.35
splicing.changePSI.th <- 0.2
```

#### Get median counts from synapse
```{r getMedianCount, include=FALSE}
counts.id = 'syn5637772'
ALL_USED_IDs = c(ALL_USED_IDs, counts = counts.id)
counts = downloadFile(counts.id)
```

#### Use co-expression results to map miRNA and methylation to mRNA
```{r mirna.mrna}
coexp.id = 'syn5713477'
ALL_USED_IDs = c(ALL_USED_IDs, coexp = coexp.id)

coexp = downloadFile(coexp.id)
```

#### Get mRNA differential expression clusters
```{r mrna}
### Download differential expression results from synapse
diffexp.mrna.id = c(mrna = 'syn5013690')
ALL_USED_IDs = c(ALL_USED_IDs, diffexp.mrna.id)

# Diff states in a sensible, quasi time-based ordering
orderedDiffState = c("DE", "EB", "ECTO", "MESO15", "MESO30", "MESO5", "SC")
Comparisons = apply(combn(orderedDiffState, 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get mrna differential expression results from synapse
diffexp.mrna = downloadFile(diffexp.mrna.id) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mrna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Extract median mrna counts for differentially expressed genes
mrna.counts = counts %>%
  filter(feature %in% unique(diffexp.mrna$feature))

# Scale expression counts wrt each gene
mrna.counts.scaled = 2^mrna.counts[,-(1:2)]
rownames(mrna.counts.scaled) = mrna.counts$feature

mrna.counts.scaled[mrna.counts.scaled<=0] = 0
mrna.counts.scaled = t(apply(mrna.counts.scaled, 1, 
                             function(x){ 
                               x = x/max(x)
                             }))
mrna.counts.scaled = na.omit(mrna.counts.scaled)
writeLines(paste('Clustering expression of',dim(mrna.counts.scaled)[1],'mrnas'))
```

```{r mrna.clust1, fig.height=8, fig.width=25}
# Finding optimum number of clusters 
clust.d = dist(mrna.counts.scaled)
clust = hclust(clust.d)
plot(clust, main = 'mrna')
mrna.cluster = cutreeDynamic(clust, 
                             distM = as.matrix(clust.d),
                             minClusterSize = 50,
                             verbose = 4, pamRespectsDendro = T,
                             pamStage = T,
                             deepSplit = 4)
names(mrna.cluster) = rownames(mrna.counts.scaled)

# Merge close modules
mrna.cluster = mergeCloseModules(t(mrna.counts.scaled), mrna.cluster, cutHeight = 0.2, verbose = 3)$colors
```

```{r mrna.clust2, fig.height=8, fig.width=10}
# Calculate medians for each cluster
clust.median = lapply(1:max(mrna.cluster), function(i, mrnaCountsScaled, mrnaCluster){
  med = rowMedians(t(mrnaCountsScaled[names(mrnaCluster)[mrnaCluster == i],]))
  names(med) = colnames(mrnaCountsScaled)
  return(med)
}, mrna.counts.scaled, mrna.cluster) %>%
  do.call(rbind,.)
rownames(clust.median) = paste('Cluster',1:dim(clust.median)[1])
clust.median = na.omit(clust.median)
writeLines(paste(dim(clust.median)[1], 'clusters are inferred in mrna'))

figure_size = list(title_gp = gpar(fontsize = 18), labels_gp = gpar(fontsize = 14))

Heatmap(clust.median, name = 'Counts', column_title = 'mRNA: Cluster medians', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        # split = mrna.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        show_row_names = T, row_names_side = 'left', show_column_names = T, show_row_dend = T,
        heatmap_legend_param = figure_size)
```

```{r mrna.clust3, fig.height=25, fig.widht=10}
Heatmap(mrna.counts.scaled, name = 'Counts', column_title = 'mRNA: Scaled median counts', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        split = mrna.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_dend = F,
        heatmap_legend_param = figure_size)

pdf(file = 'mrnaClusters.pdf', width = 15, height = 20)
for (i in unique(mrna.cluster)){
  h = Heatmap(mrna.counts.scaled[names(mrna.cluster)[mrna.cluster == i],], 
          name = 'Counts', column_title = paste('Cluster',i), 
          cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
          # split = mrna.cluster,
          column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
          row_title_gp = gpar(fontsize = 22, fontface = "bold"),
          column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
          show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_hclust = F,
          heatmap_legend_param = figure_size)
  draw(h)
}
dev.off()
```

```{r mrna.clust4, fig.height=25, fig.widht=10}
# Pick related clusters
# c(94 = 'MESO15', 78 = 'EB', 17 = 'MESO30-EB',
#   76 = 'LATE', 8 = 'ECTO', 20 = 'EARLY', 83 = 'MESO15')
Heatmap(mrna.counts.scaled, name = 'Counts', column_title = 'mRNA: Scaled median counts', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        split = mrna.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_dend = F,
        heatmap_legend_param = figure_size)

# Extract clusters into gct files
mrna.cluster.gct = data.frame(feature = names(mrna.cluster), cluster = mrna.cluster) %>%
  group_by(cluster) %>%
  summarise(feature = paste(unique(feature), collapse = ','))
```

#### Get miRNA differential expression clusters
```{r mirna}
### Download differential expression results from synapse
diffexp.mirna.id = c(mirna = 'syn5014584')
ALL_USED_IDs = c(ALL_USED_IDs, diffexp.mirna.id)

# Diff states in a sensible, quasi time-based ordering
orderedDiffState = c("DE", "EB", "ECTO", "MESO15", "MESO30", "MESO5", "SC")
Comparisons = apply(combn(orderedDiffState, 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get mirna differential expression results from synapse
diffexp.mirna = downloadFile(diffexp.mirna.id) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mirna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Extract median mirna counts for differentially expressed genes
mirna.counts = counts %>%
  filter(feature %in% unique(diffexp.mirna$feature))

# Scale expression counts wrt each gene
mirna.counts.scaled = 2^mirna.counts[,-(1:2)]
rownames(mirna.counts.scaled) = mirna.counts$feature

mirna.counts.scaled[mirna.counts.scaled<=0] = 0
mirna.counts.scaled = t(apply(mirna.counts.scaled, 1, 
                             function(x){ 
                               x = x/max(x)
                             }))
mirna.counts.scaled = na.omit(mirna.counts.scaled)
writeLines(paste('Clustering expression of',dim(mirna.counts.scaled)[1],'mirnas'))
```

```{r mirna.clust1, fig.height=8, fig.width=25}
# Finding optimum number of clusters 
clust.d = dist(mirna.counts.scaled)
clust = hclust(clust.d)
plot(clust, main = 'mirna')
mirna.cluster = cutreeDynamic(clust, 
                             distM = as.matrix(clust.d),
                             minClusterSize = 10,
                             verbose = 4, pamRespectsDendro = T,
                             pamStage = T,
                             deepSplit = 4)
names(mirna.cluster) = rownames(mirna.counts.scaled)

# Merge close modules
mirna.cluster = mergeCloseModules(t(mirna.counts.scaled), mirna.cluster, cutHeight = 0.15, verbose = 3)$colors
```

```{r mirna.clust2, fig.height=8, fig.width=8}
# Calculate medians for each cluster
clust.median = lapply(1:max(mirna.cluster), function(i, mirnaCountsScaled, mirnaCluster){
  med = rowMedians(t(mirnaCountsScaled[names(mirnaCluster)[mirnaCluster == i],]))
  names(med) = colnames(mirnaCountsScaled)
  return(med)
}, mirna.counts.scaled, mirna.cluster) %>%
  do.call(rbind,.)
rownames(clust.median) = paste('Cluster',1:dim(clust.median)[1])
clust.median = na.omit(clust.median)
writeLines(paste(dim(clust.median)[1], 'clusters are inferred in mirna'))

figure_size = list(title_gp = gpar(fontsize = 18), labels_gp = gpar(fontsize = 14))

Heatmap(clust.median, name = 'Counts', column_title = 'mirna: Cluster medians', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        # split = mirna.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        show_row_names = T, row_names_side = 'left', show_column_names = T, show_row_hclust = T,
        heatmap_legend_param = figure_size)
```

```{r mirna.clust3, fig.height=25, fig.width=10}
Heatmap(mirna.counts.scaled, name = 'Counts', column_title = 'mirna: Scaled median counts', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        split = mirna.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), #combined_name_fun =  NULL,
        show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_dend = F,
        heatmap_legend_param = figure_size)

pdf(file = 'mirnaClusters.pdf', width = 15, height = 20)
for (i in unique(mirna.cluster)){
  h = Heatmap(mirna.counts.scaled[names(mirna.cluster)[mirna.cluster == i],], 
          name = 'Counts', column_title = paste('Cluster',i), 
          cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
          # split = mirna.cluster,
          column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
          row_title_gp = gpar(fontsize = 22, fontface = "bold"),
          column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
          show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_hclust = F,
          heatmap_legend_param = figure_size)
  draw(h)
}
dev.off()

# Extract clusters into gct files
mirna.cluster.gct = data.frame(feature = names(mirna.cluster), cluster = mirna.cluster) %>%
  group_by(cluster) %>%
  summarise(feature = paste(unique(feature), collapse = ','))
```
#### Map mirna - mrna
```{r mirna.mrna}
mirna.mrna = coexp %>%
  filter(feature.assay == 'mirna', target.assay %in% c('TF', 'mrna'),
         abs(feature.lfc) >= mirna.lfc.th,
         abs(target.lfc) >= mrna.lfc.th,
         abs(coexpression) >= 0.9)

map.mrna <- function(x){
  filter(mirna.mrna, feature %in% str_split(x, ',')[[1]]) %>%
    dplyr::select(target) %>%
    unique %>%
    unlist %>%
    paste(collapse = ',')
}
  
mirna.mrna.cluster.gct = group_by(mirna.cluster.gct, cluster) %>%
  summarise(feature = map.mrna(feature))
```

#### Get DNA methylation differential expression clusters
```{r methyl}
### Download differential expression results from synapse
diffexp.methyl.id = c(methyl = 'syn4527629')
ALL_USED_IDs = c(ALL_USED_IDs, diffexp.methyl.id)

# Diff states in a sensible, quasi time-based ordering
orderedDiffState = c("DE", "EB", "ECTO", "MESO5", "SC")
Comparisons = apply(combn(orderedDiffState, 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get methyl differential expression results from synapse
diffexp.methyl = downloadFile(diffexp.methyl.id) %>%
  dplyr::select(-one_of('feature')) %>%
  dplyr::rename(feature = methProbeIDs) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= methyl.changeBeta.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, changeBETA) 

# Extract median methyl counts for differentially expressed genes
methyl.counts = counts %>%
  filter(feature %in% unique(diffexp.methyl$feature)) %>%
  dplyr::select(-MESO15, -MESO30)

# Scale expression counts wrt each gene
methyl.counts.scaled = 2^methyl.counts[,-(1:2)]
rownames(methyl.counts.scaled) = methyl.counts$feature

methyl.counts.scaled[methyl.counts.scaled<=0] = 0
methyl.counts.scaled = t(apply(methyl.counts.scaled, 1, 
                             function(x){ 
                               x = x/max(x)
                             }))
methyl.counts.scaled = na.omit(methyl.counts.scaled)
writeLines(paste('Clustering expression of',dim(methyl.counts.scaled)[1],'methyl'))
```

```{r methyl.clust1, fig.height=10, fig.width=25}
# Finding optimum number of clusters 
clust.d = dist(methyl.counts.scaled)
clust = hclust(clust.d)
plot(clust, main = 'methyl')
methyl.cluster = cutreeDynamic(clust, 
                             distM = as.matrix(clust.d),
                             minClusterSize = 100,
                             verbose = 4, pamRespectsDendro = T,
                             pamStage = T,
                             deepSplit = 4)
names(methyl.cluster) = rownames(methyl.counts.scaled)

# Merge close modules
methyl.cluster = mergeCloseModules(t(methyl.counts.scaled), methyl.cluster, cutHeight = 0.2, verbose = 3)$colors
```

```{r methyl.clust2, fig.height=10, fig.width=10}
# Calculate medians for each cluster
clust.median = lapply(1:max(methyl.cluster), function(i, methylCountsScaled, methylCluster){
  med = rowMedians(t(methylCountsScaled[names(methylCluster)[methylCluster == i],]))
  names(med) = colnames(methylCountsScaled)
  return(med)
}, methyl.counts.scaled, methyl.cluster) %>%
  do.call(rbind,.)
rownames(clust.median) = paste('Cluster',1:dim(clust.median)[1])
clust.median = na.omit(clust.median)
writeLines(paste(dim(clust.median)[1], 'clusters are inferred in methyl'))

figure_size = list(title_gp = gpar(fontsize = 18), labels_gp = gpar(fontsize = 14))

Heatmap(clust.median, name = 'Counts', column_title = 'methyl: Cluster medians', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        # split = methyl.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        show_row_names = T, row_names_side = 'left', show_column_names = T, show_row_dend = T,
        heatmap_legend_param = figure_size)
```

```{r methyl.clust3, fig.height=25, fig.width=10}
Heatmap(methyl.counts.scaled, name = 'Counts', column_title = 'methyl: Scaled median counts', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        split = methyl.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), #combined_name_fun =  NULL,
        show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_dend = F,
        heatmap_legend_param = figure_size)

pdf(file = 'methylClusters.pdf', width = 15, height = 20)
for (i in unique(methyl.cluster)){
  h = Heatmap(methyl.counts.scaled[names(methyl.cluster)[methyl.cluster == i],], 
          name = 'Counts', column_title = paste('Cluster',i), 
          cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
          # split = methyl.cluster,
          column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
          row_title_gp = gpar(fontsize = 22, fontface = "bold"),
          column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
          show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_hclust = F,
          heatmap_legend_param = figure_size)
  draw(h)
}
dev.off()

# Extract clusters into gct files
methyl.cluster.gct = data.frame(feature = names(methyl.cluster), cluster = methyl.cluster) %>%
  group_by(cluster) %>%
  summarise(feature = paste(unique(feature), collapse = ','))
```

#### Map methyl - mrna
```{r methyl.mrna}
methyl.mrna = coexp %>%
  filter(feature.assay == 'methyl', target.assay %in% c('TF', 'mrna'),
         abs(feature.cb) >= methyl.changeBeta.th,
         abs(target.lfc) >= mrna.lfc.th)

map.mrna <- function(x){
  filter(methyl.mrna, feature %in% str_split(x, ',')[[1]]) %>%
    dplyr::select(target) %>% unique %>% unlist %>%
    paste(collapse = ',')
}
  
methyl.mrna.cluster.gct = group_by(methyl.cluster.gct, cluster) %>%
  summarise(feature = map.mrna(feature))
```

### Get mRNA differentially spliced junction clusters
```{r splicing}
### Download differential expression results from synapse
diffexp.splicing.id = c(splicing = 'syn5049321')
ALL_USED_IDs = c(ALL_USED_IDs, diffexp.splicing.id)

# Diff states in a sensible, quasi time-based ordering
orderedDiffState = c("DE", "EB", "ECTO", "MESO5", "MESO15", "MESO30", "SC")
Comparisons = apply(combn(orderedDiffState, 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get splicing differential expression results from synapse
diffexp.splicing = downloadFile(diffexp.splicing.id) %>%
  dplyr::select(-one_of('feature')) %>%
  plyr::rename(c('Minor-Isoform' = 'feature')) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= splicing.changePSI.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, changePSI) 

# Extract median splicing counts for differentially expressed genes
splicing.counts = counts %>%
  filter(feature %in% unique(diffexp.splicing$feature))

# Scale expression counts wrt each gene
splicing.counts.scaled = 2^splicing.counts[,-(1:2)]
rownames(splicing.counts.scaled) = splicing.counts$feature

splicing.counts.scaled[splicing.counts.scaled<=0] = 0
splicing.counts.scaled = t(apply(splicing.counts.scaled, 1, 
                             function(x){ 
                               x = x/max(x)
                             }))
splicing.counts.scaled = na.omit(splicing.counts.scaled)
writeLines(paste('Clustering expression of',dim(splicing.counts.scaled)[1],'splicing junctions'))
```

```{r splicing.clust1, fig.height=10, fig.width=25}
# Finding optimum number of clusters 
clust.d = dist(splicing.counts.scaled)
clust = hclust(clust.d)
plot(clust, main = 'splicing')
splicing.cluster = cutreeDynamic(clust, 
                             distM = as.matrix(clust.d),
                             minClusterSize = 50,
                             verbose = 4, pamRespectsDendro = T,
                             pamStage = T,
                             deepSplit = 4)
names(splicing.cluster) = rownames(splicing.counts.scaled)

# Merge close modules
splicing.cluster = mergeCloseModules(t(splicing.counts.scaled), splicing.cluster, cutHeight = 0.15, verbose = 3)$colors
```

```{r splicing.clust2, fig.height=10, fig.width=10}
# Calculate medians for each cluster
clust.median = lapply(1:max(splicing.cluster), function(i, splicingCountsScaled, splicingCluster){
  med = rowMedians(t(splicingCountsScaled[names(splicingCluster)[splicingCluster == i],]))
  names(med) = colnames(splicingCountsScaled)
  return(med)
}, splicing.counts.scaled, splicing.cluster) %>%
  do.call(rbind,.)
rownames(clust.median) = paste('Cluster',1:dim(clust.median)[1])
clust.median = na.omit(clust.median)
writeLines(paste(dim(clust.median)[1], 'clusters are inferred in splicing'))

figure_size = list(title_gp = gpar(fontsize = 18), labels_gp = gpar(fontsize = 14))
```

```{r splicing.clust3, fig.height=25, fig.width=10}
Heatmap(clust.median, name = 'Counts', column_title = 'splicing: Cluster medians', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        # split = splicing.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
        show_row_names = T, row_names_side = 'left', show_column_names = T, show_row_dend = T,
        heatmap_legend_param = figure_size)

Heatmap(splicing.counts.scaled, name = 'Counts', column_title = 'splicing: Scaled median counts', 
        cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
        split = splicing.cluster,
        column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
        row_title_gp = gpar(fontsize = 22, fontface = "bold"),
        column_names_gp = gpar(fontsize = 18), #combined_name_fun =  NULL,
        show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_dend = F,
        heatmap_legend_param = figure_size)

pdf(file = 'splicingClusters.pdf', width = 15, height = 20)
for (i in unique(splicing.cluster)){
  h = Heatmap(splicing.counts.scaled[names(splicing.cluster)[splicing.cluster == i],], 
          name = 'Counts', column_title = paste('Cluster',i), 
          cluster_columns = F, cluster_rows=T, col = brewer.pal(5, "RdPu"), 
          # split = splicing.cluster,
          column_title_gp = gpar(fontsize = 22, fontface = "bold"), 
          row_title_gp = gpar(fontsize = 22, fontface = "bold"),
          column_names_gp = gpar(fontsize = 18), combined_name_fun = NULL,
          show_row_names = F, row_names_side = 'left', show_column_names = T, show_row_hclust = F,
          heatmap_legend_param = figure_size)
  draw(h)
}
dev.off()

# Extract clusters into gct files
splicing.cluster.gct = data.frame(feature = names(splicing.cluster), cluster = splicing.cluster) %>%
  group_by(cluster) %>%
  summarise(feature = paste(unique(feature), collapse = ','))
```

#### Map splicing - mrna
```{r splicing.mrna}
splicing.mrna = coexp %>%
  filter(feature.assay == 'splicing', target.assay %in% c('TF', 'mrna'),
         abs(feature.cpsi) >= splicing.changePSI.th,
         abs(target.lfc) >= mrna.lfc.th)

map.mrna <- function(x){
  filter(splicing.mrna, feature %in% str_split(x, ',')[[1]]) %>%
    dplyr::select(target) %>% unique %>% unlist %>%
    paste(collapse = ',')
}
  
splicing.mrna.cluster.gct = group_by(splicing.cluster.gct, cluster) %>%
  summarise(feature = map.mrna(feature))
```

### Store results in synapse
```{r synstore, cache=FALSE}
# Store mrna results in synapse
write.table(mrna.cluster.gct, file = 'mrnaCluster.gct', row.names = F, sep = '\t', quote=F)
mrna.obj = File('mrnaCluster.gct', name = "Differentially Expressed mRNA (gct format)", parentId = CODE$properties$id)
mrna.obj = synStore(mrna.obj, activityName = ActivityName, executed = thisFile, used = ALL_USED_IDs[c('counts', 'mrna')])

# Store mirna results in synapse
write.table(mirna.cluster.gct, file = 'mirnaCluster.gct', row.names = F, sep = '\t', quote=F)
mirna.obj = File('mirnaCluster.gct', name = "Differentially Expressed miRNA (gct format)", parentId = CODE$properties$id)
mirna.obj = synStore(mirna.obj, activityName = ActivityName, executed = thisFile, used = ALL_USED_IDs[c('counts', 'mirna')])

# Store mirna-mrna results in synapse
write.table(mirna.mrna.cluster.gct, file = 'mirnamrnaCluster.gct', row.names = F, sep = '\t', quote=F)
mirna.mrna.obj = File('mirnamrnaCluster.gct', name = "Differentially Coexpressed mRNA with miRNA (gct format)", parentId = CODE$properties$id)
mirna.mrna.obj = synStore(mirna.mrna.obj, activityName = ActivityName, executed = thisFile, used = ALL_USED_IDs[c('counts', 'mirna','coexp')])

# Store methyl results in synapse
write.table(methyl.cluster.gct, file = 'methylCluster.gct', row.names = F, sep = '\t', quote=F)
methyl.obj = File('methylCluster.gct', name = "Differentially Methylated Probes (gct format)", parentId = CODE$properties$id)
methyl.obj = synStore(methyl.obj, activityName = ActivityName, executed = thisFile, used = ALL_USED_IDs[c('counts', 'methyl')])

# Store methyl-mrna results in synapse
write.table(methyl.mrna.cluster.gct, file = 'methylmrnaCluster.gct', row.names = F, sep = '\t', quote=F)
methyl.mrna.obj = File('methylmrnaCluster.gct', name = "Differentially Coexpressed mRNA with DNA methylation (gct format)", parentId = CODE$properties$id)
methyl.mrna.obj = synStore(methyl.mrna.obj, activityName = ActivityName, executed = thisFile, used = ALL_USED_IDs[c('counts', 'methyl','coexp')])

# Store splicing results in synapse
write.table(splicing.cluster.gct, file = 'splicingCluster.gct', row.names = F, sep = '\t', quote=F)
splicing.obj = File('splicingCluster.gct', name = "Differentially Spliced Junctions (gct format)", parentId = CODE$properties$id)
splicing.obj = synStore(splicing.obj, activityName = ActivityName, executed = thisFile, used = ALL_USED_IDs[c('counts', 'splicing')])

# Store splicing-mrna results in synapse
write.table(splicing.mrna.cluster.gct, file = 'splicingmrnaCluster.gct', row.names = F, sep = '\t', quote=F)
splicing.mrna.obj = File('splicingmrnaCluster.gct', name = "Differentially Coexpressed mRNA with splicing junctions (gct format)", parentId = CODE$properties$id)
splicing.mrna.obj = synStore(splicing.mrna.obj, activityName = ActivityName, executed = thisFile, used = ALL_USED_IDs[c('counts', 'splicing','coexp')])
```

### Source Code
[Source R Markdown](`r thisFile`)