---
title: "Integrated differential expression analysis in conjuction with coexpression at any differentiation states"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(synapseClient)
library(knit2synapse)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'AnalyseDiffExpCoexpNet.Rmd',
                                 parentId = "syn5194922",
                                 entityName = 'Differential Expression Coexpression Networks')

```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
dev.off()
cat("\014")

# Load libraries
library(plyr)
library(dplyr)
library(data.table)
library(stringr)
library(tidyr)
library(reshape2)
library(ggplot2)
library(igraph)

library(matrixStats)
library(biomaRt)

library(knitr)
library(CovariateAnalysis)
library(knit2synapse)
library(synapseClient)
library(githubr) ## Needs the dev branch

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapseStore.params, cache=FALSE, include=FALSE}
parentId = "syn5194922"
ALL_USED_IDs = c()

# Create folder to store results in synapse
ActivityName <- 'Analyse differential expression in conjuction with coexpression between assays'

ThisFileName <- 'AnalyseDiffExpCoexpNet.Rmd'

# Github link
ThisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='diff_exp')

ThisFile <- getPermlink(repository = ThisRepo,
                        repositoryPath=paste0('code/Rmd/', ThisFileName))    

# Populate wiki with results
CODE <- Folder(name = 'Differential Expression Coexpression Networks', parentId = parentId)
CODE <- synStore(CODE)
```

#### Get all interactions from synapse
```{r all.interactions, include=FALSE}
ALL_USED_IDs = c(ALL_USED_IDs, all.interactions = 'syn5643683')
allInteractions = downloadFile('syn5643683') %>%
  dplyr::mutate(Assay = paste(feature.assay,target.assay,sep = '_'))
```

#### Get TFs list from synapse
```{r getTFs, include=FALSE}
ALL_USED_IDs = c(ALL_USED_IDs, TFdb = 'syn5672453')

# Get TF checkpoints db from synapse
TFs = read.table(synGet('syn5672453')@filePath, header=T, sep= '\t', fill = TRUE) %>%
  dplyr::filter(entrez_human != 0) %>%
  dplyr::select(entrez_human) %>%
  setnames('entrez_human','entrezgene') %>%
  dplyr::mutate(Assay = 'TF')

# Get human related mapping
Hs = useMart("ENSEMBL_MART_ENSEMBL", host="www.ensembl.org") # use this one when biomart.org is down
Hs = useDataset("hsapiens_gene_ensembl", Hs)
human_entrz2symbol = getBM(attributes = c("entrezgene","hgnc_symbol"),
                          filters = "entrezgene",                         
                          values = unique(TFs$entrezgene),
                          mart = Hs)

# Convert entrex gene ids to gene symbols
TFs = left_join(TFs, human_entrz2symbol) %>%
  dplyr::select(hgnc_symbol, Assay) %>%
  setnames('hgnc_symbol','feature') %>%
  unique

# Combine the ones from TF mapping of Enrichr
TFs = rbind(TFs,
            dplyr::filter(allInteractions, Assay == "mrna_mrna") %>%
              dplyr::select(feature) %>%  unique %>%
              dplyr::mutate(Assay = 'TF')) %>%
  unique
```

#### Get median counts from synapse
```{r getMedianCount, include=FALSE}
counts.id = 'syn5637772'
ALL_USED_IDs = c(ALL_USED_IDs, counts = counts.id)
counts = downloadFile(counts.id)
counts$Assay[counts$feature %in% TFs$feature] = 'TF'
gc()
```

#### Get differential expression from synapse
```{r download.diff.exp, include=FALSE}
### Download differential expression results from synapse
diffexp.ids = c(mrna = 'syn5013690', mirna = 'syn5014584', 
                methyl = 'syn4527629', splicing = 'syn5049321')
ALL_USED_IDs = c(ALL_USED_IDs, diffexp.ids)

orderedDiffState = c("DE", "EB", "ECTO", "MESO15", "MESO30", "MESO5", "SC")
Comparisons = apply(combn(orderedDiffState, 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get mrna and mirna
diffexp = lapply(diffexp.ids[c('mrna','mirna')], function(id){
  x = downloadFile(id) %>%
    dplyr::select(-one_of('feature')) 
  colnames(x)[1] = 'feature'
  x = x %>%
    dplyr::filter(adj.P.value <= 0.05) %>%
    tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
    dplyr::select(feature, Comparison, adj.P.value, logFC) %>%
    dplyr::filter(Comparison %in% Comparisons)
}) 

# Get methylation differential expression from synapse
diffexp[['methyl']] = downloadFile(diffexp.ids[c('methyl')]) %>%
  dplyr::select(-one_of('feature'))
colnames(diffexp[['methyl']])[1] = 'feature'
diffexp[['methyl']] = diffexp[['methyl']] %>%
  dplyr::filter(adj.P.value <= 0.05) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::select(feature, Comparison, adj.P.value, changeBETA) %>%
    dplyr::filter(Comparison %in% Comparisons)
                
# Get splicing differential expression from synapse
diffexp[['splicing']] = downloadFile(diffexp.ids[c('splicing')]) %>%
  dplyr::select(-one_of('feature'))
colnames(diffexp[['splicing']])[1] = 'feature'
diffexp[['splicing']] = diffexp[['splicing']] %>%
  dplyr::filter(adj.P.value <= 0.05) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::select(feature, Comparison, adj.P.value, changePSI) %>%
    dplyr::filter(Comparison %in% Comparisons)

# Combine list
diffexp = rbindlist(diffexp, idcol = 'Assay', fill = T)
diffexp$Assay[diffexp$feature %in% TFs$feature] = 'TF'

# Split differential expression results based on comparison
diffexp = split(diffexp, diffexp$Comparison)
diffexp = diffexp[Comparisons]
```

#### Get coexpression from synapse
```{r getCoexp, cache=FALSE, include=FALSE}
coexp.id = 'syn5662375'
ALL_USED_IDs = c(ALL_USED_IDs, coexp = coexp.id)
# coexpression = downloadFile(coexp.id)
coexpression = fread('./correlationDiffComparison.tsv', data.table=F, header=T, sep = '\t')

# Mark all TFs
ind = which(coexpression$feature %in% TFs$feature)
coexpression$feature.assay[ind] = 'TF'

ind = which(coexpression$target %in% TFs$feature)
coexpression$target.assay[ind] = 'TF'

coexpression$Assay = paste(coexpression$feature.assay,coexpression$target.assay, sep = '_')
```

#### Filter differential and co-expression results (at an fdr <= 0.05, abs(r) >= 0.7, FC >= 2, absolute Delta(beta) >= 0.35 and absolute Delta(PSI) >= 0.2)
```{r filterCoExp, include=FALSE}
filtered.Exp.Coxp = lapply(names(diffexp), function(comp, diffExp, coExp){
  # Filter differential expression
  diffExp = diffExp[[comp]] %>%
    dplyr::filter((Assay == 'TF' & abs(logFC) >= 1) |
                    (Assay == 'mrna' & abs(logFC) >= 1) |
                    (Assay == 'mirna' & abs(logFC) >= 1) |
                    (Assay == 'methyl' & abs(changeBETA) >= 0.35) |
                    (Assay == 'splicing' & abs(changePSI) >= 0.2)) %>%
    dplyr::filter(adj.P.value <= 0.05)
  
  # Filter co-expression
  coExp = dplyr::select(coExp, one_of(c("feature", "target", "feature.assay",
                                        "target.assay", "dataSource", "Assay",comp))) %>%
    setnames(comp, 'coexpression') %>%
    dplyr::filter(feature %in% diffExp$feature, 
                  target %in% diffExp$feature,
                  abs(coexpression) >= 0.7,
                  feature != target,
                  !(Assay == "mirna_mrna" & coexpression >= 0)) %>%
    dplyr::arrange(desc(abs(coexpression)))
  
  return(list(DE = diffExp, CE = coExp))
}, diffexp, coexpression)
names(filtered.Exp.Coxp) = names(diffexp)
```

#### Convert network files to graphs and store them in synapse
```{r getGraphs, include=FALSE}
activity = Activity(executed = ThisFile, used = as.character(ALL_USED_IDs), name = ActivityName)

# Create folder to store networks in synapse
all.graph.folder = Folder(name = 'Filtered Graphs', parentId = CODE$properties$id)
all.graph.folder = synStore(all.graph.folder)

# Store filtered networks in synapse 
filtered.graph = lapply(names(filtered.Exp.Coxp), function(x, filteredExpCoxp, Counts){
  # Write edges
  CE = filteredExpCoxp[[x]]$CE
  write.table(CE, file = paste0('Edges_',x,'.tsv'), sep = '\t', row.names=F, quote=F)
  obj = File(paste0('Edges_',x,'.tsv'), name = paste('Edges',x), parentId = all.graph.folder$properties$id)
  obj = synStore(obj, activity)
  
  # Write vertices
  DE = left_join(filteredExpCoxp[[x]]$DE, Counts %>% as.data.table)
  write.table(DE, file = paste0('Vertices_',x,'.tsv'), sep = '\t', row.names=F, quote=F)
  obj = File(paste0('Vertices_',x,'.tsv'), name = paste('Vertices',x), parentId = all.graph.folder$properties$id)
  obj = synStore(obj, activity)
  
  return(list(edge = CE, node = DE))
}, filtered.Exp.Coxp, counts)
names(filtered.graph) = names(filtered.Exp.Coxp)
```

#### Extract different modules of genomic signatures
```{r geneSignatures, include=FALSE}
orderedDiffState = c("SC", "DE", "MESO5", "ECTO", "MESO15", "MESO30", "EB")

tmp = data.frame(comp = names(diffexp)) %>%
  dplyr::mutate(diffState = comp) %>%
  tidyr::separate(diffState, into = c('state1', 'state2'), sep = '_vs_')
tmp1 =  dplyr::select(tmp, comp, state2, state1) %>%
  dplyr::rename(state1 = state2, state2 = state1)
allComp = rbind(tmp, tmp1)

modules = list()
relative.control = list()
for (diffState1 in orderedDiffState){
  modules[[diffState1]] = list()
  relative.control[[diffState1]] = list()
  for (diffState2 in orderedDiffState){
    if (diffState1 != diffState2){
      comp = dplyr::filter(allComp, state1 == diffState1 & state2 == diffState2) %>%
        dplyr::select(comp) %>% unlist %>% as.character
      
      DE.feature = filtered.Exp.Coxp[[comp]]$DE %>%
        dplyr::rename(feature.assay = Assay,
                      feature.fdr = adj.P.value,
                      feature.lfc = logFC,
                      feature.cb = changeBETA,
                      feature.cpsi = changePSI)
      
      DE.target = filtered.Exp.Coxp[[comp]]$DE %>%
        dplyr::rename(target = feature,
                      target.assay = Assay,
                      target.fdr = adj.P.value,
                      target.lfc = logFC,
                      target.cb = changeBETA,
                      target.cpsi = changePSI)
      
      CE = left_join(filtered.Exp.Coxp[[comp]]$CE,
                     DE.feature) %>%
        full_join(DE.target)
      
      relative.control[[diffState1]][[diffState2]] = CE
      
      # Get differentially expressed genes
      DE.mods = CE %>% filter(target.assay %in% 'TF' | target.assay %in% 'mrna')  %>%
        dplyr::mutate(S = target.lfc/abs(target.lfc)) %>%
        group_by(S) %>%
        summarise(target = paste(sort(unique(target)), collapse = ',')) %>%
        dplyr::rename(direction = S, genes = target) %>%
        dplyr::mutate(assay = 'all', action = 'none')
      
      # Get intersection modules
      TF.nonTF.mods = CE %>% filter(target.assay %in% 'TF' | target.assay %in% 'mrna')  %>%
        dplyr::mutate(CE.S = coexpression/abs(coexpression),
                      S = target.lfc/abs(target.lfc)) %>%
        dplyr::group_by(feature.assay, CE.S, S) %>%
        summarise(target = paste(sort(unique(target)), collapse = ',')) %>%
        dplyr::rename(assay = feature.assay,
                      action = CE.S,
                      direction = S,
                      genes = target)
      ind = is.na(TF.nonTF.mods$assay)
      TF.nonTF.mods[ind, c('assay','action')] = 'none'
      TF.nonTF.mods = TF.nonTF.mods  %>% filter(!(assay == 'mirna' & action == 1))
      
      # Combine and store module results
      tmp = rbindlist(list(DE.mods, TF.nonTF.mods), use.names=T) %>%
        unique
      
      if (all(c(diffState1, diffState2) == sort(c(diffState1, diffState2)))){
        tmp$action[tmp$action == 1] = 'CONCORDANT'; tmp$action[tmp$action == -1] = 'DISCORDANT'
        tmp$direction[tmp$direction == 1] = 'UP'; tmp$direction[tmp$direction == -1] = 'DOWN'
      } else {
        tmp$action[tmp$action == -1] = 'CONCORDANT'; tmp$action[tmp$action == 1] = 'DISCORDANT'
        tmp$direction[tmp$direction == -1] = 'UP'; tmp$direction[tmp$direction == 1] = 'DOWN'
      }
      
      modules[[diffState1]][[diffState2]] = tmp 
    }
  }
}

# Write all gene sets to synapse
tmp = lapply(modules, rbindlist, idcol = 'tostate', use.names=T) %>% 
  rbindlist(idcol = 'fromstate', use.names=T) %>%
  dplyr::mutate(setname = paste(fromstate, tostate, assay, action, direction, sep='_')) %>%
  dplyr::select(setname, fromstate, tostate, assay, action, direction, genes)

write.table(tmp, file = 'DifferentiationSignatures.tsv', sep = '\t', row.names=F, quote=F)
obj = File('DifferentiationSignatures.tsv', name = 'Differentiation Signatures (gene list)', parentId = CODE$properties$id)
obj = synStore(obj, activity)

# Write relative control matrix to synapse
tmp = lapply(relative.control, rbindlist, idcol = 'ToState') %>% rbindlist(idcol = 'FromState')
write.table(tmp, file = 'relativeControl.tsv', sep = '\t', row.names=F, quote=F)
obj = File('relativeControl.tsv', name = 'Relative control matrix', parentId = CODE$properties$id)
obj = synStore(obj, activity)
```

#### Relative control venn diagrams
```{r relativeControl2}
dev.off()
pdf(file = 'relativeControlVenns.pdf', width = 20, height = 15)
par(mfrow=c(7,7))
allVenns = list()
for (diffState1 in orderedDiffState){
  allVenns[[diffState1]] = list()
  for (diffState2 in orderedDiffState){
    if(diffState1 == diffState2){
      plot.new()
    }else if (diffState1 != diffState2){
      tmp = relative.control[[diffState1]][[diffState2]] %>%
        filter(target.assay == 'mrna' | target.assay == 'TF')
      tmp$feature.assay[is.na(tmp$feature.assay)] = 'All'
      
      allVenns[[diffState1]][[diffState2]] =  tmp %>% 
        dplyr::select(target, feature.assay) %>% 
        rbind(data.frame(target = unique(tmp$target), feature.assay = 'All')) %>%
        unique %>% as.matrix %>% venneuler %>%
        plot(main = paste(diffState1, '-', diffState2, '#', length(unique(tmp$target))))
    }
  }
}
dev.off()

# Write to synapse
obj = File('relativeControlVenns.pdf', name = 'Relative control Venn Diagrams', parentId = CODE$properties$id)
obj = synStore(obj, executed = ThisFile, 
               used = as.character(c(diffexp.ids, coexp.id, counts.id, ALL_USED_IDs['TFdb'])),
               activityName = ActivityName)
```

#### Relative control venn diagrams w/o MESO-15 and 30
```{r relativeControl3}
orderedDiffState2 = c("SC", "DE", "MESO5", "ECTO", "EB")
colors = c(All = 0, TF =0.2, mirna =0.4, methyl = 0.6, splicing = 0.8)

dev.off()
pdf(file = 'relativeControlVenns2.pdf', width = 13, height = 13)
par(mfrow=c(5,5))
allVenns = list()
for (diffState1 in orderedDiffState2){
  allVenns[[diffState1]] = list()
  for (diffState2 in orderedDiffState2){
    if(diffState1 == diffState2){
      plot.new()
    }else if (diffState1 != diffState2){
      tmp = relative.control[[diffState1]][[diffState2]] %>%
        filter(target.assay == 'mrna' | target.assay == 'TF')
      tmp$feature.assay[is.na(tmp$feature.assay)] = 'All'
      
      allVenns[[diffState1]][[diffState2]] =  tmp %>%
        dplyr::select(target, feature.assay) %>% 
        rbind(data.frame(target = unique(tmp$target), feature.assay = 'All')) %>%
        unique %>% as.matrix %>% venneuler %>%
        plot(main = paste(diffState1, '-', diffState2, '#', length(unique(tmp$target))))
    }
  }
}
dev.off()

# Write to synapse
obj = File('relativeControlVenns2.pdf', name = 'Relative control Venn Diagrams (WO MESO15-30)', parentId = CODE$properties$id)
obj = synStore(obj, activity)
```

#### Relative control (Bar charts)
```{r relativeControl4, fig.height=15, fig.width=15}
allBars = list()
for (diffState1 in orderedDiffState){
  allBars[[diffState1]] = list()
  for (diffState2 in orderedDiffState){
    if(diffState1 == diffState2){
      tmp = data.frame(DiffState = diffState1, Control = 'None', count = 0)
      allBars[[diffState1]][[diffState2]] = tmp
    }else if (diffState1 != diffState2){
      tmp = relative.control[[diffState1]][[diffState2]] %>%
        dplyr::filter(target.assay == 'mrna' | target.assay == 'TF') %>%
        dplyr::select(target, feature.assay) %>% dplyr::rename(feature = target, Control = feature.assay) %>%
        group_by(feature) %>% arrange(desc(Control)) %>%
        summarise(Control = paste(sort(unique(Control)), collapse = '_')) %>%
        group_by(Control) %>% summarise(count = n())
      tmp$Control[1] = 'None'
      tmp$count = tmp$count/sum(tmp$count)
      allBars[[diffState1]][[diffState2]] = tmp
    }
  }
  allBars[[diffState1]] = rbindlist(allBars[[diffState1]], idcol = 'DiffState2', use.names = T, fill = T) %>%
    dplyr::mutate(DiffState2 = factor(DiffState2, levels = orderedDiffState),
                  Control = factor(Control))
}
allBars = rbindlist(allBars, idcol = 'DiffState1', use.names = T, fill=T) %>%
  dplyr::mutate(DiffState1 = factor(DiffState1, levels = orderedDiffState))
p = ggplot(allBars, aes(x = Control, y = count)) + geom_bar(aes(fill = Control), stat = 'identity') + coord_flip()
p = p + xlab('Mode of regulation') + ylab('Fraction')
p = p + facet_grid(DiffState1 ~ DiffState2)


pdf(file = 'relativeControlBar.pdf', width = 13, height = 13)
p
dev.off()

p
# Write to synapse
obj = File('relativeControlBar.pdf', name = 'Relative control bar charts', parentId = CODE$properties$id)
obj = synStore(obj, activity)
```

#### Relative control w/o MESO-15 and 30 (Bar charts)
```{r relativeControl5, fig.height=13, fig.width=13}
allBars = list()
for (diffState1 in orderedDiffState2){
  allBars[[diffState1]] = list()
  for (diffState2 in orderedDiffState2){
    if(diffState1 == diffState2){
      tmp = data.frame(DiffState = diffState1, Control = 'None', count = 0)
      allBars[[diffState1]][[diffState2]] = tmp
    }else if (diffState1 != diffState2){
       tmp = relative.control[[diffState1]][[diffState2]] %>%
        dplyr::filter(target.assay == 'mrna' | target.assay == 'TF') %>%
        dplyr::select(target, feature.assay) %>% dplyr::rename(feature = target, Control = feature.assay) %>%
        group_by(feature) %>% arrange(desc(Control)) %>%
        summarise(Control = paste(sort(unique(Control)), collapse = '_')) %>%
        group_by(Control) %>% summarise(count = n())
      tmp$Control[1] = 'None'
      tmp$count = tmp$count/sum(tmp$count)
      allBars[[diffState1]][[diffState2]] = tmp
    }
  }
  allBars[[diffState1]] = rbindlist(allBars[[diffState1]], idcol = 'DiffState2', use.names = T, fill = T) %>%
    dplyr::mutate(DiffState2 = factor(DiffState2, levels = orderedDiffState2),
                  Control = factor(Control))
}
allBars = rbindlist(allBars, idcol = 'DiffState1', use.names = T, fill=T) %>%
  dplyr::mutate(DiffState1 = factor(DiffState1, levels = orderedDiffState2))
p = ggplot(allBars, aes(x = Control, y = count)) + geom_bar(aes(fill = Control), stat = 'identity') + coord_flip()
p = p + xlab('Mode of regulation') + ylab('Fraction')
p = p + facet_grid(DiffState1 ~ DiffState2)

pdf(file = 'relativeControlBar2.pdf', width = 13, height = 13)
p
dev.off()
p

# Write to synapse
obj = File('relativeControlBar2.pdf', name = 'Relative control bar charts (WO MESO15-30)', parentId = CODE$properties$id)
obj = synStore(obj, activity)
```

#### Understanding differnetiation with reference to a differentiation state
```{r sharedControl, fig.width=20, fig.height=15}
pdf(file = 'sharedControlVenn.pdf', width = 20, height = 20)
shared.control = list()
for (diffState in orderedDiffState2){
  shared.control[[diffState]] = list()
  for (assay in c('TF', 'nonTF')){
    shared.control[[diffState]][[assay]] = list()
    for(direction in c('UP', 'DOWN')){
    tmp = lapply(modules[[diffState]], 
                 function(x, assay, direction){
                   x %>%
                     filter(target.assay == assay, S == direction) %>%
                     dplyr::select(Genes) %>%
                     unlist %>%
                     str_split(',') %>%
                     unlist %>%
                     unique
                   }, assay, direction) %>% 
      ldply(.fun = function(x){x = data.frame(x)}) %>%
      filter(.id != 'MESO15' & .id != 'MESO30') %>%
      group_by(x) %>% summarise(.id = paste(unique(sort(.id)), collapse = '_')) %>%
      ungroup %>% group_by(.id) %>% summarise(count = n(), Genes = paste(sort(unique(x)), collapse = ',')) 
    shared.control[[diffState]][[assay]][[direction]] = tmp
    }
  }
}

tmp = lapply(shared.control, function(x){
  lapply(x, function(y){
    rbindlist(y, use.names=T, fill=T, idcol = 'Direction')
  }) %>% rbindlist(use.names=T, fill=T, idcol = 'mrna')
}) %>% rbindlist(use.names=T, fill=T, idcol = 'FromState') %>% 
  dplyr::rename(DiffState = .id)

p = list()
p[[1]] = filter(tmp, mrna == 'nonTF') %>% ggplot(aes(x = DiffState, y = count)) 
p[[1]] = p[[1]] + geom_bar(aes(fill = DiffState), stat = 'identity') + coord_flip()
p[[1]] = p[[1]] + facet_grid(mrna~FromState)

p[[2]] = filter(tmp, mrna == 'TF') %>% ggplot(aes(x = DiffState, y = count)) 
p[[2]] = p[[2]] + geom_bar(aes(fill = DiffState), stat = 'identity') + coord_flip()
p[[2]] = p[[2]] + facet_grid(mrna~FromState)

multiplot(plotlist = p, cols = 1)
dev.off()

multiplot(plotlist = p, cols = 1)
# Write to synapse
obj = File('sharedControlVenn.pdf', name = 'Shared control venn diagram', parentId = CODE$properties$id)
obj = synStore(obj, activity)
```

### Source Code
[Source R Markdown](`r ThisFile`)