---
title: Integrated differential expression analysis in conjuction with coexpression
  at any differentiation states
author: "Thanneer Perumal"
date: '`r date()`'
output: pdf_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(synapseClient)
library(knit2synapse)
library(knitr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'AnalyseDiffExpCoexpNet.Rmd',
                                 parentId = "syn5194922",
                                 entityName = 'Differential Expression Coexpression Networks')

```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

library(synapseClient)
library(knit2synapse)
library(knitr)

synapseLogin()

# Load libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)
library(githubr)

library(matrixStats)
library(biomaRt)
library(ComplexHeatmap)

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapseStore.params, cache=FALSE, include=FALSE}
parentId = "syn5194922"
ALL_USED_IDs = c()

# Create folder to store results in synapse
ActivityName <- 'Analyse differential expression in conjuction with coexpression between assays'

ThisFileName <- 'AnalyseDiffExpCoexpNet.Rmd'

# Github link
ThisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='diff_exp')

ThisFile <- getPermlink(repository = ThisRepo,
                        repositoryPath=paste0('code/Rmd/', ThisFileName))    

# Populate wiki with results
CODE <- Folder(name = 'Differential Expression Coexpression Networks', parentId = parentId)
CODE <- synStore(CODE)
```

#### Get TFs list from synapse
```{r getTFs, include=FALSE}
ALL_USED_IDs = c(ALL_USED_IDs, TFdb = 'syn5672453')

# Get TF checkpoints db from synapse
TFs = read.table(synGet('syn5672453')@filePath, header=T, sep= '\t', fill = TRUE) %>%
  dplyr::filter(entrez_human != 0) %>%
  dplyr::select(entrez_human) %>%
  setnames('entrez_human','entrezgene') %>%
  dplyr::mutate(Assay = 'TF')

# Get human related mapping
Hs = useMart("ENSEMBL_MART_ENSEMBL", host="www.ensembl.org") # use this one when biomart.org is down
Hs = useDataset("hsapiens_gene_ensembl", Hs)
human_entrz2symbol = getBM(attributes = c("entrezgene","hgnc_symbol"),
                          filters = "entrezgene",                         
                          values = unique(TFs$entrezgene),
                          mart = Hs)

# Convert entrex gene ids to gene symbols
TFs = left_join(TFs, human_entrz2symbol) %>%
  dplyr::select(hgnc_symbol, Assay) %>%
  setnames('hgnc_symbol','feature') %>%
  unique

# Combine the ones from TF mapping of Enrichr
# TFs = rbind(TFs,
#             dplyr::filter(allInteractions, Assay == "mrna_mrna") %>%
#               dplyr::select(feature) %>%  unique %>%
#               dplyr::mutate(Assay = 'TF')) %>%
#   unique
```

#### Get differential expression from synapse
```{r download.diff.exp, fig.height=10, fig.width=10, fig.align=`left`}
### Download differential expression results from synapse
diffexp.ids = c(mrna = 'syn5013690', mirna = 'syn5014584', 
                methyl = 'syn4527629', splicing = 'syn5049321')
ALL_USED_IDs = c(ALL_USED_IDs, diffexp.ids)

orderedDiffState = c("DE", "EB", "ECTO", "MESO15", "MESO30", "MESO5", "SC")
Comparisons = apply(combn(orderedDiffState, 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get mrna and mirna
diffexp = lapply(diffexp.ids[c('mrna','mirna')], function(id){
  x = downloadFile(id) %>%
    dplyr::select(-one_of('feature')) 
  colnames(x)[1] = 'feature'
  x = x %>%
    dplyr::filter(adj.P.value <= 0.01, abs(logFC) >= log2(2)) %>%
    tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
    filter(Comparison %in% Comparisons) %>%
    tidyr::separate(Comparison, into = c('from.state', 'to.state'), sep = '_vs_') %>%
    dplyr::select(feature, from.state, to.state, adj.P.value, logFC) 
  
  x1 = x %>%
    dplyr::rename(from.state = to.state,
                  to.state = from.state) %>%
    dplyr::mutate(logFC = -logFC) 
  
  x = rbindlist(list(x, x1), use.names = T, fill = T)
}) 

# Get methylation differential expression from synapse
diffexp[['methyl']] = downloadFile(diffexp.ids[c('methyl')]) %>%
  dplyr::select(-one_of('feature'))
colnames(diffexp[['methyl']])[1] = 'feature'
diffexp[['methyl']] = diffexp[['methyl']] %>%
  dplyr::filter(adj.P.value <= 0.01, abs(changeBETA) >= 0.3) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  filter(Comparison %in% Comparisons) %>%
  tidyr::separate(Comparison, into = c('from.state', 'to.state'), sep = '_vs_') %>%
  dplyr::select(feature, from.state, to.state, adj.P.value, logFC, changeBETA)
x1 = diffexp[['methyl']] %>%
  dplyr::rename(from.state = to.state,
                to.state = from.state) %>%
  dplyr::mutate(logFC = -logFC, changeBETA = -changeBETA) 
diffexp[['methyl']] = rbindlist(list(diffexp[['methyl']], x1), use.names = T, fill = T)
  
# Get splicing differential expression from synapse
diffexp[['splicing']] = downloadFile(diffexp.ids[c('splicing')]) %>%
  dplyr::select(-one_of('feature'))
colnames(diffexp[['splicing']])[1] = 'feature'
diffexp[['splicing']] = diffexp[['splicing']] %>%
  dplyr::filter(adj.P.value <= 0.01, abs(changePSI) >= 0.2) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  filter(Comparison %in% Comparisons) %>%
  tidyr::separate(Comparison, into = c('from.state', 'to.state'), sep = '_vs_') %>%
  dplyr::select(feature, from.state, to.state, adj.P.value, logFC, changePSI)
x1 = diffexp[['splicing']] %>%
  dplyr::rename(from.state = to.state,
                to.state = from.state) %>%
  dplyr::mutate(logFC = -logFC, changePSI = -changePSI) 
diffexp[['splicing']] = rbindlist(list(diffexp[['splicing']], x1), use.names = T, fill = T)

# Combine list
diffexp = rbindlist(diffexp, idcol = 'Assay', fill = T) %>%
  dplyr::mutate(direction = logFC/abs(logFC)) 
diffexp$Assay[diffexp$feature %in% TFs$feature] = 'TF'

# Get differential expression summary
tmp = diffexp %>%
  group_by(Assay, from.state, direction) %>%
  summarise(count = length(unique(feature))) %>%
  spread(Assay, count) %>%
  dplyr::mutate(from.state = factor(from.state, levels = c("SC", "DE", "MESO5", "ECTO", "MESO15", "MESO30" , "EB")),
                direction = factor(direction, labels = c('-1' = 'down', '1' = 'up'))) %>%
  dplyr::arrange(from.state) %>%
  dplyr::select(from.state, direction, TF, mrna, mirna, methyl, splicing) %>%
  dplyr::rename(DiffState = from.state)

writeLines('Number of unique molecular entities that are differentially expressed')
kable(tmp)

ggplot(tmp %>% gather(Assay, count, -DiffState, -direction),
       aes(x = DiffState, y = count, fill = Assay)) + geom_bar(stat = 'identity', position = 'dodge')
```
#### Markers for differentiation states
```{r diffStateMarkers, fig.height=10, fig.width=10, fig.align=`left`}
# Combine MESO15 and MESO30 and take them as MESO state
diffexp.tmp = diffexp %>%
  filter(from.state != 'EB', from.state != 'MESO5',
         to.state != 'EB', to.state != 'MESO5') %>%
  dplyr::mutate(from.state = gsub('15','',from.state),
                to.state = gsub('15','',to.state),
                from.state = gsub('30','',from.state),
                to.state = gsub('30','',to.state))

markers.up = lapply(unique(diffexp.tmp$from.state), function(x){
  ftr = filter(diffexp.tmp, from.state == x, direction == 1) %>%
    group_by(feature, Assay) %>% summarise(count = length(unique(to.state))) %>%
    filter((Assay %in% c('TF', 'mrna', 'mirna', 'splicing') && count == 3) 
           || (Assay %in% c('methyl') && count == 2)) %>%
    dplyr::select(feature) %>% unlist
  for (state in setdiff(unique(diffexp.tmp$from.state), x)){
    ftr = setdiff(ftr, 
                  filter(diffexp.tmp, from.state == state, direction == 1) %>%
                    group_by(feature, Assay) %>% summarise(count = length(unique(to.state))) %>%
                    filter((Assay %in% c('TF', 'mrna', 'mirna', 'splicing') && count == 3) 
                           || (Assay %in% c('methyl') && count == 2)) %>%
                    dplyr::select(feature) %>% unlist)
  }
  filter(diffexp.tmp, from.state == x, feature %in% ftr, direction == 1)
}) %>% 
  rbindlist(use.names = T, fill = T)

markers.down = lapply(unique(diffexp.tmp$from.state), function(x){
  ftr = filter(diffexp.tmp, from.state == x, direction == -1) %>%
    group_by(feature, Assay) %>% summarise(count = length(unique(to.state))) %>%
    filter((Assay %in% c('TF', 'mrna', 'mirna', 'splicing') && count == 3) 
           || (Assay %in% c('methyl') && count == 2)) %>%
    dplyr::select(feature) %>% unlist
  for (state in setdiff(unique(diffexp.tmp$from.state), x)){
    ftr = setdiff(ftr, 
                  filter(diffexp.tmp, from.state == state, direction == -1) %>%
                    group_by(feature, Assay) %>% summarise(count = length(unique(to.state))) %>%
                    filter((Assay %in% c('TF', 'mrna', 'mirna', 'splicing') && count == 3) 
                           || (Assay %in% c('methyl') && count == 2)) %>%
                    dplyr::select(feature) %>% unlist)
  }
  filter(diffexp.tmp, from.state == x, feature %in% ftr, direction == -1)
}) %>% 
  rbindlist(use.names = T, fill = T)

# Get differential expression summary
tmp = rbindlist(list(markers.up, markers.down), use.names = T, fill = T) %>%
  group_by(Assay, from.state, direction) %>%
  summarise(count = length(unique(feature))) %>%
  spread(Assay, count) %>%
  dplyr::mutate(from.state = factor(from.state, levels = c("SC", "DE", "MESO", "ECTO")),
                direction = factor(direction, labels = c('-1' = 'down', '1' = 'up'))) %>%
  dplyr::arrange(from.state) %>%
  dplyr::select(from.state, direction, TF, mrna, mirna, methyl, splicing) %>%
  dplyr::rename(DiffState = from.state)

writeLines('Number of unique molecular entities that are specific to a differentiation state')
kable(tmp)

# Get differential expression summary
tmp = rbindlist(list(markers.up, markers.down), use.names = T, fill = T) %>%
  group_by(Assay, from.state, direction) %>%
  summarise(count = length(unique(feature)),
            feature = paste(unique(feature), collapse = ','))
```

#### Get logCPM counts and covariates and plot
```{r cpm.cov, fig.height=10, fig.width=10, fig.align=`left`}
covariates.id = 'syn5011149'
ALL_USED_IDs = c(ALL_USED_IDs, covariates.id)
covariates = downloadFile(covariates.id)

covariates = covariates[,c('UID', 'Diffname_short', 'Gender', 'Reprogramming_Gene_Combination', 'Cell_Type_of_Origin_Level2')]
covariates = sapply(covariates, as.factor)
rownames(covariates) = covariates[,'UID']

logcpm.id = 'syn5011095'
ALL_USED_IDs = c(ALL_USED_IDs, logcpm.id)
logcpm = downloadFile(logcpm.id)
rownames(logcpm) = logcpm$GeneName
logcpm$GeneName = NULL
logcpm[logcpm<0] = 0

tmp = rbindlist(list(markers.up, markers.down)) %>% 
  filter(Assay %in% c('TF','mrna')) %>% 
  dplyr::select(feature) %>% unlist %>% unique

writeLines('Clustering samples based on extracted signatures')
count = logcpm[tmp, ]
count = scale(count)
count = t(scale(t(count)))

ha = HeatmapAnnotation(covariates[,-(1)])
Heatmap(count, top_annotation = ha, show_row_names = F, show_column_names = F, name = 'logCPM')

writeLines('Clustering of EBs based on extracted signatures')

count = logcpm[tmp, rownames(covariates)[covariates[,'Diffname_short'] == 'EB']]
count = scale(count)
count = t(scale(t(count)))

ha = HeatmapAnnotation(covariates[rownames(covariates)[covariates[,'Diffname_short'] == 'EB'],-(1)])
Heatmap(count, top_annotation = ha, show_row_names = F, show_column_names = F, name = 'logCPM')
```

### Analyse targets of miRNAs, methylation and splicing
#### Get coexpression from synapse
```{r getCoexp, cache=FALSE, eval = FALSE}
coexp.id = 'syn5662375'
ALL_USED_IDs = c(ALL_USED_IDs, coexp = coexp.id)
coexpression = downloadFile(coexp.id) %>%
  tidyr::gather(Comparison, coexpression, 
                -feature, -target, -feature.assay, -target.assay,
                -dataSource, -Assay)
# coexpression = fread('./correlationDiffComparison.tsv', data.table=F, header=T, sep = '\t')

# Mark all TFs
ind = which(coexpression$feature %in% TFs$feature)
coexpression$feature.assay[ind] = 'TF'

ind = which(coexpression$target %in% TFs$feature)
coexpression$target.assay[ind] = 'TF'

coexpression$Assay = paste(coexpression$feature.assay,coexpression$target.assay, sep = '_')

# Filter coexpression
coexpression = coexpression %>%
  filter(feature %in% diffexp$feature, target %in% diffexp$feature, Assay != 'mrna_mrna', Assay != 'mrna_TF') %>%
  ddply(.(Assay), .fun = function(x){
    Assay = unique(x$Assay)
    if (Assay == 'mirna_mrna' || Assay == 'mirna_TF'){
      cutoff = quantile(x$coexpression,0.1)
      x = filter(x, coexpression <= cutoff)
    } else if (Assay == 'splicing_mrna' || Assay == 'splicing_TF') {
      x = x
    } else if (Assay == 'mrna_splicing' || Assay == 'TF_splicing') {
      cutoff = quantile(x$coexpression,0.9, na.rm = T)
      x = filter(x, coexpression >= cutoff)
    } else {
      cutoff = quantile(abs(x$coexpression),0.9, na.rm = T)
      x = filter(x, abs(coexpression) >= cutoff)
    }
  }) %>%
  tidyr::separate(Comparison, c('from.state','to.state'), sep = '_vs_')

tmp1 = coexpression %>%
  dplyr::rename(from.state = to.state,
                to.state = from.state) 

coexpression = rbindlist(list(coexpression, tmp1), use.names = T, fill = T)
rm(list = 'tmp1')
gc()

coexpression = coexpression %>% dplyr::mutate(feature.target = paste(feature, target, sep = '_'))
```

### Source Code
[Source R Markdown](`r ThisFile`)