---
title: "Differential Expression Analysis for eXpress aligned mRNA data with mixed effect modeling performed at indivudal differentiation stages"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('knitr')
library('stringr')
library('knit2synapse')

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

# devtools::install_github("knit2synapse-1", username = "th1vairam", ref= "dev")
# knitToFolderEntity(file = "./DiffExpAnal_mRNA_mixedEffects_EachDiffState_V2.Rmd", 
#                    parentId = 'syn3256504',
#                    entityName = 'Differential Expression Analysis mRNA Mixed Effects EachDiffState', 
#                    overwrite=F)

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameter
COVARIATES_FOLDER = 'syn4557064'
parentId = 'syn3256504'

# Diffstate Names
diffstate = list(DE = 'DE', EB = 'EB', ECTO = 'ECTO', MES05 = 'MESO5', SC = 'SC')
# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Gender', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')
ContCovariates = c('PassageAtThaw', 'PassageAtHarvest')
```
Obtain processed counts, design, weights and covariates matrix from synapse
```{r getdata, cache=TRUE, include=FALSE}
All.Files = synQuery(paste0('select * from file where parentId == "',COVARIATES_FOLDER,'"'))

# Function to download expression counts, 
getData <- function(diffstate, All.Files, FactorCovariates, ContCovariates){
  writeLines('------------------------------------------------------------------------------')
  writeLines(paste('Downloading differentiation state:',as.character(diffstate)))
  writeLines('------------------------------------------------------------------------------')

  # Get count matrix
  PROC.COUNT_OBJ = synGet(All.Files$file.id[grep(paste('Processed logCPM Counts',diffstate), All.Files$file.name)])
  ALL_USED_IDs = PROC.COUNT_OBJ$properties$id
  PROC.COUNT = fread(getFileLocation(PROC.COUNT_OBJ), data.table=FALSE)
  row.names(PROC.COUNT) = PROC.COUNT[,1]
  PROC.COUNT = as.matrix(PROC.COUNT[,-(1)])

  # Get design matrix
  DESIGN_OBJ = synGet(All.Files$file.id[grep(paste('Adjusted Design',diffstate), All.Files$file.name)])
  ALL_USED_IDs = c(ALL_USED_IDs, DESIGN_OBJ$properties$id)
  DESIGN = fread(getFileLocation(DESIGN_OBJ), data.table=FALSE)
  row.names(DESIGN) = DESIGN[,1]
  DESIGN = as.matrix(DESIGN[,-(1)])

  # Get covariates matrix
  COVARIATES_OBJ = synGet(All.Files$file.id[grep(paste('Adjusted Covariates',diffstate), All.Files$file.name)])
  ALL_USED_IDs = c(ALL_USED_IDs, COVARIATES_OBJ$properties$id)
  COVARIATES = read.table(getFileLocation(COVARIATES_OBJ), sep='\t', header=T, row.names=1)
  
  # Convert factor covariates to factors
  FactorCovariates = intersect(colnames(COVARIATES),FactorCovariates)
  ContCovariates = intersect(colnames(COVARIATES), ContCovariates)
  COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
  COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)

  writeLines('------------------------------------------------------------------------------')
  return(list( PROC.COUNT = PROC.COUNT,
               COVARIATES = COVARIATES,              
               DESIGN = DESIGN,
               ALL_USED_IDs = ALL_USED_IDs))
  
}

# Get data from synapse
DATA = lapply(diffstate, getData, All.Files, FactorCovariates, ContCovariates)
```
### Differential Expression Analysis At Each Differentiation Stage
Obtain differentially expressed genes with the following covariates (as obtained from covariates analysis) in the model:
* Differentation Stage: DE
  * Fixed Effects: Cell Line Type, run
  * Random Effects: Donor ID
* Differentation Stage: EB
  * Fixed Effects: Gender, Donor Life Stage
  * Random Effects: Donor ID
* Differentation Stage: ECTO
  * Fixed Effects: run
  * Random Effects: Donor ID
* Differentation Stage: MESO5
  * Fixed Effects: run
  * Random Effects: Donor ID
* Differentation Stage: SC
  * Fixed Effects: Gender, Donor Life Stage, run
  * Random Effects: donor ID
```{r differential.expression, cache=FALSE}
# Function to calculate differential expression at each differentiation stages
getDifferentialExpression <- function(diffstate, DATA){
  writeLines('------------------------------------------------------------------------------')
  writeLines(paste('Performing differential expression for',as.character(diffstate)))
  writeLines('------------------------------------------------------------------------------')
  
  # Post adjusted design matrix
  DESIGN = DATA[[diffstate]]$DESIGN
  print(colnames(DESIGN))
  
  # Estimate sample weights
  GENE.EXPRESSION = voom(DATA[[diffstate]]$PROC.COUNT,
                         design = DESIGN,                         
                         plot=F)
  
  # Calculate correlation between donors
  CORRELATION <- duplicateCorrelation(GENE.EXPRESSION,                                        
                                      block= DATA[[diffstate]]$COVARIATES$Donor_ID)
  
  # Estimate sample weights with block correlation effects
  GENE.EXPRESSION = voom(DATA[[diffstate]]$PROC.COUNT,
                         design = DESIGN,                         
                         block = DATA[[diffstate]]$COVARIATES$Donor_ID,
                         correlation = CORRELATION$cor,
                         plot=F)
  
  # Fit linear model using mixed effects design
  FIT = lmFit(GENE.EXPRESSION,
              block = DATA[[diffstate]]$COVARIATES$Donor_ID,
              correlation = CORRELATION$cor)
  
  # Make contrast to check differential expression between different differentiation stages
  CONT.TO.CHECK = setdiff(colnames(DATA[[diffstate]]$COVARIATES),'run')
  CONT.TO.CHECK = CONT.TO.CHECK[!!sapply(CONT.TO.CHECK, function(x, DESIGN) length(grep(x,colnames(DESIGN))), DESIGN)]
  
  ALL.CONT = c()
  for (name in CONT.TO.CHECK){
    CONT.NAMES <- colnames(DESIGN)[grep(name,colnames(DESIGN))]
    CONT.NAMES <- combn(CONT.NAMES,2)
    CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')
  
    CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
    colnames(CONT) <- sapply(colnames(CONT),
                             function(x, name, diffstate){ 
                               x <- gsub(name,'',x);
                               x <- gsub('-','_vs_',x);
                               x <- paste(diffstate,x,sep='__')
                             }, 
                             name, 
                             diffstate)
    ALL.CONT = cbind(ALL.CONT, CONT)
  }

  # Refit contrasts
  FIT.CONTRAST <- contrasts.fit(FIT,ALL.CONT)
    
  # Estimate moderated t-statistics
  FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
  # Obtain all the differential expession combinations
  DIFF.EXP <- list()
  DIFF.EXP$logFC <- data.frame(row.names = rownames(DATA[[diffstate]]$PROC.COUNT))
  DIFF.EXP$adj.P.Val <- data.frame(row.names = rownames(DATA[[diffstate]]$PROC.COUNT))
  DIFF.EXP$SIG.SETS <- data.frame()

  for (name in colnames(ALL.CONT)){
    tmp <- topTable(FIT.CONTRAST, coef=name, number=dim(DATA[[diffstate]]$PROC.COUNT)[1])    
    DIFF.EXP$logFC[,name] <- tmp[rownames(DIFF.EXP$logFC),'logFC']
    DIFF.EXP$adj.P.Val[,name] <- tmp[rownames(DIFF.EXP$adj.P.Val),'adj.P.Val'] 
  
    DIFF.EXP$SIG.SETS <- rbind(DIFF.EXP$SIG.SETS,
                               getUpDownGenes(DIFF.EXP$adj.P.Val[,name], 
                                              DIFF.EXP$logFC[,name], 
                                              rownames(DIFF.EXP$logFC), 
                                              name, 
                                              FC_CUTOFF = 0))
  }

  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and log2FC >= 1
  DIFF.EXP$SIG.EXP.POS <- DIFF.EXP$adj.P.Val<=0.05 & DIFF.EXP$logFC >= 1
  DIFF.EXP$NUM.SIG.EXP.POS <- colSums(DIFF.EXP$SIG.EXP.POS)
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
  DIFF.EXP$SIG.EXP.NEG <- DIFF.EXP$adj.P.Val<=0.05 & DIFF.EXP$logFC <= -1
  DIFF.EXP$NUM.SIG.EXP.NEG <- colSums(DIFF.EXP$SIG.EXP.NEG)
    
  return(list(DIFF.EXP = DIFF.EXP,
              FIT = FIT,
              FIT.CONTRAST = FIT.CONTRAST,
              GENE.EXPRESSION = GENE.EXPRESSION,
              CORRELATION = CORRELATION))
}

# Compute differential expression at each differentiation stages
DEXP = lapply(diffstate, getDifferentialExpression, DATA)
```
Number of differentially expressed genes at DE stage for adj.P.Val <= 0.05 and logFC >= 1 or <= -1
```{r print.imp.genes.DE}
tmp1 = rownameToFirstColumn(as.data.frame(DEXP$DE$DIFF.EXP$NUM.SIG.EXP.POS),'ContrastNames')
colnames(tmp1) = c('Contrast Names','Up Regulated')
tmp2 = rownameToFirstColumn(as.data.frame(DEXP$DE$DIFF.EXP$NUM.SIG.EXP.NEG),'ContrastNames')
colnames(tmp2) = c('Contrast Names','Down Regulated')

kable(merge(tmp1,tmp2,by = 'Contrast Names', all = T), row.names = F)
```
Number of differentially expressed genes at EB stage for adj.P.Val <= 0.05 and logFC >= 1 or <= -1
```{r print.imp.genes.EB}
tmp1 = rownameToFirstColumn(as.data.frame(DEXP$EB$DIFF.EXP$NUM.SIG.EXP.POS),'ContrastNames')
colnames(tmp1) = c('Contrast Names','Up Regulated')
tmp2 = rownameToFirstColumn(as.data.frame(DEXP$EB$DIFF.EXP$NUM.SIG.EXP.NEG),'ContrastNames')
colnames(tmp2) = c('Contrast Names','Down Regulated')

kable(merge(tmp1,tmp2,by = 'Contrast Names', all = T), row.names = F)
```
Number of differentially expressed genes at ECTO stage for adj.P.Val <= 0.05 and logFC >= 1 or <= -1
```{r print.imp.genes.ECTO}
tmp1 = rownameToFirstColumn(as.data.frame(DEXP$ECTO$DIFF.EXP$NUM.SIG.EXP.POS),'ContrastNames')
colnames(tmp1) = c('Contrast Names','Up Regulated')
tmp2 = rownameToFirstColumn(as.data.frame(DEXP$ECTO$DIFF.EXP$NUM.SIG.EXP.NEG),'ContrastNames')
colnames(tmp2) = c('Contrast Names','Down Regulated')

kable(merge(tmp1,tmp2,by = 'Contrast Names', all = T), row.names = F)
```
Number of differentially expressed genes at MESO5 stage for adj.P.Val <= 0.05 and logFC >= 1 or <= -1
```{r print.imp.genes.MESO5}
tmp1 = rownameToFirstColumn(as.data.frame(DEXP$MESO5$DIFF.EXP$NUM.SIG.EXP.POS),'ContrastNames')
colnames(tmp1) = c('Contrast Names','Up Regulated')
tmp2 = rownameToFirstColumn(as.data.frame(DEXP$MESO5$DIFF.EXP$NUM.SIG.EXP.NEG),'ContrastNames')
colnames(tmp2) = c('Contrast Names','Down Regulated')

kable(merge(tmp1,tmp2,by = 'Contrast Names', all = T), row.names = F)
```
Number of differentially expressed genes at SC stage for adj.P.Val <= 0.05 and logFC >= 1 or <= -1
```{r print.imp.genes.SC}
tmp1 = rownameToFirstColumn(as.data.frame(DEXP$SC$DIFF.EXP$NUM.SIG.EXP.POS),'ContrastNames')
colnames(tmp1) = c('Contrast Names','Up Regulated')
tmp2 = rownameToFirstColumn(as.data.frame(DEXP$SC$DIFF.EXP$NUM.SIG.EXP.NEG),'ContrastNames')
colnames(tmp2) = c('Contrast Names','Down Regulated')

kable(merge(tmp1,tmp2,by = 'Contrast Names', all = T), row.names = F)
```
```{r merge.results}
logFC = join_all(lapply(DEXP, function(x){rownameToFirstColumn(x$DIFF.EXP$logFC, 'GeneNames')}), 
                 by = 'GeneNames', match = 'all')
adjPVal = join_all(lapply(DEXP, function(x){rownameToFirstColumn(x$DIFF.EXP$adj.P.Val, 'GeneNames')}), 
                 by = 'GeneNames', match = 'all')
comparisonSet = rbindlist(lapply(DEXP, function(x){x$DIFF.EXP$SIG.SETS}))
ALL_USED_IDs = as.character(unlist(lapply(DATA, function(x){return(x$ALL_USED_IDs)})))
```
Get pvalue and fold changes of some important genes

Fold change:
```{r FC}
kable(filter(logFC, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```
Adjusted Pvalue:
```{r PVAL}
kable(filter(adjPVal, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```
### Results
Store files in synapse 
```{r synapse.store, include = FALSE, eval=TRUE}
activityName='Differential Expression Analysis of eXpress aligned data with mixed effects model at each differentiation stages'
  
thisFileName <- 'DiffExpAnal_mRNA_mixedEffects_EachDiffState_V2.Rmd'
  
# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='mRNA')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Create results folder
CODE <- Folder('./DiffExpAnal_mRNA_mixedEffects_EachDiffState_V2.Rmd',name = 'Differential Expression Analysis mRNA Mixed Effects EachDiffState',parentId = parentId)
CODE <- synStore(CODE)
      
# Store logFC
write.table(logFC,file='./DiffExpAnal_mRNA_mixedEffects_EachDiffState_logFC.tsv',sep='\t',row.names=F,quote=F)
FC_OBJ <- File('./DiffExpAnal_mRNA_mixedEffects_EachDiffState_logFC.tsv',name = 'logFC',parentId = CODE$properties$id)
FC_OBJ <- synStore(FC_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)
    
# Store adjusted pvalue
write.table(adjPVal,file='./DiffExpAnal_mRNA_mixedEffects_EachDiffState_adjPval.tsv',sep='\t',row.names=F,quote=F)
PVAL_OBJ <- File('./DiffExpAnal_mRNA_mixedEffects_EachDiffState_adjPval.tsv',name = 'Adjusted Pvalue',parentId = CODE$properties$id)
PVAL_OBJ <- synStore(PVAL_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)

# Store differentially expressed genes for comparison
write.table(comparisonSet,file='./DiffExpAnal_mRNA_mixedEffects_EachDiffState_diffGenes.tsv',sep='\t',row.names=F,quote=F)
SIG.SETS_OBJ <- File('./DiffExpAnal_mRNA_mixedEffects_EachDiffState_diffGenes.tsv',name = 'Differentily Expressed Genes comparison list',parentId = CODE$properties$id)
annotations(SIG.SETS_OBJ) <- list(Pvalue='0.05',Foldchage='0')
SIG.SETS_OBJ <- synStore(SIG.SETS_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=thisFile)
```
|  *Results*                                                      |  *SynapseID*                    |
|  -------                                                        |   ---------                     |
|  Log fold change                                                |  `r paste(FC_OBJ$properties$id,FC_OBJ$properties$versionNumber,sep='.')`           |
|  Adjusted pvalues                                               |  `r paste(PVAL_OBJ$properties$id,PVAL_OBJ$properties$versionNumber,sep='.')`         |
|  Differentially expressed genes (with fold change and pvalue)   |  `r paste(SIG.SETS_OBJ$properties$id,SIG.SETS_OBJ$properties$versionNumber,sep='.')` |

### Executed: Source code
`r paste0('[Source R Markdown](',print(thisFile),')')`