---
title: "Differential expression analysis for eXpress aligned mRNA data with mixed effects modeling at each differentiation stages with EB derived ecto scores"
author: "Thanneer Perumal"
date: "`r date()`"
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('RColorBrewer')
library('ggplot2')

library('psych')
library('reshape2')

library('limma')
library('vcd')
library('data.table')
library('plyr')
library('dplyr')

library('knitr')
library('stringr')
library('synapseClient')
library('rGithubClient') ## Needs the dev branch

library(CovariateAnalysis)

synapseLogin()

# knit2synapse::knitToFolderEntity(file = "./DiffExpAnal_mRNA_mixedEffects_EachDiffState.Rmd", 
#                                  entityName = 'Differential Expression Analysis mRNA Mixed Effects EachDiffState', 
#                                  parentId = 'syn5008933', 
#                                  overwrite=F, knitmd = T)

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R", full.names = TRUE)
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r fxns, include=FALSE}
# Function to rotate variables
rotateVars <- function(x, first) {
  others <- setdiff(x, first)
  c(first, others)
}

# Function to make design matrix
makeDesignMatrix <- function(covariates, comparison, covariateColOrder=NA) {
  # Make a design matrix by reordering the column covariates
  # and column-binding each separate matrix for each column together
  # with complete listing of dummy variables (no reference levels)
  # and also remove any linearly dependent columns
  
  vars <- ifelse(is.na(covariateColOrder), 
                 colnames(covariates),
                 covariateColOrder)
  
  newVars <- rotateVars(vars, comparison)
  
  DESIGN = getDesignMatrix(covariates[, newVars], Intercept = F)
  DESIGN = DESIGN$design[, linColumnFinder(DESIGN$design)$indepCols]
  
  return(DESIGN)
}

# Perform differential expression analysis
doDiffExpr <- function(countMat, designMat, weights, block, comparison, diffState) {
  # Helper function to do all pairwise differential expression on a single covariate
  # from the whole covariate matrix
  
  # Make contrast  
  contrastNames <- colnames(designMat)[grep(comparison, colnames(designMat))]
  contrastNames <- combn(contrastNames, 2)
  contrastNames <- apply(contrastNames, 2, paste, collapse='-')
  
  cont <- makeContrasts(contrasts=contrastNames, 
                        levels=colnames(designMat))
  
  colnames(cont) <- sapply(colnames(cont),
                           function(x, diffState){x <- gsub(comparison, '', x);
                                                  x <- gsub('-','_vs_',x);
                                                  x <- paste(diffState, x, sep='__')}, diffState)
  
  # Voom normalisations
  expr <- voom(countMat, design = designMat)
  
  # Calculate correlation between donors
  correlation <- duplicateCorrelation(expr,
                                      block=block)
  
  if(abs(correlation$cor) == 1 || abs(correlation$cor) == 0){
    cat('Random effects correlation cannot be estimated\n')
    # Assign NA values to all results
    foo1 = list()
    
    foo1$logFC = matrix(NA,1,dim(cont)[2]); 
    colnames(foo1$logFC) = colnames(cont)
    rownames(foo1$logFC) = rownames(countMat)[1]
    
    foo1$adj.P.Val = foo1$logFC
    
    foo1$SIG.SETS <- matrix(NA, 1, 4);
    colnames(foo1$SIG.SETS) = c("GeneSymbol", "logFC", "adj.P.value", "Comparison")
    
    foo1$SIG.EXP.POS <- foo1$logFC
    foo1$NUM.SIG.EXP.POS <- foo1$logFC; rownames(foo1$NUM.SIG.EXP.POS) = c()
    foo1$SIG.EXP.NEG <- foo1$logFC
    foo1$NUM.SIG.EXP.NEG <- foo1$logFC; rownames(foo1$NUM.SIG.EXP.NEG) = c()
    
    return(foo1)
  }
  
  expr = voom(countMat, design = designMat, correlation = correlation$cor)
    
  # Fit linear model using mixed effects design
  fit = lmFit(expr, 
              block=block, 
              correlation = correlation$cor)
  
  # Refit contrasts
  fitContrast <- contrasts.fit(fit, cont)
  
  # Estimate moderated t-statistics
  fitContrast <- eBayes(fitContrast)
  
  # Obtain all the differential expession combinations
  foo <- list()
  foo$logFC <- data.frame(row.names = rownames(countMat))
  foo$adj.P.Val <- data.frame(row.names = rownames(countMat))
  foo$SIG.SETS <- data.frame()
  
  for (i in colnames(cont)){
    tmp <- topTable(fitContrast, coef=i, number=dim(countMat)[1])    
    foo$logFC[,i] <- tmp[rownames(foo$logFC),'logFC']
    foo$adj.P.Val[,i] <- tmp[rownames(foo$adj.P.Val),'adj.P.Val'] 
    
    foo$SIG.SETS <- rbind(foo$SIG.SETS,
                          getUpDownGenes(foo$adj.P.Val[,i], foo$logFC[,i], 
                                         rownames(foo$logFC), i, FC_CUTOFF = 0))
  }
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 0
  foo$SIG.EXP.POS <- foo$adj.P.Val<=0.05 & foo$logFC >= 0
  foo$NUM.SIG.EXP.POS <- colSums(foo$SIG.EXP.POS, na.rm = T)
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= 0
  foo$SIG.EXP.NEG <- foo$adj.P.Val<=0.05 & foo$logFC <= 0
  foo$NUM.SIG.EXP.NEG <- colSums(foo$SIG.EXP.NEG, na.rm = T)
  
  return(foo)
}
```

### Download data
```{r setup, include=FALSE, cache=FALSE}
# Input Parameters
SOURCE.FOLDER_ID = 'syn5015149'
ALL.FILES = synQuery(paste0('select Diffname_short,id,fileType from file where parentId == "',SOURCE.FOLDER_ID,'"'))

SYNAPSE_STORE = T
parentId = 'syn5008933'

# Specify factor and continuous covariates pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Race', 'Ethnicity' , 'Gender', 'Disease', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')

ContCovariates = c('PassageAtThaw', 'PassageAtHarvest')

# Differentiation states
DIFF.STATE = c('SC','DE','EB')
```
Obtain processed count, design, weights and covariates matrix from Synapse.
```{r getdata, cache=TRUE, include=FALSE}
downloadData <- function(diffState, allFiles, 
                         factorCovariates, contCovariates){
  
  cat("----\n",
      paste0('#### Differentiation state: ',as.character(diffState), "\n"),
      sep="")
  
  # Get processed counts matrix
  processedCounts.id = allFiles %>% 
    dplyr::filter(file.Diffname_short == diffState, file.fileType == "GenomicMatrix")
  processedCounts.obj = synGet(processedCounts.id$file.id)
  allUsedIDs = processedPSI.obj$properties$id
  processedCounts = read.table(processedCounts.obj@filePath, sep = '\t', header = T, row.names = 1, quote='', check.names = F)
  
  # Get sample covariates
  covariates.id = allFiles %>% 
    dplyr::filter(file.Diffname_short == diffState, file.fileType == "Covariates")
  covariates.obj = synGet(covariates.id$file.id)
  allUsedIDs = c(allUsedIDs, covariates.obj$properties$id)
  covariates = read.table(getFileLocation(covariates.obj), sep = '\t', row.names=1, header=T)
  
  processedCounts = processedCounts[, rownames(covariates)]
  
  # Convert factor covariates to factors
  factorCovariates = intersect(colnames(covariates), factorCovariates)
  covariates[,factorCovariates] = lapply(covariates[,factorCovariates], factor)
  contCovariates = intersect(colnames(covariates), contCovariates)
  covariates[,contCovariates] = lapply(covariates[,contCovariates], as.numeric)
  
  return(list( COUNTS = processedCounts,
               COVARIATES = covariates,
               ALL_USED_IDs = allUsedIDs,
               postAdjustedCovars = str_split(annotations(covariates.obj)$postAdjustCovars, ',')[[1]]))
}

RAW.COUNTS = lapply(DIFF.STATE, downloadData, ALL.FILES,
                    FactorCovariates, ContCovariates)
names(RAW.COUNTS) = DIFF.STATE
```

### Differential expression analysis (at each differentiation state)
In each diffState, for each covariate, we fit a linear model using `limma` with only the variables obtained from the covariate analysis

The design matrix is formed by including all variables for each factor encoded as dummy variables, reordered so that the variable of interest for comparison is first. With this formulation and the available data, many design matrices are degenerate. Hence, we also remove any columns that are linear combinations of other columns. Because of this, models between separate covariates (separate columns in the covariate matrix) cannot be compared, as each comparison is performed with a separately defined model.

```{r diff.exp}
performDiffExpAnal <- function(diffState, rawCounts){
  cat("----\n",
      paste0('#### Differentiation state: ',as.character(diffState), "\n"),
      sep="")
  
  cat("Get differentially expressed genes using limma package with following coefficients in the linear model:\n")
  
  cat("----\n",
      "Get differentially expressed genes using limma package with following coefficients in the linear model:\n",
      paste0('Fixed Effects:',paste(rawCounts[[diffState]]$postAdjustedCovars, collapse = ','), "\n"),
      "Random Effects: Donor_ID\n")
  
  diffExp = list()
  for (comparison in setdiff(rawCounts[[diffState]]$postAdjustedCovars, c("run","lane"))){
    cat("----\n",
        paste0("Performing differential expression analysis for comparison:", comparison,"\n"))
    
    covariateColOrder = rotateVars(rawCounts[[diffState]]$postAdjustedCovars, comparison)
    
    design <- makeDesignMatrix(covariates = rawCounts[[diffState]]$COVARIATES, 
                               comparison = comparison, 
                               covariateColOrder = covariateColOrder)
    
    diffExp[[comparison]] <- doDiffExpr(countMat = rawCounts[[diffState]]$COUNTS, designMat = design,
                                        weights = NULL, block = rawCounts[[diffState]]$COVARIATES$Donor_ID,
                                        comparison = comparison, diffState = diffState)
    
#     cat(paste0('Number of differentially expressed genes between ',
#                comparison,
#                ' at FDR <= 0.05 and logFC >= 0 or logFC <= 0:\n'))
#     
    #     tmp <- data.frame('NUM. OF UP REG. GENES' = t(diffExp[[comparison]]$NUM.SIG.EXP.POS),
    #                       'NUM. OF DOWN REG. GENES' = t(diffExp[[comparison]]$NUM.SIG.EXP.NEG))
    #     
    #     if (dim(tmp)[1] != 0){      
    #       print(kable(tmp))
    #     } else {
    #       cat('Not feasible to calculate \n')
    #     }
  }
  
  allFC <- llply(lapply(diffExp, function(x){return(x$logFC)}),
                 rownameToFirstColumn, colname = "GeneSymbol") 
  
  results = list()
  results$FC <- plyr::join_all(allFC, by = 'GeneSymbol', match = 'all')
  
  allPVal <- llply(lapply(diffExp, function(x){ return(x$adj.P.Val)}),
                   rownameToFirstColumn, colname="GeneSymbol")
  
  results$PVAL <- plyr::join_all(allPVal, by = 'GeneSymbol', match = 'all')
  
  allSig <- llply(lapply(diffExp, function(x){ 
    return(x$SIG.EXP.POS + x$SIG.EXP.NEG)
  }), rownameToFirstColumn, colname="GeneSymbol")
  
  results$SIG <- join_all(allSig, by = 'GeneSymbol', match = 'all')
  
  results$SIG.SETS <- rbindlist(lapply(diffExp, function(x){ return(as.data.frame(x$SIG.SETS)) }))

  return(results)
}
DIFF.EXP = lapply(DIFF.STATE, performDiffExpAnal, RAW.COUNTS)
names(DIFF.EXP) = DIFF.STATE
```

### Store files in Synapse
Store logFC, adjusted p-values and differentially expressed genes.
```{r synapse.store, include = FALSE, eval=TRUE, cache=FALSE}
activityName='Differential expression analysis of eXpress aligned mRNA with mixed effects model at each differentiation states'

thisFileName <- 'DiffExpAnal_mRNA_mixedEffects_EachDiffState.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName="mRNA")

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Create folder to store the results and populate analysis wiki
CODE <- Folder(name = 'Differential Expression Analysis mRNA Mixed Effects EachDiffState', parentId = parentId)
CODE <- synStore(CODE)

storeFilesInSynapse <- function(diffState, diffExp, rawCounts, code) {
  print(diffState)
  
  diffExp = diffExp[[diffState]]
  rawCounts = rawCounts[[diffState]]
  
  # Store logFC
  write.table(diffExp$FC, file= paste0('./DiffExpAnal_mRNA_mixedEffects_',diffState,'_logFC.tsv'),
              sep='\t',row.names=F,quote=F)
  
  fc.obj <- File(paste0('./DiffExpAnal_mRNA_mixedEffects_',diffState,'_logFC.tsv'),
                 name = paste(diffState,'logFC'), 
                 parentId = code$properties$id)
  synSetAnnotations(fc.obj) <- list(dataType="mRNA", 
                                    fileType="FoldChangeMatrix", 
                                    Diffname_short=diffState)
  
  fc.obj <- synStore(fc.obj, used = rawCounts$ALL_USED_IDs, activityName = activityName, executed=thisFile)
  
  # Store adjusted pvalue
  write.table(diffExp$PVAL,file=paste0('./DiffExpAnal_mRNA_mixedEffects_',diffState,'_adjPval.tsv'),
              sep='\t', row.names=F, quote=F)
  
  pval.obj <- File(paste0('./DiffExpAnal_mRNA_mixedEffects_',diffState,'_adjPval.tsv'),
                   name = paste(diffState,'Adjusted Pvalue'), parentId = CODE$properties$id)
  synSetAnnotations(pval.obj) <- list(dataType="mRNA",
                                      fileType="PvalueMatrix", 
                                      Diffname_short=diffState)
  pval.obj <- synStore(pval.obj, used = rawCounts$ALL_USED_IDs, activityName = activityName, executed=thisFile)
  
  # Store differentially spliced junctions for comparison
  write.table(diffExp$SIG.SETS, file=paste0('./DiffExpAnal_mRNA_mixedEffects_',diffState,'_diffGenes.tsv'),
              sep='\t', row.names=F, quote=F)
  
  sig.sets.obj <- File(paste0('./DiffExpAnal_mRNA_mixedEffects_',diffState,'_diffGenes.tsv'),
                       name = paste(diffState,'Differentially expressed genes comparison list'),
                       parentId = CODE$properties$id)
  
  synSetAnnotations(sig.sets.obj) <- list(dataType="mRNA", 
                                          fileType="DifferentialmRNAExpressionList", 
                                          Diffname_short=diffState,
                                          q = 0.05, logFC = 0)
  sig.sets.obj <- synStore(sig.sets.obj, 
                           used = rawCounts$ALL_USED_IDs, 
                           activityName = activityName, 
                           executed=thisFile)
  
  return(list(FC = fc.obj$properties$id, 
              PVAL = pval.obj$properties$id, 
              SIG.SETS = sig.sets.obj$properties$id))
}

STORED.FILES = sapply(DIFF.STATE, storeFilesInSynapse, 
                      DIFF.EXP, RAW.COUNTS,CODE)
```
Related results can be found here
```{r results.files}
kable(STORED.FILES)
```  
### Executed: Source code
`r paste0('[Source R Markdown](',print(thisFile),')')`