---
title: "Differential expression analysis for eXpress aligned mRNA data with mixed effects modeling at SC state with EB derived ecto scores"
author: "Thanneer Perumal"
date: "`r date()`"
---
```{r knit2syn, eval = F, echo = T}
library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./DiffExpAnal_mRNA_mixedEffects_SC_withECTOScores.Rmd",
                                 entityName = 'Differential Expression Analysis mRNA Mixed Effects SC WithECTOScores',
                                 parentId = 'syn5008933')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library(synapseClient)
library(knitr)
library(githubr)

library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(glmnet)

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r fxns, include=FALSE}
# Function to rotate variables
rotateVars <- function(x, first) {
  others <- setdiff(x, first)
  c(first, others)
}

# Function to make design matrix
makeDesignMatrix <- function(covariates, comparison, covariateColOrder=NA) {
  # Make a design matrix by reordering the column covariates
  # and column-binding each separate matrix for each column together
  # with complete listing of dummy variables (no reference levels)
  # and also remove any linearly dependent columns
  
  vars <- ifelse(is.na(covariateColOrder), 
                 colnames(covariates),
                 covariateColOrder)
  
  newVars <- rotateVars(vars, comparison)
  
  DESIGN = getDesignMatrix(covariates[, newVars], Intercept = F)
  DESIGN = DESIGN$design[, linColumnFinder(DESIGN$design)$indepCols]
  
  return(DESIGN)
}

# Perform differential expression analysis
doDiffExpr <- function(countMat, designMat, weights, block, comparison, diffState) {
  # Helper function to do all pairwise differential expression on a single covariate
  # from the whole covariate matrix
  
  # Make contrast  
  contrastNames <- colnames(designMat)[grep(comparison, colnames(designMat))]
  # contrastNames <- combn(contrastNames, 2)
  # contrastNames <- apply(contrastNames, 2, paste, collapse='-')
  # 
  cont <- makeContrasts(contrasts=contrastNames, 
                        levels=colnames(designMat))
  
  # Voom normalisations
  expr <- voom(countMat, design = designMat)
  
  # Calculate correlation between donors
  correlation <- duplicateCorrelation(expr,
                                      block=block)
  
  if(abs(correlation$cor) == 1 || abs(correlation$cor) == 0){
    cat('Random effects correlation cannot be estimated\n')
    # Assign NA values to all results
    foo1 = list()
    
    foo1$logFC = matrix(NA,1,dim(cont)[2]); 
    colnames(foo1$logFC) = colnames(cont)
    rownames(foo1$logFC) = rownames(countMat)[1]
    
    foo1$adj.P.Val = foo1$logFC
    
    foo1$SIG.SETS <- matrix(NA, 1, 4);
    colnames(foo1$SIG.SETS) = c("GeneSymbol", "logFC", "adj.P.value", "Comparison")
    
    foo1$SIG.EXP.POS <- foo1$logFC
    foo1$NUM.SIG.EXP.POS <- foo1$logFC; rownames(foo1$NUM.SIG.EXP.POS) = c()
    foo1$SIG.EXP.NEG <- foo1$logFC
    foo1$NUM.SIG.EXP.NEG <- foo1$logFC; rownames(foo1$NUM.SIG.EXP.NEG) = c()
    
    return(foo1)
  }
  
  expr = voom(countMat, design = designMat, correlation = correlation$cor)
    
  # Fit linear model using mixed effects design
  fit = lmFit(expr, 
              block=block, 
              correlation = correlation$cor)
  
  # Refit contrasts
  fitContrast <- contrasts.fit(fit, cont)
  
  # Estimate moderated t-statistics
  fitContrast <- eBayes(fitContrast)
  
  # Obtain all the differential expession combinations
  foo <- list()
  foo$logFC <- data.frame(row.names = rownames(countMat))
  foo$adj.P.Val <- data.frame(row.names = rownames(countMat))
  foo$SIG.SETS <- data.frame()
  
  for (i in colnames(cont)){
    tmp <- topTable(fitContrast, coef=i, number=dim(countMat)[1])    
    foo$logFC[,i] <- tmp[rownames(foo$logFC),'logFC']
    foo$adj.P.Val[,i] <- tmp[rownames(foo$adj.P.Val),'adj.P.Val'] 
    
    foo$SIG.SETS <- rbind(foo$SIG.SETS,
                          getUpDownGenes(foo$adj.P.Val[,i], foo$logFC[,i], 
                                         rownames(foo$logFC), i, FC_CUTOFF = 0))
  }
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 0
  foo$SIG.EXP.POS <- foo$adj.P.Val<=0.05 & foo$logFC >= 0
  foo$NUM.SIG.EXP.POS <- colSums(foo$SIG.EXP.POS, na.rm = T)
  
  # Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= 0
  foo$SIG.EXP.NEG <- foo$adj.P.Val<=0.05 & foo$logFC <= 0
  foo$NUM.SIG.EXP.NEG <- colSums(foo$SIG.EXP.NEG, na.rm = T)
  
  return(foo)
}
```

### Download data
```{r setup, include=FALSE, cache=FALSE}
# Input Parameters
SOURCE.FOLDER_ID = 'syn7268440'
ALL.FILES = synQuery(paste0('select Diffname_short,id,fileType from file where parentId == "',SOURCE.FOLDER_ID,'"'))

SYNAPSE_STORE = T
parentId = 'syn5008933'

# Specify factor and continuous covariates pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Race', 'Ethnicity' , 'Gender', 'Disease', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')

ContCovariates = c('PassageAtThaw', 'PassageAtHarvest', 'ecto.scores')
```
Obtain processed counts, design, weights and covariates matrix from Synapse
```{r getdata, cache=TRUE, include=FALSE}
# Get processed counts matrix
PROC.COUNTS.ID = ALL.FILES %>%
  dplyr::filter(file.Diffname_short == 'SC', file.fileType == "GenomicMatrix")
ALL.USED.IDs = PROC.COUNTS.ID$file.id
PROC.COUNTS = read.table(synGet(PROC.COUNTS.ID$file.id)@filePath,
                         sep = '\t', header = T, row.names = 1, quote='', check.names = F)

# Get sample covariates
COV.ID = ALL.FILES %>%
  dplyr::filter(file.Diffname_short == 'SC', file.fileType == "Covariates")
ALL.USED.IDs = c(ALL.USED.IDs, COV.ID$file.id)
COV = read.table(synGet(COV.ID$file.id)@filePath, sep = '\t', row.names=1, header=T)
  
# Convert factor covariates to factors
FactorCovariates = intersect(colnames(COV), FactorCovariates)
COV[,FactorCovariates] = lapply(COV[,FactorCovariates], factor)
ContCovariates = intersect(colnames(COV), ContCovariates)
COV[,ContCovariates] = lapply(COV[,ContCovariates], as.numeric)
  
# Get covariates to adjust from synapse
postAdjustedCovars = c('ecto.scores',
                       str_split(annotations(synGet(COV.ID$file.id))$postAdjustCovars, ',')[[1]]) %>%
  unique

# Remove few SC and EB samples form this analysis for validation purpose
METADATA_OBJ = synTableQuery(paste('SELECT * FROM','syn3156503',sep=' '))
ALL.USED.IDs <- c(ALL.USED.IDs, METADATA_OBJ@schema)
METADATA = METADATA_OBJ@values

CELL.LINE.MAPPING = data.frame(UID = rownames(COV),
                               ecto.scores = COV$ecto.scores) %>%
  left_join(METADATA %>%
              dplyr::select(UID, Diffname_short, C4_Cell_Line_ID))
```

### Differential expression analysis (wrt ecto.scores)
In each diffState, for each covariate, we fit a linear model using `limma` with only the variables obtained from the covariate analysis. Here primary variable of interest is ecto.scores
```{r diff.exp}
ALL.DEXP = list()
ALL.AUC = c()
for(ncount in 1:5){
  
  train.id = sample(unique(CELL.LINE.MAPPING$C4_Cell_Line_ID), 30)
  test.id = setdiff(unique(CELL.LINE.MAPPING$C4_Cell_Line_ID), train.id)
  
  train.UID = CELL.LINE.MAPPING$UID[CELL.LINE.MAPPING$C4_Cell_Line_ID %in% train.id]
  test.UID = CELL.LINE.MAPPING$UID[CELL.LINE.MAPPING$C4_Cell_Line_ID %in% test.id]

  cat("Get differentially expressed genes using limma package with following coefficients in the linear model:\n")
  cat("----\n",
      "Get differentially expressed genes using limma package with following coefficients in the linear model:\n",
      paste0('Fixed Effects:',paste(postAdjustedCovars, collapse = ','), "\n"),
      "Random Effects: Donor_ID\n")
  
  # Get desing matrix
  DESIGN.MAT <- getDesignMatrix(COV[train.UID, postAdjustedCovars, drop = F], Intercept = F)
  DESIGN.MAT = DESIGN.MAT$design[,linColumnFinder(DESIGN.MAT$design)$indepCols,drop=F]

  # Voom normalisations
  EXPR <- voom(PROC.COUNTS[,train.UID], design = DESIGN.MAT)
  
  # Calculate correlated effects of donors
  CORRELATION <- duplicateCorrelation(EXPR, block = COV[train.UID, 'Donor_ID'])

  # Re-estimate voom normalisations
  EXPR <- voom(PROC.COUNTS[,train.UID], design = DESIGN.MAT, correlation = CORRELATION$cor)
  
  # Fit linear model using mixed effects design
  FIT = lmFit(EXPR)

  # Make contrast  
  CONT <- makeContrasts(contrasts = 'ecto.scores', levels=colnames(DESIGN.MAT))

  # Refit contrasts
  FIT.CONT <- contrasts.fit(FIT, CONT)

  # Estimate moderated t-statistics
  FIT.CONT <- eBayes(FIT.CONT)
  
  # Get differentially expressed genes
  DEXP <- topTable(FIT.CONT, coef = 1, number = dim(PROC.COUNTS)[1]) %>%
    rownameToFirstColumn('hgnc_symbol')

  DEXP %>%
    dplyr::mutate(Comparison = 'ecto.scores') %>%
    filter(adj.P.Val <= 0.05) %>%
    summarise(count = n(), Genes = paste(unique(hgnc_symbol), collapse = ',')) %>%
    kable
  
  ALL.DEXP[[ncount]] = DEXP

  # Predict ecto.scores using genomic features
  x.data = EXPR$E[DEXP$hgnc_symbol[DEXP$adj.P.Val <= 0.05],rownames(DESIGN.MAT)] %>% t %>% scale
  y.data = DESIGN.MAT[,'ecto.scores']

  fit.cv = cv.glmnet(x.data, y.data, type.measure = 'class')
  
  # Voom normalisations
  ALL.EXPR <- voom(PROC.COUNTS, design = NULL)
  x.test = ALL.EXPR$E[DEXP$hgnc_symbol[DEXP$adj.P.Val <= 0.05],test.UID] %>% t %>% scale
  y.test = COV[test.UID,'ecto.scores']
  y.test[y.test<0.5] = 0;y.test[y.test>=0.5] = 1
  
  fit.test = predict.cv.glmnet(fit.cv, newx = x.test, s = "lambda.min", type = 'response')
  
  ALL.AUC[ncount] = glmnet::auc(y.test, fit.test)
}
boxplot(ALL.AUC, main = 'AUC')

genes = lapply(ALL.DEXP, function(x){
  x$hgnc_symbol[x$adj.P.Val <= 0.05]
}) %>%
  Reduce(intersect,.)
# coeff = as.matrix(coef.glmnet(fit.cv, s = 'lambda.min'))   
# 
#   DESIGN.MAT = DESIGN.MAT[order(DESIGN.MAT[,'ecto.scores']),]
#   ha = HeatmapAnnotation(DESIGN.MAT[,'ecto.scores', drop = F])
#   Heatmap(EXPR$E[DEXP$hgnc_symbol[DEXP$adj.P.Val <= 0.05],rownames(DESIGN.MAT)] %>% t %>% scale %>% t,
#           top_annotation = ha,
#           cluster_columns = F,
#           col = colorRamp2(c(2,0,-2), c('blue','white','red')),
#           name = 'Scaled logCPM')
# 

```

### Store files in Synapse
Store logFC, adjusted p-values and differentially expressed genes.
```{r synapse.store, include = FALSE, eval=TRUE, cache=FALSE}
activityName='Differential expression analysis of eXpress aligned mRNA with mixed effects model at each differentiation states'

thisFileName <- 'DiffExpAnal_mRNA_mixedEffects_SC_withECTOScores.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName="mRNA")

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Create folder to store the results and populate analysis wiki
CODE <- Folder(name = 'Differential Expression Analysis mRNA Mixed Effects SC WithECTOScores', parentId = parentId)
CODE <- synStore(CODE)

save(list=ls, file = 'Results.RData')
obj = File(file = 'Results.RData', name = 'Results', parnentId = CODE$properties$id)
obj = synStore(obj, used = ALL.USED.IDs, executed = thisFile, activityName = activityName)
```