---
title: "Identifying specific signatures of differentiation states from all molecular assays (i.e., mRNA, miRNA, DNA methylation probes and splicing junctions)"
author: "Thanneer Perumal"
date: '`r date()`'
output: html_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'DiffStateSpecificSignatures.Rmd',
                                 parentId = "syn5194922",
                                 entityName = 'Differentiation State Specific Signatures')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
library(matrixStats)

library(fpc)
library(WGCNA)
library(flashClust)

library(circlize)
colfunc <- colorRampPalette(c("white", "red"))

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapseStore.params}
parentId = "syn5194922"
ALL_USED_IDs = c()

ActivityName <- 'Get diff state specific signatures from all three different assays'

# Get the latest commit of this code from github
thisFileName <- 'DiffStateSpecificSignatures.Rmd'

thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='diff_exp')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Populate wiki with results
CODE <- Folder(name = 'Differentiation State Specific Signatures', parentId = parentId)
CODE <- synStore(CODE)
```

```{r set.thresholds}
pval.th <- 0.01

mrna.lfc.th <- log2(2)
mirna.lfc.th <- log2(2)
methyl.changeBeta.th <- 0.3
splicing.changePSI.th <- 0.2
```

### Data download from synapse
#### Get raw logCPM counts from synapse
```{r getRawCount, include=FALSE}
expr.id = c(mrna = 'syn5011095', mirna = 'syn5014454', methyl = 'syn4487642', splicing = 'syn5048714')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(expr.id))
expr = lapply(expr.id, downloadFile)
```

#### Get covariates from synapse
```{r getRawCount, include=FALSE}
covariates.id = c(mrna = 'syn5011149', mirna = 'syn5014460', methyl = 'syn4487669', splicing = 'syn5048719')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(covariates.id))
covariates = lapply(covariates.id, downloadFile)
```

#### Get differential expression results from synapse
```{r diffexp}
### Download differential expression results from synapse
diffexp.id = c(mrna = 'syn5013690', mirna = 'syn5014584', methyl = 'syn4527629', splicing = 'syn5049321')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(diffexp.id))

# Diff states in a sensible, quasi time-based ordering
orderedDiffState = c("SC", "MESO5", "DE", "MESO15", "MESO30", "ECTO", "EB")
Comparisons = apply(combn(sort(orderedDiffState), 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get mrna differential expression results from synapse
diffexp = list()
diffexp$mrna = downloadFile(diffexp.id['mrna']) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mrna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Get mirna differential expression results from synapse
diffexp$mirna = downloadFile(diffexp.id['mirna']) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mirna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Get methyl differential expression results from synapse
diffexp$methyl = downloadFile(diffexp.id['methyl']) %>%
  dplyr::select(-one_of('feature')) %>%
  dplyr::rename(feature = methProbeIDs) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(changeBETA) >= methyl.changeBeta.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC, changeBETA) 

# Get splicing differential expression results from synapse
diffexp$splicing = downloadFile(diffexp.id['splicing']) %>%
  dplyr::select(-one_of('feature')) %>%
  plyr::rename(c('Minor-Isoform' = 'feature')) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(changePSI) >= splicing.changePSI.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC, changePSI) 

diffexp = rbindlist(diffexp, use.names = T, fill = T, idcol = 'Assay') %>%
  tidyr::separate(Comparison, c('from.state','to.state'), sep = '_vs_')

diffexp = rbindlist(list(diffexp,
                         diffexp %>%
                           dplyr::mutate(
                             from.state = diffexp$to.state,
                             to.state = diffexp$from.state,
                             logFC = -logFC, changeBETA = -changeBETA, changePSI = -changePSI)),
                    use.names = T, fill =T) %>%
  dplyr::mutate(Direction = logFC/-abs(logFC))
```

#### Get co-expression from synapse
```{r coexpp}
coexp.id = 'syn5662375'
ALL_USED_IDs = c(ALL_USED_IDs, coexp.id)
coexp = downloadFile(coexp.id) %>%
  filter(feature.assay != 'splicing', target.assay != 'splicing') %>%
  gather(comparison, coexpression, -feature, -target, -feature.assay, -target.assay, -dataSource, -Assay) %>% 
  dplyr::select(-Assay,-dataSource) %>%
  separate(comparison, c('from.state', 'to.state'), sep = '_vs_') %>%
  dplyr::mutate(interaction.assay = paste(feature.assay, target.assay, sep = '_')) %>%
  dplyr::filter(!(interaction.assay == 'mirna_mrna' & coexpression > 0))

writeLines('80 percentile values for each interaction assay')
kable(coexp %>%
  group_by(interaction.assay) %>%
  summarise(Qntl_80P = quantile(abs(coexpression), 0.8, na.rm = T)))

coexp = coexp %>%
  filter((interaction.assay == 'methyl_mirna' & abs(coexpression) >= 0.45) |
           (interaction.assay == 'methyl_mrna' & abs(coexpression) >= 0.38) |
           (interaction.assay == 'mirna_mrna' & abs(coexpression) >= 0.48) |
           (interaction.assay == 'mrna_mirna' & abs(coexpression) >= 0.47) |
           (interaction.assay == 'mrna_mrna' & abs(coexpression) >= 0.45))

coexp = rbindlist(list(coexp,
                       coexp %>%
                         dplyr::rename(from.state = to.state, to.state = from.state)),
                  use.names = T, fill = T)
```

### Differential expression analysis summary 
```{r combine}
writeLines('Unique entities that are differentially expressed in our data are')
kable(group_by(diffexp, from.state, Assay, Direction) %>%
        summarise(count = length(unique(feature))) %>%
        tidyr::spread(from.state, count) %>%
        dplyr::mutate(Direction = factor(Direction, labels = c('-1' = 'Down', '1' = 'Up'))))  

# Combine diffexp and coexpression
all.int = right_join(coexp,
                     diffexp %>%
                     dplyr::rename(feature.assay = Assay, feature.fdr = adj.P.value, feature.lfc = logFC, 
                                   feature.cb = changeBETA, feature.psi = changePSI, feature.direction = Direction)) %>%
  left_join(diffexp %>%
              dplyr::rename(target = feature, target.assay = Assay, target.fdr = adj.P.value, target.lfc = logFC, 
                            target.cb = changeBETA, target.psi = changePSI, target.direction = Direction)) %>%
  dplyr::filter(!(feature.assay %in% c('mirna', 'methyl') & is.na(target.assay)))

writeLines('Unique entities that are differentially expressed and coexpressed in the same comparison in our data are')
kable(group_by(all.int, from.state, feature.assay, feature.direction) %>%
        summarise(count = length(unique(feature))) %>%
        tidyr::spread(from.state, count) %>%
        dplyr::rename(Direction = feature.direction, Assay = feature.assay) %>%
        dplyr::mutate(Direction = factor(Direction, labels = c('-1' = 'Down', '1' = 'Up'))) %>%
        dplyr::filter(Assay %in% c('mirna','methyl')))
```

### Differentiation stages
Clustering samples based on gene signatures of differentiation stages from GO 
```{r mesendo.test, fig.height= 20, fig.width=15}
# These signatures are 
DE.Sig = c('ONECUT1', 'OTX2', 'GATA6', 'NKX2-1', 'GATA4', 'BPTF', 'HHEX', 'SOX7', 'NKX2-5', 'PAX9', 'MMP14', 'HDAC1')
MESO.Sig = c('BTK', 'FOXC1', 'TEAD2', 'LEF1', 'T', 'SRF', 'FOXH1', 'KDM6A', 'IKZF1', 'SMAD1', 'TBX1', 'FOXC2', 'SNAI1',
             'ECSIT', 'HNF1A', 'HTT', 'HAND1', 'TCF15', 'WNT5A', 'ETV2', 'IRX3', 'EYA1', 'EYA2', 'KDM6B', 'ZFPM2', 'ZIC2',
             'CITED2', 'LHX2', 'TP63', 'HES7', 'OSR1', 'CHURC1', 'FOXF1', 'TBX6')
ECTO.Sig = c('ZBTB17', 'GRHL3', 'ZBTB7B', 'ERF')
SC.Sig = c('ZSCAN10', 'LDB1', 'ZFP36L2', 'PAX8', 'DPPA4', 'ZNF358', 'NFYA', 'KDM2B', 'PBX1', 'PRDM14', 'CNOT1', 'HES5',
           'CNOT2', 'CNOT3', 'EPAS1', 'PRO')

markers = rbind(data.frame(diff.state = 'DE', signature = DE.Sig),
                data.frame(diff.state = 'MESO', signature = MESO.Sig),
                data.frame(diff.state = 'ECTO', signature = ECTO.Sig),
                data.frame(diff.state = 'SC', signature = SC.Sig))

tmp = filter(expr$mrna, GeneName %in% markers$signature)
rownames(tmp) = tmp$GeneName
tmp$GeneName = NULL

markers = filter(markers, signature %in% rownames(tmp))
rownames(markers) = markers$signature

tmp1 = covariates$mrna
rownames(tmp1) = tmp1$UID
ha = HeatmapAnnotation(tmp1[,'Diffname_short', drop = F])

Heatmap(tmp, top_annotation = ha, split = markers[rownames(tmp), 'diff.state'])
```
From the above heatmap we only get four clusters of differentiation stages. Especially ECTO, SC, EB-MESO15-MESO30, and DE-MESO5 clusters. Therefore, for extracting unique signatures of differentiation stages, we are using this four clusters, excluding EB (which is potentially hypothesised to be a mixture of SC and all germ layers)

### Indentify entities that are differentially expressed between one diff state and all the other diff states (called signatures here after)
Here four groups SC, MESENDODERM, MESODERM and ECTODERM are considered
```{r extract.specific.ent, fig.height=12, fig.width=15}
# Seperate gene sets based on transitions
specific.signatures = all.int %>% 
  dplyr::select(feature, feature.assay, feature.direction, from.state, to.state) %>% 
  unique %>% 
  dplyr::mutate(from.state = gsub('MESO5','MESENDO',from.state), from.state = gsub('DE','MESENDO',from.state),
                to.state = gsub('MESO5','MESENDO',to.state), to.state = gsub('DE','MESENDO',to.state),
                from.state = gsub('MESO15','MESO',from.state), from.state = gsub('MESO30','MESO',from.state),
                to.state = gsub('MESO15','MESO',to.state), to.state = gsub('MESO30','MESO',to.state)) %>%
  dplyr::filter(from.state != 'EB', to.state != 'EB', !(from.state == 'MESO' & to.state == 'MESO'), !(from.state == 'MESENDO' & to.state == 'MESENDO')) %>%
  group_by(feature, feature.assay, feature.direction, from.state) %>%
  summarise(ncount = length(unique(to.state)), to.state = paste(unique(to.state), collapse = ',')) %>%
  filter((ncount == 3 & feature.assay != 'methyl') | (ncount == 2 & feature.assay == 'methyl')) %>%
  data.frame

splicing.annot = expr$splicing[,c(1, 227:236)]
expr$splicing = expr$splicing[,1:226]

for (assay in unique(specific.signatures$feature.assay)){
  tmp = expr[[assay]]
  rownames(tmp) = tmp[,1]
  tmp[,1] = NULL
  tmp[is.na(tmp)] = 0
  
  tmp = tmp[rownames(tmp) %in% unique(specific.signatures$feature),]
  tmp1 = covariates[[assay]]
  rownames(tmp1) = tmp1$UID
  ha = HeatmapAnnotation(tmp1[,'Diffname_short', drop = F])

  writeLines(paste('Heatmap of',dim(tmp)[1],assay))
  draw(Heatmap(tmp, top_annotation = ha, show_row_names = F, show_column_names = F))
  
  PC = prcomp(tmp, scale. = T, center = T)
  
  plotdata = data.frame(UID = rownames(PC$rotation), PC1 = PC$rotation[,1], PC2 = PC$rotation[,2]) %>%
    left_join(tmp1)
  
  p = ggplot(plotdata, aes(x = PC1, y = PC2, color = Diffname_short, shape = Diffname_short)) + geom_point()
  p = p + scale_shape_manual(values = 1:length(unique(plotdata$Diffname_short)))
  print(p)
}

writeLines('# of entities that are differentially expressed between one diff state and all the other diff state are')
kable(specific.signatures %>%
        group_by(from.state, feature.assay, feature.direction) %>%
        summarise(ncount = length(unique(feature))) %>%
        spread(from.state, ncount))

# Store results in synapse
specific.signatures = specific.signatures %>%
  dplyr::select(-ncount) %>%
  dplyr::arrange(from.state, feature.assay, feature.direction, to.state)
```


### Indentify entities that are differentially expressed between one diff state and all the other diff states (called signatures here after)
Here three groups SC, DE, MESO5 are considered
```{r extract.specific.ent2, fig.height=12, fig.width=15}
# Seperate gene sets based on transitions
specific.signatures2 = all.int %>% 
  dplyr::select(feature, feature.assay, feature.direction, from.state, to.state) %>% 
  unique %>% 
  dplyr::filter(!(from.state %in% c('EB', 'MESO15', 'MESO30', 'ECTO')),
                !(to.state %in% c('EB', 'MESO15', 'MESO30', 'ECTO'))) %>%
  group_by(feature, feature.assay, feature.direction, from.state) %>%
  summarise(ncount = length(unique(to.state)), to.state = paste(unique(to.state), collapse = ',')) %>%
  filter((ncount == 2 & feature.assay != 'methyl') | (ncount == 2 & feature.assay == 'methyl')) %>%
  data.frame

writeLines('# of entities that are differentially expressed between one diff state and all the other diff state are')
kable(specific.signatures2 %>%
        group_by(from.state, feature.assay, feature.direction) %>%
        summarise(ncount = length(unique(feature))) %>%
        spread(from.state, ncount))

for (assay in unique(specific.signatures2$feature.assay)){
  tmp = expr[[assay]]
  rownames(tmp) = tmp[,1]
  tmp[,1] = NULL
  tmp[is.na(tmp)] = 0
  
  tmp1 = covariates[[assay]]
  rownames(tmp1) = tmp1$UID
  tmp1 = tmp1[tmp1$Diffname_short %in% c('SC','MESO5','DE'),]
  
  tmp = tmp[rownames(tmp) %in% unique(specific.signatures2$feature),rownames(tmp1)]
  
  ha = HeatmapAnnotation(tmp1[,'Diffname_short', drop = F])

  writeLines(paste('Heatmap of',dim(tmp)[1],assay))
  draw(Heatmap(tmp, top_annotation = ha, show_row_names = F, show_column_names = F))
  
  PC = prcomp(tmp, scale. = T, center = T)
  
  plotdata = data.frame(UID = rownames(PC$rotation), PC1 = PC$rotation[,1], PC2 = PC$rotation[,2]) %>%
    left_join(tmp1)
  
  p = ggplot(plotdata, aes(x = PC1, y = PC2, color = Diffname_short, shape = Diffname_short)) + geom_point()
  p = p + scale_shape_manual(values = 1:length(unique(plotdata$Diffname_short)))
  print(p)
}

# Store results in synapse
specific.signatures2 = specific.signatures2 %>%
  dplyr::select(-ncount) %>%
  dplyr::arrange(from.state, feature.assay, feature.direction, to.state)
```

### Store results in synapse
```{r synstore, cache=FALSE}
# Store signatures in synapse
write.table(specific.signatures, file = 'GlobalSignatures.tsv', row.names = F, sep = '\t', quote=F)
obj = File('GlobalSignatures.tsv', name = "Signatures (Global)", parentId = CODE$properties$id)
obj = synStore(obj, activityName = ActivityName, executed = thisFile, used = as.character(ALL_USED_IDs))

# Store signatures in synapse
write.table(specific.signatures2, file = 'MESENDOSignatures.tsv', row.names = F, sep = '\t', quote=F)
obj = File('MESENDOSignatures.tsv', name = "Signatures (SC-MESO5-DE)", parentId = CODE$properties$id)
obj = synStore(obj, activityName = ActivityName, executed = thisFile, used = as.character(ALL_USED_IDs))
```

### Source Code
[Source R Markdown](`r thisFile`)