---
title: "Differential expression analysis for eXpress aligned mRNA data with mixed effects modeling at SC state with EB polarisation scores"
author: "Thanneer Perumal"
date: "`r date()`"
---
```{r knit2syn, eval = F, echo = T}
library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./DiffExpAnal_mRNA_mixedEffects_SC_WithEBPolScores.Rmd",
                                 entityName = 'Differential Expression Analysis mRNA Mixed Effects SC WithEBPolScores',
                                 parentId = 'syn5008933')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, cache=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library(synapseClient)
library(knitr)
library(githubr)

library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(glmnet)
library(ComplexHeatmap)
library(circlize)

synapseLogin()

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE, cache=FALSE}
# Input Parameters
SOURCE.FOLDER_ID = 'syn7291553'
ALL.FILES = synQuery(paste0('select Diffname_short,id,fileType from file where parentId == "',SOURCE.FOLDER_ID,'"'))

SYNAPSE_STORE = T
parentId = 'syn5008933'

# Specify factor and continuous covariates pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Race', 'Ethnicity' , 'Gender', 'Disease', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')

ContCovariates = c('PassageAtThaw', 'PassageAtHarvest', 'eb.pol.scores')
```
Obtain processed counts, design, weights and covariates matrix from Synapse
```{r getdata, cache=TRUE, include=FALSE}
# Get processed counts matrix
PROC.COUNTS.ID = ALL.FILES %>%
  dplyr::filter(file.Diffname_short == 'SC', file.fileType == "GenomicMatrix")
ALL.USED.IDs = PROC.COUNTS.ID$file.id
PROC.COUNTS = read.table(synGet(PROC.COUNTS.ID$file.id)@filePath,
                         sep = '\t', header = T, row.names = 1, quote='', check.names = F)

# Get sample covariates
COV.ID = ALL.FILES %>%
  dplyr::filter(file.Diffname_short == 'SC', file.fileType == "Covariates")
ALL.USED.IDs = c(ALL.USED.IDs, COV.ID$file.id)
COV = read.table(synGet(COV.ID$file.id)@filePath, sep = '\t', row.names=1, header=T)
  
# Convert factor covariates to factors
FactorCovariates = intersect(colnames(COV), FactorCovariates)
COV[,FactorCovariates] = lapply(COV[,FactorCovariates], factor)
ContCovariates = intersect(colnames(COV), ContCovariates)
COV[,ContCovariates] = lapply(COV[,ContCovariates], as.numeric)
  
# Get covariates to adjust from synapse
postAdjustedCovars = c('eb.pol.scores',
                       str_split(annotations(synGet(COV.ID$file.id))$postAdjustCovars, ',')[[1]]) %>%
  unique

# Remove few SC and EB samples form this analysis for validation purpose
METADATA_OBJ = synTableQuery(paste('SELECT * FROM','syn3156503',sep=' '))
ALL.USED.IDs <- c(ALL.USED.IDs, METADATA_OBJ@schema)
METADATA = METADATA_OBJ@values

CELL.LINE.MAPPING = data.frame(UID = rownames(COV),
                               eb.pol.scores = COV$eb.pol.scores) %>%
  left_join(METADATA %>%
              dplyr::select(UID, Diffname_short, C4_Cell_Line_ID))
```

### Differential expression analysis (wrt eb.pol.scores)
In each diffState, for each covariate, we fit a linear model using `limma` with only the variables obtained from the covariate analysis. Here primary variable of interest is eb.pol.scores

Differential expression with only 30 cell lines
```{r diff.exp, include = FALSE}
ALL.DEXP = list()
ALL.AUC = c()
for(ncount in 1:100){
  
  train.id = sample(unique(CELL.LINE.MAPPING$C4_Cell_Line_ID), 30)
  test.id = setdiff(unique(CELL.LINE.MAPPING$C4_Cell_Line_ID), train.id)
  
  train.UID = CELL.LINE.MAPPING$UID[CELL.LINE.MAPPING$C4_Cell_Line_ID %in% train.id]
  test.UID = CELL.LINE.MAPPING$UID[CELL.LINE.MAPPING$C4_Cell_Line_ID %in% test.id]

  cat("Get differentially expressed genes using limma package with following coefficients in the linear model:\n")
  cat("----\n",
      "Get differentially expressed genes using limma package with following coefficients in the linear model:\n",
      paste0('Fixed Effects:',paste(postAdjustedCovars, collapse = ','), "\n"),
      "Random Effects: Donor_ID\n")
  
  # Get desing matrix
  DESIGN.MAT <- getDesignMatrix(COV[train.UID, postAdjustedCovars, drop = F], Intercept = F)
  DESIGN.MAT = DESIGN.MAT$design[,linColumnFinder(DESIGN.MAT$design)$indepCols,drop=F]

  # Voom normalisations
  EXPR <- voom(PROC.COUNTS[,train.UID], design = DESIGN.MAT)
  
  # Calculate correlated effects of donors
  CORRELATION <- duplicateCorrelation(EXPR, block = COV[train.UID, 'Donor_ID'])

  # Re-estimate voom normalisations
  EXPR <- voom(PROC.COUNTS[,train.UID], design = DESIGN.MAT, correlation = CORRELATION$cor)
  
  # Fit linear model using mixed effects design
  FIT = lmFit(EXPR)

  # Make contrast  
  CONT <- makeContrasts(contrasts = 'eb.pol.scores', levels=colnames(DESIGN.MAT))

  # Refit contrasts
  FIT.CONT <- contrasts.fit(FIT, CONT)

  # Estimate moderated t-statistics
  FIT.CONT <- eBayes(FIT.CONT)
  
  # Get differentially expressed genes
  DEXP <- topTable(FIT.CONT, coef = 1, number = dim(PROC.COUNTS)[1]) %>%
    rownameToFirstColumn('hgnc_symbol')

  # DEXP %>%
  #   dplyr::mutate(Comparison = 'eb.pol.scores') %>%
  #   filter(adj.P.Val <= 0.05) %>%
  #   summarise(count = n(), Genes = paste(unique(hgnc_symbol), collapse = ',')) %>%
  #   kable

  # Voom normalisations
  ALL.EXPR <- voom(PROC.COUNTS, design = NULL)
  
  # Logistic regression prediction
  if (sum(DEXP$adj.P.Val <= 0.05) >= 100){
    # Predict eb.pol.scores using genomic features
    x.data = EXPR$E[DEXP$hgnc_symbol[DEXP$adj.P.Val <= 0.05],rownames(DESIGN.MAT)] %>% t %>% scale
    y.data = DESIGN.MAT[,'eb.pol.scores']

    fit.cv = cv.glmnet(x.data, y.data, type.measure = 'class')
  
    x.test = ALL.EXPR$E[DEXP$hgnc_symbol[DEXP$adj.P.Val <= 0.05],test.UID] %>% t %>% scale
    y.test = COV[test.UID,'eb.pol.scores']
    y.test[y.test<0.5] = 0;y.test[y.test>=0.5] = 1
    
    fit.test = predict.cv.glmnet(fit.cv, newx = x.test, s = "lambda.min", type = 'response')
    
    ALL.AUC = c(ALL.AUC, glmnet::auc(y.test, fit.test))
    ALL.DEXP = c(ALL.DEXP, list(DEXP))
  }
  print(ncount)
}
```

```{r bp}
boxplot(ALL.AUC, main = 'AUC')
```

Differential expression with all cell lines
```{r diff.exp2, fig.height=20, fig.width=15}
cat("Get differentially expressed genes using limma package with following coefficients in the linear model:\n")
cat("----\n",
    "Get differentially expressed genes using limma package with following coefficients in the linear model:\n",
    paste0('Fixed Effects:',paste(postAdjustedCovars, collapse = ','), "\n"),
    "Random Effects: Donor_ID\n")
  
# Get desing matrix
DESIGN.MAT <- getDesignMatrix(COV[, postAdjustedCovars, drop = F], Intercept = F)
DESIGN.MAT = DESIGN.MAT$design[,linColumnFinder(DESIGN.MAT$design)$indepCols,drop=F]

# Voom normalisations
EXPR <- voom(PROC.COUNTS, design = DESIGN.MAT)
  
# Calculate correlated effects of donors
CORRELATION <- duplicateCorrelation(EXPR, block = COV[, 'Donor_ID'])

# Re-estimate voom normalisations
EXPR <- voom(PROC.COUNTS, design = DESIGN.MAT, correlation = CORRELATION$cor)
  
# Fit linear model using mixed effects design
FIT = lmFit(EXPR)

# Make contrast  
CONT <- makeContrasts(contrasts = 'eb.pol.scores', levels=colnames(DESIGN.MAT))

# Refit contrasts
FIT.CONT <- contrasts.fit(FIT, CONT)

# Estimate moderated t-statistics
FIT.CONT <- eBayes(FIT.CONT)
  
# Get differentially expressed genes
DEXP <- topTable(FIT.CONT, coef = 1, number = dim(PROC.COUNTS)[1]) %>%
  rownameToFirstColumn('hgnc_symbol')

DEXP %>%
  dplyr::mutate(Comparison = 'eb.pol.scores') %>%
  filter(adj.P.Val <= 0.05) %>%
  summarise(count = n(), Genes = paste(unique(hgnc_symbol), collapse = ',')) %>%
  kable

# Predict eb.pol.scores using genomic features
x.data = EXPR$E[DEXP$hgnc_symbol[DEXP$adj.P.Val <= 0.05],rownames(DESIGN.MAT)] %>% t %>% scale
y.data = DESIGN.MAT[,'eb.pol.scores']

fit.cv = cv.glmnet(x.data, y.data, type.measure = 'class')
fit.test = predict.cv.glmnet(fit.cv, newx = x.data, s = "lambda.min")
p = ggplot(data.frame(y.train = y.data, pred.prob = fit.test) %>%
             dplyr::rename(pred.prob = X1), aes(x = y.train, y = pred.prob)) + geom_point()
p = p + geom_smooth(method = 'lm')
p

DESIGN.MAT = DESIGN.MAT[order(DESIGN.MAT[,c('eb.pol.scores')]),]
ha = HeatmapAnnotation(DESIGN.MAT[,c('eb.pol.scores','Genderfemale'),drop=F])
Heatmap(t(x.data[rownames(DESIGN.MAT), ]),
        col = colorRamp2(c(-2,0,2), c('blue','white','red')),
        top_annotation = ha,
        cluster_columns = F,
        name = 'Scaled logCPM')
```

With predictors of penalised logistic regression model
```{r heatmap, fig.height=10}
coef = as.matrix(coef.cv.glmnet(fit.cv))
Heatmap(t(x.data[rownames(DESIGN.MAT), colnames(x.data) %in% rownames(coef)[coef != 0]]),
        col = colorRamp2(c(-2,0,2), c('blue','white','red')),
        top_annotation = ha,
        cluster_columns = F,
        name = 'Scaled logCPM')
```

Investigate gender relationship with eb.pol.scores
```{r explore}
covariates = COVARIATES$SC %>%
  rownameToFirstColumn('UID')

assoc.stat = getAssociationStatistics(covariates[,c('Gender','eb.pol.scores')])

p = ggplot(covariates, aes(x = Gender, y = eb.pol.scores)) + geom_violin()
p = p + ggtitle(paste('Estimate:', round(assoc.stat$ESTIMATE[1,2], 3), ', Pval:', round(assoc.stat$PVAL[1,2],2)))
p
```

### Store files in Synapse
Store logFC, adjusted p-values and differentially expressed genes.
```{r synapse.store, include = FALSE, eval=TRUE, cache=FALSE}
activityName='Differential expression analysis of eXpress aligned mRNA with mixed effects model at each differentiation states'

thisFileName <- 'DiffExpAnal_mRNA_mixedEffects_SC_WithEBPolScores.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName="mRNA")

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Create folder to store the results and populate analysis wiki
CODE <- Folder(name = 'Differential Expression Analysis mRNA Mixed Effects SC WithEBPolScores', parentId = parentId)
CODE <- synStore(CODE)

save(list=ls(), file = 'Results.RData')
obj = File('Results.RData', name = 'Results', parentId = CODE$properties$id)
obj = synStore(obj, used = ALL.USED.IDs, executed = thisFile, activityName = activityName)

write.table(DEXP, file = 'DifferentialExpression.tsv', sep = '\t', row.names = F, quote=T)
obj = File('DifferentialExpression.tsv', name = 'Differential Expression', parentId = CODE$properties$id)
obj = synStore(obj, used = ALL.USED.IDs, executed = thisFile, activityName = activityName)
```