---
title: "Differential expression analysis for eXpress aligned data with mixed effects modeling"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('ctv')
library('psych')
library('reshape2')
library('vcd')
library('erer')
library('fpc')
library('data.table')
library('dplyr')
library('knitr')
library('stringr')

## Needs the dev branch
library(rGithubClient)

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

synapseLogin()

# source('/mnt/Github/knit2synapse-1//R/knitFile2Synapse.R')
# knitFile2Synapse(file = "./DiffExpAnal_Express_mixedEffects.Rmd", owner = 'syn3972059', name = "Differential Expression Analysis Express Mixed Effects",overwrite=F)

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```
### Download data
```{r setup, include=FALSE}
# Input Parameters
COUNT_ID = 'syn3446250'
METADATA_ID = 'syn3156503'

SYNAPSE_STORE = T
parentId = 'syn3471223'

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c("Diffname_short", "Gender", "Reprogramming_Vector_Type_Level2", "Cell_Type_of_Origin_Level2", "Donor_Life_Stage", "Disease", "run","Donor_ID")
ContCovariates = NULL
```
Synapse id of count matrix used for the analysis is `r COUNT_ID` and the synapse id of meta data table used for the analysis is `r METADATA_ID`. 

Factor covariates considered for analysis are  `r paste(gsub('_','\\\\_',FactorCovariates),collapse=',')`, and continuous covariates considered for the analysis are  `r paste(gsub('_','\\\\_',ContCovariates),collapse=',')`.

Obtain count matrix and metadata from synapse.
```{r getdata, cache=TRUE, include=FALSE}
# Get count matrix
COUNT_OBJ = synGet(COUNT_ID)
ALL_USED_IDs = COUNT_OBJ$properties$id
COUNT = fread(getFileLocation(COUNT_OBJ), data.table=FALSE)
row.names(COUNT) = COUNT[,1]

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_OBJ@schema
METADATA = METADATA_OBJ@values
```
Preprocess counts matrix and metadata.
```{r preprocessing, include=FALSE}
# Preprocess metadata
METADATA[METADATA == 'N/A'] = NA

# Replace all special characters with blank
myFix <- function(x) str_replace_all(x, '[^[:alnum:]]', '')
METADATA <- METADATA %>%
  dplyr::mutate_each(funs(myFix), -UID, -C4_Cell_Line_ID, -biologicalSampleName) # fix them but don't touch some columns

# Set rownames
rownames(METADATA) = METADATA$UID
```
### Preprocess data
* Remove somatic samples and samples with not type.
* Remove samples that failed QC and samples classified as exclude.
* Remove samples with abnormal karyotypes.
```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  METADATA %>%
  filter(Diffname_short != "") %>%
  filter(UID %in% colnames(COUNT)) %>%
  filter(Cell_Type == "PSC") %>%  
  filter(pass_qc == "TRUE") %>%
  filter(exclude != "TRUE") %>%
  filter(C4_Karyotype_Result != "abnormal")

REMOVED_UID <- setdiff(METADATA$UID, metadata_filtered$UID)
METADATA <- METADATA[metadata_filtered$UID,]
COUNT <- COUNT[, METADATA$UID]
```
The following `r length(REMOVED_UID)` samples were removed:

`r paste(gsub('_','\\\\_',REMOVED_UID), collapse= ',')` 

### CPM Normalisation
Determine design matrix for normalisation and differential expression analysis. 

Remove genes that have less than 1 cpm counts in at least 50% of each of the individual differentiation stages.
```{r cpmnormalisation}
tmp <- tapply(colnames(COUNT),
              factor(METADATA$Diffname_short),
              function(cols,COUNT){PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[,cols])},
              COUNT)

ugenes <- c()
for (i in unique(METADATA$Diffname_short)) {
  ugenes <- unique(c(ugenes,tmp[[i]]$filteredExprMatrix$genes[,1]))
}

COUNT <- COUNT[ugenes,,drop=F]
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT,MIN_GENE_CPM=0,
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
```{r covariates.clustering}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
```
### Differential expression 
Obtain differentially expressed genes with the following covariates (as obtained from covariates analysis) in the model: 
* Fixed Effects: Diffname short, Gender, Cell Type of Origin Level2, Reprogramming Vector Type Level2, Donor Life Stage, lane, run
* Random Effects: Donor ID
#### Differentiation stages
```{r diffstate, cache=TRUE, include=FALSE}
# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,c("Diffname_short", "Gender", "Cell_Type_of_Origin_Level2", "Reprogramming_Gene_Combination", "Reprogramming_Vector_Type", "Disease", "run")], Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
    
# Estimate correlation of random effects
EXPR = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DESIGN, plot=F)

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression between different differentiation stages
CONT.NAMES <- colnames(DESIGN)[grep('Diffname_short',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Diffname_short','',x);
                                                    x <- gsub('-','_',x)})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
DIFFNAME <- list()
DIFFNAME$logFC <- data.frame(row.names = rownames(EXPR$E))
DIFFNAME$adj.P.Val <- data.frame(row.names = rownames(EXPR$E))
DIFFNAME$SIG.SETS <- data.frame()

for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR$E)[1])    
  DIFFNAME$logFC[,i] <- tmp[rownames(DIFFNAME$logFC),'logFC']
  DIFFNAME$adj.P.Val[,i] <- tmp[rownames(DIFFNAME$adj.P.Val),'adj.P.Val'] 
  
  DIFFNAME$SIG.SETS <- rbind(DIFFNAME$SIG.SETS,
                             getUpDownGenes(DIFFNAME$adj.P.Val[,i], DIFFNAME$logFC[,i], rownames(DIFFNAME$logFC), i))
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
DIFFNAME$SIG.EXP.POS <- DIFFNAME$adj.P.Val<=0.05 & DIFFNAME$logFC >= 1   
DIFFNAME$NUM.SIG.EXP.POS <- colSums(DIFFNAME$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
DIFFNAME$SIG.EXP.NEG <- DIFFNAME$adj.P.Val<=0.05 & DIFFNAME$logFC <= -1   
DIFFNAME$NUM.SIG.EXP.NEG <- colSums(DIFFNAME$SIG.EXP.NEG)

save(DIFFNAME,file='DifferentialExpressionDiffName_mixedEffects.RData')
```
Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Number of differentially expressed genes at FDR <= 0.05 and logFC >= 1 or logFC <= -1
```{r significant.sets1,cache=TRUE}
tmp <- cbind(as.data.frame(DIFFNAME$NUM.SIG.EXP.POS),as.data.frame(DIFFNAME$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```
#### Gender 
```{r gender, cache=TRUE}
# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,c("Gender", "Diffname_short", "Cell_Type_of_Origin_Level2", "Reprogramming_Gene_Combination", "Reprogramming_Vector_Type", "Disease", "run")], Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
    
# Estimate correlation of random effects
EXPR = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DESIGN, plot=F)

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression
CONT.NAMES <- colnames(DESIGN)[grep('Gender',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Gender','',x);
                                                    x <- gsub('-','_',x)})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)
  
# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
GENDER <- list()
GENDER$logFC <- data.frame(row.names = rownames(EXPR$E))
GENDER$adj.P.Val <- data.frame(row.names = rownames(EXPR$E))
GENDER$SIG.SETS <- data.frame()

for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR$E)[1])    
  GENDER$logFC[,i] <- tmp[rownames(GENDER$logFC),'logFC']
  GENDER$adj.P.Val[,i] <- tmp[rownames(GENDER$adj.P.Val),'adj.P.Val'] 
  
  GENDER$SIG.SETS <- rbind(GENDER$SIG.SETS,
                             getUpDownGenes(GENDER$adj.P.Val[,i], GENDER$logFC[,i], rownames(GENDER$logFC), i))
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
GENDER$SIG.EXP.POS <- GENDER$adj.P.Val<=0.05 & GENDER$logFC >= 1   
GENDER$NUM.SIG.EXP.POS <- colSums(GENDER$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
GENDER$SIG.EXP.NEG <- GENDER$adj.P.Val<=0.05 & GENDER$logFC <= -1   
GENDER$NUM.SIG.EXP.NEG <- colSums(GENDER$SIG.EXP.NEG)

save(GENDER,file='DifferentialExpressionGender_mixedEffects.RData')
```
Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Number of differentially expressed genes at FDR <= 0.05 and absolute logFC >= 1 or logFC <= -1
```{r significant.sets2,cache=TRUE}
tmp <- cbind(as.data.frame(GENDER$NUM.SIG.EXP.POS),as.data.frame(GENDER$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```
#### Cell Type of Origin Level2
```{r cell.type.of.origin, cache=TRUE}
# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,c("Cell_Type_of_Origin_Level2", "Gender", "Diffname_short",  "Reprogramming_Gene_Combination", "Reprogramming_Vector_Type", "Disease", "run")], Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
    
# Estimate correlation of random effects
EXPR = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DESIGN, plot=F)

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression
CONT.NAMES <- colnames(DESIGN)[grep('Cell_Type_of_Origin_Level2',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Cell_Type_of_Origin_Level2','',x);
                                                    x <- gsub('-','_',x)})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
CELL.TYPE.OF.ORIGIN <- list()
CELL.TYPE.OF.ORIGIN$logFC <- data.frame(row.names = rownames(EXPR$E))
CELL.TYPE.OF.ORIGIN$adj.P.Val <- data.frame(row.names = rownames(EXPR$E))
CELL.TYPE.OF.ORIGIN$SIG.SETS <- data.frame()

for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR$E)[1])    
  CELL.TYPE.OF.ORIGIN$logFC[,i] <- tmp[rownames(CELL.TYPE.OF.ORIGIN$logFC),'logFC']
  CELL.TYPE.OF.ORIGIN$adj.P.Val[,i] <- tmp[rownames(CELL.TYPE.OF.ORIGIN$adj.P.Val),'adj.P.Val'] 
  
  CELL.TYPE.OF.ORIGIN$SIG.SETS <- rbind(CELL.TYPE.OF.ORIGIN$SIG.SETS,
                             getUpDownGenes(CELL.TYPE.OF.ORIGIN$adj.P.Val[,i], CELL.TYPE.OF.ORIGIN$logFC[,i], rownames(CELL.TYPE.OF.ORIGIN$logFC), i))
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
CELL.TYPE.OF.ORIGIN$SIG.EXP.POS <- CELL.TYPE.OF.ORIGIN$adj.P.Val<=0.05 & CELL.TYPE.OF.ORIGIN$logFC >= 1   
CELL.TYPE.OF.ORIGIN$NUM.SIG.EXP.POS <- colSums(CELL.TYPE.OF.ORIGIN$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
CELL.TYPE.OF.ORIGIN$SIG.EXP.NEG <- CELL.TYPE.OF.ORIGIN$adj.P.Val<=0.05 & CELL.TYPE.OF.ORIGIN$logFC <= -1   
CELL.TYPE.OF.ORIGIN$NUM.SIG.EXP.NEG <- colSums(CELL.TYPE.OF.ORIGIN$SIG.EXP.NEG)

save(CELL.TYPE.OF.ORIGIN,file='DifferentialExpressionCellTypeOfOrigin.RData')
```
Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Number of differentially expressed genes at FDR <= 0.05 and logFC >= 1 or logFC <= -1
```{r significant.sets3,cache=TRUE}
tmp <- cbind(as.data.frame(CELL.TYPE.OF.ORIGIN$NUM.SIG.EXP.POS),as.data.frame(CELL.TYPE.OF.ORIGIN$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```
#### Reprogramming Gene Combination
```{r reprogramming.gene.combo, cache=TRUE}
# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,c("Reprogramming_Gene_Combination", "Diffname_short",  "Gender", "Cell_Type_of_Origin_Level2", "Reprogramming_Vector_Type", "Disease", "run")], Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
    
# Estimate correlation of random effects
EXPR = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DESIGN, plot=F)

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression
CONT.NAMES <- colnames(DESIGN)[grep('Reprogramming_Gene_Combination',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Reprogramming_Gene_Combination','',x);
                                                    x <- gsub('-','_',x)})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
REP.GENE.COMBO <- list()
REP.GENE.COMBO$logFC <- data.frame(row.names = rownames(EXPR$E))
REP.GENE.COMBO$adj.P.Val <- data.frame(row.names = rownames(EXPR$E))
REP.GENE.COMBO$SIG.SETS <- data.frame()

for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR$E)[1])    
  REP.GENE.COMBO$logFC[,i] <- tmp[rownames(REP.GENE.COMBO$logFC),'logFC']
  REP.GENE.COMBO$adj.P.Val[,i] <- tmp[rownames(REP.GENE.COMBO$adj.P.Val),'adj.P.Val'] 
  
  REP.GENE.COMBO$SIG.SETS <- rbind(REP.GENE.COMBO$SIG.SETS,
                             getUpDownGenes(REP.GENE.COMBO$adj.P.Val[,i], REP.GENE.COMBO$logFC[,i], rownames(REP.GENE.COMBO$logFC), i))
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
REP.GENE.COMBO$SIG.EXP.POS <- REP.GENE.COMBO$adj.P.Val<=0.05 & REP.GENE.COMBO$logFC >= 1   
REP.GENE.COMBO$NUM.SIG.EXP.POS <- colSums(REP.GENE.COMBO$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
REP.GENE.COMBO$SIG.EXP.NEG <- REP.GENE.COMBO$adj.P.Val<=0.05 & REP.GENE.COMBO$logFC <= -1   
REP.GENE.COMBO$NUM.SIG.EXP.NEG <- colSums(REP.GENE.COMBO$SIG.EXP.NEG)

save(REP.GENE.COMBO,file='DifferentialExpressionRepGeneCombo.RData')
```
Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Number of differentially expressed genes at FDR <= 0.05 and logFC >= 1 or logFC <= -1
```{r significant.sets4,cache=TRUE}
tmp <- cbind(as.data.frame(REP.GENE.COMBO$NUM.SIG.EXP.POS),as.data.frame(REP.GENE.COMBO$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```
#### Reprogramming Vector Type
```{r reprogramming.vector, cache=TRUE}
# Post adjusted design matrix
DESIGN = getDesignMatrix(COVARIATES[,c("Reprogramming_Vector_Type", "Diffname_short",  "Gender", "Cell_Type_of_Origin_Level2", "Reprogramming_Gene_Combination", "Disease", "run")], Intercept = F)
DESIGN = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]
    
# Estimate correlation of random effects
EXPR = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DESIGN, plot=F)

# Calculate correlation between donors
correlation <- duplicateCorrelation(EXPR, block=COVARIATES$Donor_ID)

# Fit linear model using mixed effects design
FIT = lmFit(EXPR, block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Make contrast to check differential expression
CONT.NAMES <- colnames(DESIGN)[grep('Reprogramming_Vector_Type',colnames(DESIGN))]
CONT.NAMES <- combn(CONT.NAMES,2)
CONT.NAMES <- apply(CONT.NAMES,2,paste,collapse='-')

CONT <- makeContrasts(contrasts=CONT.NAMES,levels=colnames(FIT$coefficients))
colnames(CONT) <- sapply(colnames(CONT),function(x){x <- gsub('Reprogramming_Vector_Type','',x);
                                                    x <- gsub('-','_',x)})

# Refit contrasts
FIT.CONTRAST <- contrasts.fit(FIT,CONT)

# Estimate moderated t-statistics
FIT.CONTRAST <- eBayes(FIT.CONTRAST)
  
# Obtain all the differential expession combinations
REP.VECT.TYPE <- list()
REP.VECT.TYPE$logFC <- data.frame(row.names = rownames(EXPR$E))
REP.VECT.TYPE$adj.P.Val <- data.frame(row.names = rownames(EXPR$E))
REP.VECT.TYPE$SIG.SETS <- data.frame()

for (i in colnames(CONT)){
  tmp <- topTable(FIT.CONTRAST, coef=i, number=dim(EXPR$E)[1])    
  REP.VECT.TYPE$logFC[,i] <- tmp[rownames(REP.VECT.TYPE$logFC),'logFC']
  REP.VECT.TYPE$adj.P.Val[,i] <- tmp[rownames(REP.VECT.TYPE$adj.P.Val),'adj.P.Val'] 
  
  REP.VECT.TYPE$SIG.SETS <- rbind(REP.VECT.TYPE$SIG.SETS,
                             getUpDownGenes(REP.VECT.TYPE$adj.P.Val[,i], REP.VECT.TYPE$logFC[,i], rownames(REP.VECT.TYPE$logFC), i))
}

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC >= 1
REP.VECT.TYPE$SIG.EXP.POS <- REP.VECT.TYPE$adj.P.Val<=0.05 & REP.VECT.TYPE$logFC >= 1   
REP.VECT.TYPE$NUM.SIG.EXP.POS <- colSums(REP.VECT.TYPE$SIG.EXP.POS)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.05 and logFC <= -1
REP.VECT.TYPE$SIG.EXP.NEG <- REP.VECT.TYPE$adj.P.Val<=0.05 & REP.VECT.TYPE$logFC <= -1   
REP.VECT.TYPE$NUM.SIG.EXP.NEG <- colSums(REP.VECT.TYPE$SIG.EXP.NEG)

save(REP.VECT.TYPE,file='DifferentialExpressionRepVectType.RData')
```
Get differentially expressed genes using limma package with following co-efficients `r colnames(DESIGN)` in linear model

Number of differentially expressed genes at FDR <= 0.05 and logFC >= 1 or logFC <= -1
```{r significant.sets5,cache=TRUE}
tmp <- cbind(as.data.frame(REP.VECT.TYPE$NUM.SIG.EXP.POS),as.data.frame(REP.VECT.TYPE$NUM.SIG.EXP.NEG))
colnames(tmp) <- c('NUMBER OF SIG. EXP. GENES (+ve)','NUMBER OF SIG. EXP. GENES (-ve)')
kable(tmp)
```
#### Get pvalue and fold changes of some important genes
```{r print.imp.genes}
FC <- join_all(list(rownameToFirstColumn(DIFFNAME$logFC,'GeneNames'),
                    rownameToFirstColumn(GENDER$logFC,'GeneNames'),
                    rownameToFirstColumn(CELL.TYPE.OF.ORIGIN$logFC,'GeneNames'),
                    rownameToFirstColumn(REP.GENE.COMBO$logFC,'GeneNames'),
                    rownameToFirstColumn(REP.VECT.TYPE$logFC,'GeneNames')),
               by = 'GeneNames',
               match = 'all')

PVAL <- join_all(list(rownameToFirstColumn(DIFFNAME$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(GENDER$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(CELL.TYPE.OF.ORIGIN$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(REP.GENE.COMBO$adj.P.Val,'GeneNames'),
                      rownameToFirstColumn(REP.VECT.TYPE$adj.P.Val,'GeneNames')),
                 by = 'GeneNames',
                 match = 'all')

SIG <- join_all(list(rownameToFirstColumn(DIFFNAME$SIG.EXP.POS+DIFFNAME$SIG.EXP.NEG,'GeneNames'),
                     rownameToFirstColumn(GENDER$SIG.EXP.POS+GENDER$SIG.EXP.NEG,'GeneNames'),
                     rownameToFirstColumn(CELL.TYPE.OF.ORIGIN$SIG.EXP.POS+CELL.TYPE.OF.ORIGIN$SIG.EXP.NEG,'GeneNames'),
                     rownameToFirstColumn(REP.GENE.COMBO$SIG.EXP.POS+REP.GENE.COMBO$SIG.EXP.NEG,'GeneNames'),
                     rownameToFirstColumn(REP.VECT.TYPE$SIG.EXP.POS+REP.VECT.TYPE$SIG.EXP.NEG,'GeneNames')),
                by = 'GeneNames',
                match = 'all')

SIG.SETS <- rbind(DIFFNAME$SIG.SETS,
                  GENDER$SIG.SETS,
                  CELL.TYPE.OF.ORIGIN$SIG.SETS,
                  REP.GENE.COMBO$SIG.SETS,
                  REP.VECT.TYPE$SIG.SETS)

print('Fold change:')
kable(filter(FC, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
print('Adjusted Pvalue:')
kable(filter(PVAL, GeneNames %in% c('LARGE','FRG1','FRG1B','PEG10','SPEG')))
```
#### Gender effect in differentially expressed genes
ECDFs of Ccoefficient of variation between male and female cells at each differentiation stages
```{r gender.effect, cache=TRUE,fig.height=4,fig.width=4}
ind <- which(rowSums(SIG[,2:12]) != 0)

GENDER.VAR <- list()
for (Diffname in  c("SC","EB","ECTO","MESO5","DE")){
  for (Gender in levels(COVARIATES$Gender)){
    TEMP.EXPR <- EXPR$E[ind, COVARIATES$Diffname_short == Diffname & COVARIATES$Gender == Gender]
    GENDER.VAR[[Diffname]][[Gender]] <- apply(TEMP.EXPR,1,mean)/apply(TEMP.EXPR,1,sd)
  }
  plotdata <- as.data.frame(rbind(cbind(GENDER.VAR[[Diffname]]$male,'male'),cbind(GENDER.VAR[[Diffname]]$female,'female')))
  plotdata[,1] <- as.numeric(as.character(plotdata[,1]))
  colnames(plotdata) <- c('CV','Gender')
  GENDER.VAR[[Diffname]]$plot <- ggplot(plotdata, aes(x = CV)) + stat_ecdf(aes(colour = Gender)) + ggtitle(Diffname)  
  print(GENDER.VAR[[Diffname]]$plot)
}
```
No significant differnece between male vs female stem cells at each of the differentiation stages in terms of their variance

Perform enrichment analysis for differentiation
```{r enrichment.analysis, include=FALSE}
# Write input files to folder
inputFolderName <- paste0(getwd(),'/GOElite_input')
if (file.exists(inputFolderName))
  unlink(inputFolderName, recursive = T, force = T)
dir.create(inputFolderName, showWarnings = TRUE, recursive = FALSE)

# Function to write input files for GO-Elite
writeInputFiles <- function(cont.name, FC, PVAL, inputFolderName){
  input <- FC[(FC[,cont.name]>=1 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up',sep='_'),
                 paste(input$GeneNames,collapse=','))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'up','txt',sep='.')    
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  input <- FC[(FC[,cont.name]<=-1 & PVAL[,cont.name]<=0.05),c('GeneNames',cont.name),drop=F]
  input <- dplyr::mutate(input,SystemCode = 'Sy')
  input <- input[,c(1,3,2)]
  colnames(input)[3] <- 'logFC'
  
  Genes <- rbind(Genes,cbind(paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down',sep='_'),
                             paste(input$GeneNames,collapse=',')))
  
  FName <- paste(str_replace_all(cont.name,'[^[:alnum:]]','_'),'down','txt',sep='.')  
  FName <- write.table(input,paste(inputFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)
  
  return(Genes)
}
DEXPP.GENES <- lapply(colnames(FC)[-(1)],writeInputFiles, FC, PVAL, inputFolderName)
DEXPP.GENES <- plyr::ldply(DEXPP.GENES) 
colnames(DEXPP.GENES) <- c('Contrast.Names','Gene.Symbols')

# Write denominator file to folder
denominatorFolderName <- paste0(getwd(),'/GOElite_denominator')
if (file.exists(denominatorFolderName))
  unlink(denominatorFolderName, recursive = T, force = T)
dir.create(denominatorFolderName, showWarnings = TRUE, recursive = FALSE)
tmp <- FC[,c('GeneNames'),drop=F] 
tmp <- dplyr::mutate(tmp,SystemCode = 'Sy')
FName <- 'Denominaor.txt'
write.table(tmp,paste(denominatorFolderName,FName,sep='/'),sep='\t',row.names = F,col.names = T, quote=F)

# Create output dir
outputFolderName <- paste0(getwd(),'/GOElite_output')
if (file.exists(outputFolderName))
  unlink(outputFolderName, recursive = T, force = T)
dir.create(outputFolderName, showWarnings = TRUE, recursive = FALSE)

# system(paste('python',
#              '/mnt/data/GO-Elite_v.1.2.5-Py/GO_Elite.py',
#              '--species','Hs',
#              '--version','EnsMart72Plus',
#              '--input', inputFolderName,
#              '--denom',denominatorFolderName,
#              '--output',outputFolderName))
```
### Store files in synapse
```{r synapse.store, include = FALSE, eval=TRUE}
if(SYNAPSE_STORE){
  activityName='Differential Expression Analysis of eXpress aligned data with mixed effects model'
  
  thisFileName <- 'DiffExpAnal_Express_mixedEffects.Rmd'
  
  # Github link
  thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                      ref="branch", 
                      refName='diff_exp')

  thisFile <- getPermlink(repository = thisRepo,
                          repositoryPath=paste0('code/Rmd/', thisFileName))
    
  CODE <- File('./DiffExpAnal_Express_mixedEffects.Rmd',name = 'Differential Expression Analysis Express Mixed Effects',parentId = parentId)
  CODE <- synStore(CODE, used = ALL_USED_IDs,activityName = activityName, executed=thisFile)
      
  write.table(FC,file='./DiffnameShort_FoldChange_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  FC <- File('./DiffnameShort_FoldChange_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects logFC',parentId = parentId)
  FC <- synStore(FC, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
    
  write.table(PVAL,file='./DiffnameShort_Pvalue_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  PVAL <- File('./DiffnameShort_Pvalue_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects pvalue',parentId = parentId)
  PVAL <- synStore(PVAL, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
  
  write.table(DEXPP.GENES,file='./DiffnameShort_DiffGenes_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  DEXPP <- File('./DiffnameShort_DiffGenes_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects genelist',parentId = parentId)
  annotations(DEXPP) <- list(Pvalue='0.05',Foldchage='+/-1')
  DEXPP <- synStore(DEXPP, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
  
  write.table(SIG.SETS,file='./AllSignificantGenes_Express_mixedEffects.tsv',sep='\t',row.names=F,quote=F)
  SIG.SETS_OBJ <- File('./AllSignificantGenes_Express_mixedEffects.tsv',name = 'Differential Expression Analysis Express Mixed Effects DiffExpGenes',parentId = parentId)
  annotations(SIG.SETS_OBJ) <- list(Pvalue='0.05',Foldchage='+/-1')
  SIG.SETS_OBJ <- synStore(SIG.SETS_OBJ, used = ALL_USED_IDs, activityName = activityName, executed=CODE$properties$id)
}
```
|  *Results*                                                      |  *SynapseID*                    |
|  -------                                                        |   ---------                     |
|  Log fold change                                                |  `r FC$properties$id`           |
|  Adjusted pvalues                                               |  `r PVAL$properties$id`         |
|  Differentially expressed genes                                 |  `r DEXPP$properties$id`        |
|  Differentially expressed genes (with fold change and pvalue)   |  `r SIG.SETS_OBJ$properties$id` |