---
title: Identifying molecular signatures of differentiation states from ainter omic assays (i.e., mRNA, miRNA, DNA methylation probes and splicing junctions)
author: "Thanneer Perumal"
date: '`r date()`'
output: pdf_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'GermlineSpecificSignatures.Rmd',
                                 parentId = "syn5194922",
                                 entityName = 'Germline Specific Signatures')
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
library(matrixStats)

library(fpc)
library(WGCNA)
library(flashClust)

library(synapseClient)
library(knitr)
library(githubr)

synapseLogin()

library(circlize)
colfunc <- colorRampPalette(c("white", "red"))

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  fig.align = 'left',
  results = 'asis',
  out.width = '8in')
```

```{r synapseStore.params, include=FALSE}
parentId = "syn5194922"
ALL_USED_IDs = c()

ActivityName <- 'Get diff state specific signatures from all three different assays'

# Get the latest commit of this code from github
thisFileName <- 'GermlineSpecificSignatures.Rmd'

thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='diff_exp')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))

# Populate wiki with results
CODE <- Folder(name = 'Germline Specific Signatures', parentId = parentId)
CODE <- synStore(CODE)
```
Threshold parameters
```{r set.thresholds, echo=TRUE}
pval.th <- 0.01

mrna.lfc.th <- log2(2)
mirna.lfc.th <- log2(2)
methyl.changeBeta.th <- 0.3
splicing.changePSI.th <- 0.2
```

### Data download from synapse
#### Get raw logCPM counts from synapse
```{r getRawCount, include=FALSE}
expr.id = c(mrna = 'syn5011095', mirna = 'syn5014454', methyl = 'syn4487642', splicing = 'syn5048714')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(expr.id))
expr = lapply(expr.id, downloadFile)
```

#### Get covariates from synapse
```{r covars, include=FALSE}
covariates.id = c(mrna = 'syn5011149', mirna = 'syn5014460', methyl = 'syn4487669', splicing = 'syn5048719')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(covariates.id))
covariates = lapply(covariates.id, downloadFile)
```

#### Get differential expression results from synapse
```{r diffexp, include=FALSE}
### Download differential expression results from synapse
diffexp.id = c(mrna = 'syn5013690', mirna = 'syn5014584', methyl = 'syn4527629', splicing = 'syn5049321')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(diffexp.id))

# Diff states in a sensible, quasi time-based ordering
orderedDiffState = sort(c("SC", "MESO5", "DE", "MESO15", "MESO30", "ECTO", "EB"))
Comparisons = apply(combn(sort(orderedDiffState), 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get mrna differential expression results from synapse
diffexp = list()
diffexp$mrna = downloadFile(diffexp.id['mrna']) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mrna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Get mirna differential expression results from synapse
diffexp$mirna = downloadFile(diffexp.id['mirna']) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mirna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Get methyl differential expression results from synapse
diffexp$methyl = downloadFile(diffexp.id['methyl']) %>%
  dplyr::select(-one_of('feature')) %>%
  dplyr::rename(feature = methProbeIDs) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(changeBETA) >= methyl.changeBeta.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC, changeBETA) 

# Get splicing differential expression results from synapse
diffexp$splicing = downloadFile(diffexp.id['splicing']) %>%
  dplyr::select(-one_of('feature')) %>%
  plyr::rename(c('Minor-Isoform' = 'feature')) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(changePSI) >= splicing.changePSI.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC, changePSI) 

diffexp = rbindlist(diffexp, use.names = T, fill = T, idcol = 'Assay') %>%
  tidyr::separate(Comparison, c('from.state','to.state'), sep = '_vs_')

diffexp = rbindlist(list(diffexp,
                         diffexp %>%
                           dplyr::mutate(
                             from.state = diffexp$to.state,
                             to.state = diffexp$from.state,
                             logFC = -logFC, changeBETA = -changeBETA, changePSI = -changePSI)),
                    use.names = T, fill =T) %>%
  dplyr::mutate(Direction = logFC/abs(logFC))
```

#### Get co-expression data from synapse
```{r coexpp, include=FALSE}
coexp.id = 'syn5662375'
ALL_USED_IDs = c(ALL_USED_IDs, coexp.id)
coexp = downloadFile(coexp.id) %>%
  filter(feature.assay != 'splicing', target.assay != 'splicing') %>%
  gather(comparison, coexpression, -feature, -target, -feature.assay, -target.assay, -dataSource, -Assay) %>% 
  dplyr::select(-Assay,-dataSource) %>%
  separate(comparison, c('from.state', 'to.state'), sep = '_vs_') %>%
  dplyr::mutate(interaction.assay = paste(feature.assay, target.assay, sep = '_')) %>%
  dplyr::filter(!(interaction.assay == 'mirna_mrna' & coexpression > 0))

writeLines('80 percentile values for each interaction assay')
kable(coexp %>%
  group_by(interaction.assay) %>%
  summarise(Qntl_80P = quantile(abs(coexpression), 0.8, na.rm = T)))

coexp = coexp %>%
  filter((interaction.assay == 'methyl_mirna' & abs(coexpression) >= 0.45) |
           (interaction.assay == 'methyl_mrna' & abs(coexpression) >= 0.38) |
           (interaction.assay == 'mirna_mrna' & abs(coexpression) >= 0.48) |
           (interaction.assay == 'mrna_mirna' & abs(coexpression) >= 0.47) |
           (interaction.assay == 'mrna_mrna' & abs(coexpression) >= 0.45))

coexp = rbindlist(list(coexp,
                       coexp %>%
                         dplyr::rename(from.state = to.state, to.state = from.state)),
                  use.names = T, fill = T)
```

### Differential expression analysis summary 
```{r combine}
writeLines('Number of entities that are differentially expressed in our data are')
kable(group_by(diffexp, from.state, Assay, Direction) %>%
        summarise(count = length(unique(feature))) %>%
        tidyr::spread(from.state, count) %>%
        dplyr::mutate(Direction = factor(Direction, labels = c('-1' = 'Down', '1' = 'Up'))))  

# Combine diffexp and coexpression (criteria is the entities in the network should be both differentially expressed and co-expressed in a diff state)
all.int = inner_join(coexp,
                     diffexp %>%
                       dplyr::rename(feature.assay = Assay, feature.fdr = adj.P.value, feature.lfc = logFC, 
                                     feature.cb = changeBETA, feature.psi = changePSI, feature.direction = Direction)) %>%
  inner_join(diffexp %>%
              dplyr::rename(target = feature, target.assay = Assay, target.fdr = adj.P.value, target.lfc = logFC, 
                            target.cb = changeBETA, target.psi = changePSI, target.direction = Direction)) 

writeLines('Unique miRNA and methylation probes that are differentially expressed and coexpressed in the same comparison in our data are')
kable(group_by(all.int, from.state, feature.assay, feature.direction) %>%
        summarise(count = length(unique(feature))) %>%
        tidyr::spread(from.state, count) %>%
        dplyr::rename(Direction = feature.direction, Assay = feature.assay) %>%
        dplyr::mutate(Direction = factor(Direction, labels = c('-1' = 'Down', '1' = 'Up'))) %>%
        dplyr::filter(Assay %in% c('mirna','methyl')))
```

### Differentiation stages
Clustering samples based on gene signatures of differentiation stages from GO 
```{r mesendo.test, out.height= '10in'}
# These signatures are 
DE.Sig = c('ONECUT1', 'OTX2', 'GATA6', 'NKX2-1', 'GATA4', 'BPTF', 'HHEX', 'SOX7', 'NKX2-5', 'PAX9', 'MMP14', 'HDAC1')
MESO.Sig = c('BTK', 'FOXC1', 'TEAD2', 'LEF1', 'T', 'SRF', 'FOXH1', 'KDM6A', 'IKZF1', 'SMAD1', 'TBX1', 'FOXC2', 'SNAI1',
             'ECSIT', 'HNF1A', 'HTT', 'HAND1', 'TCF15', 'WNT5A', 'ETV2', 'IRX3', 'EYA1', 'EYA2', 'KDM6B', 'ZFPM2', 'ZIC2',
             'CITED2', 'LHX2', 'TP63', 'HES7', 'OSR1', 'CHURC1', 'FOXF1', 'TBX6')
ECTO.Sig = c('ZBTB17', 'GRHL3', 'ZBTB7B', 'ERF')
SC.Sig = c('ZSCAN10', 'LDB1', 'ZFP36L2', 'PAX8', 'DPPA4', 'ZNF358', 'NFYA', 'KDM2B', 'PBX1', 'PRDM14', 'CNOT1', 'HES5',
           'CNOT2', 'CNOT3', 'EPAS1', 'PRO')

markers = rbind(data.frame(diff.state = 'DE', signature = DE.Sig),
                data.frame(diff.state = 'MESO', signature = MESO.Sig),
                data.frame(diff.state = 'ECTO', signature = ECTO.Sig),
                data.frame(diff.state = 'SC', signature = SC.Sig))

tmp = filter(expr$mrna, GeneName %in% markers$signature)
rownames(tmp) = tmp$GeneName
tmp$GeneName = NULL

markers = filter(markers, signature %in% rownames(tmp))
rownames(markers) = markers$signature

tmp1 = covariates$mrna
rownames(tmp1) = tmp1$UID
tmp1 = tmp1[tmp1$Diffname_short != 'EB',]

tmp = scale(tmp[,rownames(tmp1)])
tmp = t(scale(t(tmp)))

ha = HeatmapAnnotation(tmp1[,'Diffname_short', drop = F], col = list(
  Diffname_short = c('SC' = 'violet', 'MESO5' = 'blue', 'DE' = 'green', 'ECTO' = 'yellow',  
                     'MESO15' = 'brown', 'MESO30' = 'purple')
))
Heatmap(tmp, name = 'Scaled logCPM', top_annotation = ha, split = markers[rownames(tmp), 'diff.state'],
        col = colorRamp2(c(-2,0,2), c("blue","black","yellow")),
        show_column_names = F, show_row_names = F)
```

```{r mesendo.test1}
PC = prcomp(tmp, scale. = T, center = T)
  
plotdata = data.frame(UID = rownames(PC$rotation), PC1 = PC$rotation[,1], PC2 = PC$rotation[,2]) %>%
  left_join(tmp1)
  
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
p = ggplot(plotdata, aes(x = PC1, y = PC2, color = Diffname_short, shape = Diffname_short)) + geom_point()
p = p + scale_shape_manual(values = c(0,2,4,13,14,15)) + scale_color_manual(values = cbPalette[c(1,2,3,4,6,7)])
print(p)
```
From the above heatmap we get four big clusters of differentiation stages. Especially ECTO, SC, MESO15-MESO30, and DE-MESO5 clusters. Therefore, for extracting unique signatures of differentiation stages, we are using this four clusters, excluding EB (which is potentially hypothesised to be a mixture of SC and all germ layers)

### Indentify entities that are differentially expressed between the germline clusters (called here as signature of germlines)
Here three groups MESENDODERM, MESODERM and ECTODERM are considered
```{r extract.specific.ent}
# Seperate gene sets based on transitions
specific.signatures = diffexp %>%
  ddply(.(Assay), .fun = function(x){
    if (unique(x$Assay) == 'mirna' || unique(x$Assay) == 'methyl')
      x = filter(x, feature %in% all.int$feature)
    return(x)
  }) %>%
  dplyr::rename(feature.assay = Assay, feature.direction = Direction) %>% 
  dplyr::select(feature, feature.assay, feature.direction, from.state, to.state) %>% 
  unique %>% 
  dplyr::mutate(from.state = gsub('MESO5','MESENDO',from.state), from.state = gsub('DE','MESENDO',from.state),
                to.state = gsub('MESO5','MESENDO',to.state), to.state = gsub('DE','MESENDO',to.state),
                from.state = gsub('MESO15','MESO',from.state), from.state = gsub('MESO30','MESO',from.state),
                to.state = gsub('MESO15','MESO',to.state), to.state = gsub('MESO30','MESO',to.state)) %>%
  dplyr::filter(from.state != 'EB', to.state != 'EB',
                from.state != 'SC', to.state != 'SC',
                !(from.state == 'MESO' & to.state == 'MESO'), 
                !(from.state == 'MESENDO' & to.state == 'MESENDO')) %>%
  group_by(feature, feature.assay, feature.direction, from.state) %>%
  summarise(ncount = length(unique(to.state)), to.state = paste(unique(to.state), collapse = ',')) %>%
  filter((ncount == 2 & feature.assay != 'methyl') | (ncount == 1 & feature.assay == 'methyl')) %>%
  data.frame

writeLines('Number of entities that are differentially expressed between one diff cluster with respect to all the other are')
kable(specific.signatures %>%
        group_by(from.state, feature.assay, feature.direction) %>%
        summarise(ncount = length(unique(feature))) %>%
        spread(from.state, ncount))

# Extract results to store in synapse
specific.signatures.store = specific.signatures %>%
  dplyr::select(feature, feature.assay, feature.direction, from.state) %>%
  left_join(diffexp %>%
              dplyr::rename(feature.direction = Direction, feature.assay = Assay) %>%
              dplyr::mutate(from.state = gsub('MESO5','MESENDO',from.state), from.state = gsub('DE','MESENDO',from.state),
                            to.state = gsub('MESO5','MESENDO',to.state), to.state = gsub('DE','MESENDO',to.state),
                            from.state = gsub('MESO15','MESO',from.state), from.state = gsub('MESO30','MESO',from.state),
                            to.state = gsub('MESO15','MESO',to.state), to.state = gsub('MESO30','MESO',to.state)) %>%
              dplyr::filter(from.state != 'EB', to.state != 'EB', !(from.state == 'MESO' & to.state == 'MESO'), 
                            !(from.state == 'MESENDO' & to.state == 'MESENDO'))) %>%
  filter(!is.na(to.state))
```
Visualise mRNA expression profiles
```{r extract.specific.ent.mrna, out.height= '10in'}
mrna = expr[['mrna']]
rownames(mrna) = mrna[,1]
mrna[,1] = NULL
mrna[is.na(mrna)] = 0
  
mrna.cov = covariates[['mrna']]
rownames(mrna.cov) = mrna.cov$UID

tmp.mrna = mrna[rownames(mrna) %in% unique(specific.signatures$feature),
                mrna.cov$UID[mrna.cov$Diffname_short !='SC' & mrna.cov$Diffname_short !='EB']]
tmp.mrna = scale(tmp.mrna)
tmp.mrna = t(scale(t(tmp.mrna)))

mrna = mrna[rownames(mrna) %in% unique(specific.signatures$feature),rownames(mrna.cov)]
mrna = scale(mrna)
mrna = t(scale(t(mrna)))

mrna.cov = split(mrna.cov, mrna.cov$Diffname_short)
mrna = lapply(mrna.cov, function(x, mrna){
  mrna = mrna[,x$UID]
  clust = hclust(dist(t(mrna)))
  mrna = mrna[,clust$order]
}, mrna)
  
mrna.cov = do.call(rbind, mrna.cov[c('SC','DE','MESO5','ECTO','MESO15','MESO30','EB')])
colnames(mrna.cov) = gsub('Diffname_short','DifferentiationStage', colnames(mrna.cov))
mrna = do.call(cbind, mrna[c('SC','DE','MESO5','ECTO','MESO15','MESO30','EB')])

clust = pamk(tmp.mrna, krange = 2:10, seed = 123456, usepam = F)
clust = rownameToFirstColumn(clust$pamobject$clustering, 'hgnc_symbol') %>%
  dplyr::rename(cluster = DF) %>%
  dplyr::mutate(cluster = factor(cluster, labels = c('1' = 'MESENDO', '2' = 'ECTO', '3' = 'LATE MESO')))
  
# Top Annotation
ha = HeatmapAnnotation(mrna.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'MESO5' = 'blue', 'DE' = 'green', 
                                                             'ECTO' = 'yellow', 'MESO15' = 'brown', 'MESO30' = 'purple',
                                                             'EB' = 'cyan')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('LATE MESO' = 'orange', 'MESENDO' = 'cyan', 'ECTO'= 'grey', 'MESO' = 'pink')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(mrna)[1],'mrna'))
h1 = Heatmap(mrna, name = 'Scaled logCPM', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-2,0,2), c("blue","black","yellow")),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h1)

# PC plot mrna
PC = prcomp(t(tmp.mrna), scale. = T, center = T)

# Reproject all samples in to the PC space
pc = sapply(1:dim(mrna)[2], function(i, expr,pcs){
  tmp = lm(expr ~ 0 + ., cbind(data.frame(expr = scale(expr[rownames(pcs),i])), pcs))$coefficients
}, mrna, PC$rotation) %>% t
rownames(pc) = colnames(mrna)  
  
plotdata = data.frame(UID = rownames(pc), PC1 = pc[,1], PC2 = pc[,2], assay = 'mrna', row = 1, col = 1) %>%
  left_join(mrna.cov[,c('UID', 'DifferentiationStage'), drop=F])
```
Visualise miRNA expression profiles
```{r extract.specific.ent.mirna, out.height= '10in'}
mirna = expr[['mirna']]
rownames(mirna) = mirna[,1]
mirna[,1] = NULL
mirna[is.na(mirna)] = 0
  
mirna.cov = covariates[['mirna']]
rownames(mirna.cov) = mirna.cov$UID
# mirna.cov = mirna.cov[mirna.cov$Diffname_short != 'EB',]
  
tmp.mirna = mirna[rownames(mirna) %in% unique(specific.signatures$feature),
                  mirna.cov$UID[mirna.cov$Diffname_short !='SC' & mirna.cov$Diffname_short !='EB']]
tmp.mirna = scale(tmp.mirna)
tmp.mirna = t(scale(t(tmp.mirna)))

mirna = mirna[rownames(mirna) %in% unique(specific.signatures$feature),rownames(mirna.cov)]
mirna = scale(mirna)
mirna = t(scale(t(mirna)))

mirna.cov = split(mirna.cov, mirna.cov$Diffname_short)
mirna = lapply(mirna.cov, function(x, mirna){
  mirna = mirna[,x$UID]
  clust = hclust(dist(t(mirna)))
  mirna = mirna[,clust$order]
}, mirna)
  
mirna.cov = do.call(rbind, mirna.cov[c('SC','DE','MESO5','ECTO','MESO15','MESO30','EB')])
colnames(mirna.cov) = gsub('Diffname_short','DifferentiationStage', colnames(mirna.cov))
mirna = do.call(cbind, mirna[c('SC','DE','MESO5','ECTO','MESO15','MESO30','EB')])
  
clust = pamk(tmp.mirna, krange = 2:10, seed = 123456, usepam = T)
clust = rownameToFirstColumn(clust$pamobject$clustering, 'hgnc_symbol') %>%
  dplyr::rename(cluster = DF) %>%
  dplyr::mutate(cluster = factor(cluster, labels = c('1' = 'LATE', '2' = 'EARLY')))
  
# Top Annotation
ha = HeatmapAnnotation(mirna.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'MESO5' = 'blue', 'DE' = 'green', 
                                                             'ECTO' = 'yellow', 'MESO15' = 'brown', 'MESO30' = 'purple',
                                                             'EB' = 'turquoise')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('EARLY' = 'orange', 'LATE' = 'cyan')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(mirna)[1],'mirna'))
h2 = Heatmap(mirna, name = 'Scaled logCPM', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-1, 0, 1), c('blue','black','yellow')),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h2)

# PC plot mirna
PC = prcomp(t(tmp.mirna), scale. = T, center = T)

# Reproject all samples in to the PC space
pc = sapply(1:dim(mirna)[2], function(i, expr, pcs){
  tmp = lm(expr ~ 0 + ., cbind(data.frame(expr = scale(expr[rownames(pcs),i])), pcs))$coefficients
}, mirna, PC$rotation) %>% t
rownames(pc) = colnames(mirna)  
  
plotdata = rbind(plotdata, data.frame(UID = rownames(pc), PC1 = pc[,1], PC2 = pc[,2], assay = 'mirna', row = 2, col = 1) %>%
  left_join(mirna.cov[,c('UID', 'DifferentiationStage'), drop=F]))
```
Visualise methylation expression profiles
```{r extract.specific.ent.methyl, out.height= '10in'}
methyl = expr[['methyl']]
rownames(methyl) = methyl[,1]
methyl[,1] = NULL
methyl[is.na(methyl)] = 0
  
methyl.cov = covariates[['methyl']]
rownames(methyl.cov) = methyl.cov$UID
# methyl.cov = methyl.cov[methyl.cov$Diffname_short != 'EB',]
  
tmp.methyl = methyl[rownames(methyl) %in% unique(specific.signatures$feature),
                    methyl.cov$UID[methyl.cov$Diffname_short !='SC' & methyl.cov$Diffname_short !='EB']]
tmp.methyl = scale(tmp.methyl)
tmp.methyl = t(scale(t(tmp.methyl)))

methyl = methyl[rownames(methyl) %in% unique(specific.signatures$feature),rownames(methyl.cov)]
methyl = scale(methyl)
methyl = t(scale(t(methyl)))

methyl.cov = split(methyl.cov, methyl.cov$Diffname_short)
methyl = lapply(methyl.cov, function(x, methyl){
  methyl = methyl[,x$UID]
  clust = hclust(dist(t(methyl)))
  methyl = methyl[,clust$order]
}, methyl)
  
methyl.cov = do.call(rbind, methyl.cov[c('SC','DE','MESO5','ECTO','EB')])
colnames(methyl.cov) = gsub('Diffname_short','DifferentiationStage', colnames(methyl.cov))
methyl = do.call(cbind, methyl[c('SC','DE','MESO5','ECTO','EB')])
  
clust = pamk(tmp.methyl, krange = 2:10, seed = 123456, usepam = T)
clust = rownameToFirstColumn(clust$pamobject$clustering, 'hgnc_symbol') %>%
  dplyr::rename(cluster = DF) %>%
  dplyr::mutate(cluster = factor(cluster, labels = c('1' = 'MESENDO', '2' = 'ECTO')))
  
# Top Annotation
ha = HeatmapAnnotation(methyl.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'MESO5' = 'blue', 'DE' = 'green', 
                                                             'ECTO' = 'yellow', 'MESO15' = 'brown', 'MESO30' = 'purple',
                                                             'EB' = 'turquoise')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('MESENDO' = 'orange', 'ECTO' = 'cyan')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(methyl)[1],'methyl'))
h3 = Heatmap(methyl, name = 'Scaled Beta', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-1.5,0,1.5), c('blue','black','yellow')),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h3)

# PC plot methyl
PC = prcomp(t(tmp.methyl), scale. = T, center = T)

# Reproject all samples in to the PC space
pc = sapply(1:dim(methyl)[2], function(i, expr, pcs){
  tmp = lm(expr ~ 0 + ., cbind(data.frame(expr = scale(expr[rownames(pcs),i])), pcs))$coefficients
}, methyl, PC$rotation) %>% t
rownames(pc) = colnames(methyl)  
  
plotdata = rbind(plotdata, data.frame(UID = rownames(pc), PC1 = pc[,1], PC2 = pc[,2], assay = 'methyl', row = 2, col = 1) %>%
  left_join(methyl.cov[,c('UID', 'DifferentiationStage'), drop=F]))
```
Visualise splicing expression profiles
```{r extract.specific.ent.splicing, out.height= '10in'}
# Remove annotations from splicing
splicing.annot = expr$splicing[,c(1, 227:236)]
expr$splicing = expr$splicing[,1:226]

splicing = expr[['splicing']]
rownames(splicing) = splicing[,1]
splicing[,1] = NULL
splicing[is.na(splicing)] = 0
  
splicing.cov = covariates[['splicing']]
rownames(splicing.cov) = splicing.cov$UID
# splicing.cov = splicing.cov[splicing.cov$Diffname_short != 'EB',]
  
tmp.splicing = splicing[rownames(splicing) %in% unique(specific.signatures$feature),
                        splicing.cov$UID[splicing.cov$Diffname_short !='SC' & splicing.cov$Diffname_short !='EB']]
tmp.splicing = scale(tmp.splicing)
tmp.splicing = t(scale(t(tmp.splicing)))

splicing = splicing[rownames(splicing) %in% unique(specific.signatures$feature),rownames(splicing.cov)]
splicing = scale(splicing)
splicing = t(scale(t(splicing)))

splicing.cov = split(splicing.cov, splicing.cov$Diffname_short)
splicing = lapply(splicing.cov, function(x, splicing){
  splicing = splicing[,x$UID]
  clust = hclust(dist(t(splicing)))
  splicing = splicing[,clust$order]
}, splicing)
  
splicing.cov = do.call(rbind, splicing.cov[c('SC','DE','MESO5','ECTO','MESO15','MESO30','EB')])
colnames(splicing.cov) = gsub('Diffname_short','DifferentiationStage', colnames(splicing.cov))
splicing = do.call(cbind, splicing[c('SC','DE','MESO5','ECTO','MESO15','MESO30','EB')])
  
clust = pamk(tmp.splicing, krange = 2:10, seed = 123456, usepam = T)
clust = rownameToFirstColumn(clust$pamobject$clustering, 'hgnc_symbol') %>%
  dplyr::rename(cluster = DF) %>%
  dplyr::mutate(cluster = factor(cluster, labels = c('1' = '1', '2' = '2')))
  
# Top Annotation
ha = HeatmapAnnotation(splicing.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'MESO5' = 'blue', 'DE' = 'green', 
                                                             'ECTO' = 'yellow', 'MESO15' = 'brown', 'MESO30' = 'purple',
                                                             'EB' = 'turquoise')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('1' = 'orange', '2' = 'cyan')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(splicing)[1],'spliced junctions'))
h4 = Heatmap(splicing, name = 'Scaled PSI', top_annotation = ha,
        col = colorRamp2(c(-2,0,2), c('blue','black','yellow')),
        split = clust$cluster,
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h4)

# PC plot splicing
PC = prcomp(t(tmp.splicing), scale. = T, center = T)

# Reproject all samples in to the PC space
pc = sapply(1:dim(splicing)[2], function(i, expr, pcs){
  tmp = lm(expr ~ 0 + ., cbind(data.frame(expr = scale(expr[rownames(pcs),i])), pcs))$coefficients
}, splicing, PC$rotation) %>% t
rownames(pc) = colnames(splicing)  
  
plotdata = rbind(plotdata, data.frame(UID = rownames(pc), PC1 = pc[,1], PC2 = pc[,2], assay = 'splicing', row = 2, col = 1) %>%
  left_join(splicing.cov[,c('UID', 'DifferentiationStage'), drop=F]))
```
All samples projected in to the pc space of germ layers
```{r plot.pca, fig.height=15, fig.width=10}
pl = list()
plotdata$DifferentiationStage = factor(plotdata$DifferentiationStage,
                                       levels = c("SC","EB","DE","MESO5","ECTO","MESO15","MESO30"))
for (assays in levels(plotdata$assay)){
  p = ggplot(filter(plotdata, assay == assays), 
             aes(x = PC1, y = PC2, color = DifferentiationStage, shape = DifferentiationStage)) 
  p = p + geom_point(stroke = 0.75) + theme_bw()
  p = p + scale_shape_manual(values = c(0,2,5,8,12,13,14)) + scale_color_manual(values = cbPalette[c(1:4,6:8)])
  p = p + theme(legend.position = 'top')
  p = p + ggtitle(assays) + scale_size_manual(values = 4)
  
  pl[[assays]] = p
}
multiplot(plotlist = pl, cols = 2)
```

```{r association.test, fig.height=6, fig.width=6}
library(CovariateAnalysis)
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

for (assays in unique(plotdata$assay)){
  for(DiffState in c('SC','EB')){
    writeLines(paste('DiffState:',DiffState,'Assay:',assays))
    tmp = filter(plotdata, assay == assays, DifferentiationStage == DiffState) %>%
      left_join(covariates[[assays]]) 
    ind = colnames(tmp) %in% c('PC1', 'PC2', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 
                    'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 
                    'Gender', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 
                    'Reprogramming_Vector_Type',  'PassageAtThaw', 'PassageAtHarvest')
    tmp = tmp[,ind]
    
    if(assays != 'splicing'){
      tmp$Culture_Conditions2 = tmp$Culture_Conditions
      tmp$Culture_Conditions2[tmp$Culture_Conditions != 'unknown' &
                                tmp$Culture_Conditions != 'Stromalprimed'] = 'Stromalnonprimed'
    }
    ind1 = which(sapply(tmp, class) == "character")
    ind2 = setdiff(1:dim(tmp)[2], ind1)
    tmp[is.na(tmp)] = 0
    tmp[,ind1] = lapply(tmp[,ind1], factor)
    tmp[,ind2] = lapply(tmp[,ind2], as.integer)
    
    ind1 = ind1[which((sapply(tmp[,ind1], levels) %>% sapply(length)) != 1)]
    tmp = tmp[,c(ind1,ind2)]
    
    tmp1 = getAssociationStatistics(tmp, PVAL = 0.1)
    ggheatmap.show(tmp1$plot, col.width=0.4, row.width=0.25)

    writeLines ('Variables associated with PC1 at an FDR 0.05 are')
    setdiff(colnames(tmp1$PVAL)[which(tmp1$PVAL['PC1',] <= 0.05)], c('PC1', 'PC2'))
    # writeLines('Effect size are')
    # print(tmp1$ESTIMATE['PC1', which(tmp1$PVAL['PC1',] <= 0.05)])

    writeLines ('Variables associated with PC2 at an FDR 0.05 are')
    setdiff(colnames(tmp1$PVAL)[which(tmp1$PVAL['PC2',] <= 0.05)], c('PC1', 'PC2'))
    # writeLines('Effect size are')
    # print(tmp1$ESTIMATE['PC2', which(tmp1$PVAL['PC2',] <= 0.05)])
  }
}

```

```{r pcs.culture.condition, fig.height=12, fig.width=10}
tmp = lapply(covariates[c('mrna','mirna','methyl')], function(x){
  x = dplyr::select(x, UID, Diffname_short, Culture_Conditions)
  x$Culture_Conditions[x$Culture_Conditions != 'unknown' &
                         x$Culture_Conditions != 'Stromalprimed'] = 'Stromalnonprimed'
  x$Culture_Conditions = factor(x$Culture_Conditions)
  return(x)
}) %>%
  rbindlist(idcol = 'assay', fill = T, use.names=T) %>% data.table %>%
  left_join(plotdata %>% data.table) %>%  
  filter(assay != 'splicing') %>%
  droplevels

plotlist = list()
for (assays in c('mrna','mirna','methyl')){
  p = ggplot(tmp %>% filter(assay == assays, Culture_Conditions != 'unknown'), 
             aes(x = PC1, y = PC2, color = Culture_Conditions, shape = DifferentiationStage))
  p = p + geom_point() + theme_bw() + scale_shape_manual(values = 1:7)
  plotlist[[assays]] = p + ggtitle(assays)
}
multiplot(plotlist = plotlist, cols = 1)
```

### Perform enrichment analysis
```{r filter.genesets}
# Get mrna specific genesets
genesets.to.test.mrna = specific.signatures %>%
  filter(feature.assay == 'mrna') %>%
  dplyr::mutate(feature.direction = factor(feature.direction, labels = c('-1' = 'DOWN', '1' = 'UP'))) %>%
  dlply(.(from.state, feature.assay, feature.direction), .fun = function(x){
    unlist(x$feature) %>% unique %>% as.character
  })

# Get mirna, methylation specific genesets
genesets.to.test.others = specific.signatures.store %>%
  filter(feature.assay != 'mrna', feature.assay != 'splicing') %>%
  dplyr::select(-adj.P.value, -logFC, -changeBETA, -changePSI) %>%
  left_join(all.int  %>%
              dplyr::mutate(from.state = gsub('MESO5','MESENDO',from.state), from.state = gsub('DE','MESENDO',from.state),
                            to.state = gsub('MESO5','MESENDO',to.state), to.state = gsub('DE','MESENDO',to.state),
                            from.state = gsub('MESO15','MESO',from.state), from.state = gsub('MESO30','MESO',from.state),
                            to.state = gsub('MESO15','MESO',to.state), to.state = gsub('MESO30','MESO',to.state)) %>%
              dplyr::filter(from.state != 'EB', to.state != 'EB', !(from.state == 'MESO' & to.state == 'MESO'), 
                            !(from.state == 'MESENDO' & to.state == 'MESENDO'))) %>%
  dplyr::filter(!is.na(target)) %>%
  dplyr::mutate(feature.direction = factor(feature.direction, labels = c('-1' = 'DOWN', '1' = 'UP')),
                target.direction = factor(target.direction, labels = c('-1' = 'DOWN', '1' = 'UP'))) %>%
  dlply(.(from.state, feature.assay, feature.direction, target.direction),
        .fun = function(x){
          unique(x$feature)
  })
genesets.to.test.others = genesets.to.test.others[setdiff(1:length(genesets.to.test.others), grep('DOWN.DOWN', names(genesets.to.test.others)))]
genesets.to.test.others = genesets.to.test.others[setdiff(1:length(genesets.to.test.others), grep('UP.UP', names(genesets.to.test.others)))]

# Get splicing specific genesets
genesets.to.test.splicing = specific.signatures.store %>%
  filter(feature.assay == 'splicing') %>%
  dplyr::select(-adj.P.value, -logFC, -changeBETA, -changePSI) %>%
  left_join(splicing.annot  %>%
              dplyr::select(one_of(c('Minor-Isoform', 'Symbol'))) %>%
              plyr::rename(c('Minor-Isoform' = 'feature'))) %>%
  dplyr::mutate(feature.direction = factor(feature.direction, labels = c('-1' = 'DOWN', '1' = 'UP'))) %>%
  dlply(.(from.state, feature.assay, feature.direction),
        .fun = function(x){
          unique(x$Symbol)
  })

genesets.to.test = c(genesets.to.test.mrna, genesets.to.test.splicing, genesets.to.test.others)
genes.in.background = expr[['mrna']]$GeneName %>% unique()
```

```{r perform.enrichment, eval=FALSE}
# Download gene sets
GENE.SETS_ID = 'syn4867851'
ALL_USED_IDs = c(ALL_USED_IDs, GENE.SETS_ID)
load(synGet(GENE.SETS_ID)@filePath)

# Choose gene sets for testing
gsets = c("BioCarta", "GO_Biological_Process", "KEGG_2015", "Reactome", "WikiPathways_2015")
GeneSets = GeneSets[gsets]
GeneSets = filterGeneSets(GeneSets, genes.in.background, minSize = 2, maxSize = 5000)
GeneSets = mapply(function(x, y){
  names(x) = paste(y,names(x),sep = '_')
  return(x)
}, GeneSets, names(GeneSets))
GeneSets = do.call(c, GeneSets)

# Function to perform Fishers enrichment analysis
fisherEnrichment <- function(genesInSignificantSet, # A character vector of differentially expressed or some significant genes to test
                             genesInGeneSet, # A character vector of genes in gene set like GO annotations, pathways etc...
                             genesInBackground # Background genes that are 
){
  genesInSignificantSet = intersect(genesInSignificantSet, genesInBackground) # back ground filtering
  genesInNonSignificantSet = base::setdiff(genesInBackground, genesInSignificantSet)
  genesInGeneSet = intersect(genesInGeneSet, genesInBackground) # back ground filtering
  genesOutGeneSet = base::setdiff(genesInBackground,genesInGeneSet)
  
  pval = fisher.test(
    matrix(c(length(intersect(genesInGeneSet, genesInSignificantSet)),             
             length(intersect(genesInGeneSet, genesInNonSignificantSet)),
             length(intersect(genesOutGeneSet, genesInSignificantSet)),
             length(intersect(genesOutGeneSet, genesInNonSignificantSet))), 
           nrow=2, ncol=2),
    alternative="greater")
  OR = (length(intersect(genesInGeneSet, genesInSignificantSet)) * length(intersect(genesOutGeneSet, genesInNonSignificantSet))) / (length(intersect(genesInGeneSet, genesInNonSignificantSet)) * length(intersect(genesOutGeneSet, genesInSignificantSet)))
  return(data.frame(pval = pval$p.value,
                    ngenes = length(genesInGeneSet),
                    noverlap = length(intersect(genesInGeneSet, genesInSignificantSet)),
                    Odds.Ratio = OR,
                    Genes = paste(intersect(genesInGeneSet, genesInSignificantSet), collapse = '|')))
}

library(doParallel)
library(foreach)

cl = makeCluster(14)
registerDoParallel(cl)

enrichment.results = foreach(i = 1:length(genesets.to.test),
                             .export = c('genesets.to.test', 'genes.in.background', 'GeneSets', 'fisherEnrichment'),
                             .packages = c('dplyr', 'data.table')) %dopar% {
  tmp = lapply(GeneSets, function(x){fisherEnrichment(genesets.to.test[[i]], x, genes.in.background)}) %>%
    rbindlist(idcol = 'GeneSetName') %>%
    dplyr::mutate(fdr = p.adjust(pval, method = 'BH'))
                             }
names(enrichment.results) = names(genesets.to.test)
enrichment.results = rbindlist(enrichment.results, idcol = 'ComparisonName')
```

### Store results in synapse
```{r synstore, cache=FALSE}
tfs.id = 'syn6040938'
ALL_USED_IDs = c(ALL_USED_IDs, tfs.id)
tfs = downloadFile(tfs.id)

# Store signatures in synapse
tmp = specific.signatures.store
tmp$feature.assay[tmp$feature %in% tfs$feature] = 'TF'
write.table(tmp, file = 'GlobalSignatures.tsv', row.names = F, sep = '\t', quote=F)
obj = File('GlobalSignatures.tsv', name = "Signatures (Global)", parentId = CODE$properties$id)
obj = synStore(obj, activityName = ActivityName, executed = thisFile, used = as.character(ALL_USED_IDs))

# Store enrichment results in synapse
# write.table(enrichment.results, file = 'EnrichmentResults.tsv', row.names = F, sep = '\t', quote=F)
# obj = File('EnrichmentResults.tsv', name = "Enrichment Results (Global)", parentId = CODE$properties$id)
# obj = synStore(obj, activityName = ActivityName, executed = thisFile, used = as.character(ALL_USED_IDs))
```

```{r top.features}
writeLines('Top 5 differentially expressed TFs, and nonTFs signatures')
# Get top n differentially expresed entities
tmp %>% 
  filter(feature.assay %in% c('mrna', 'TF')) %>%
  dplyr::group_by(feature, feature.assay, feature.direction, from.state) %>% 
  summarise(median.lfc = median(logFC)) %>% 
  group_by(from.state, feature.assay, feature.direction) %>% 
  top_n(5, median.lfc) %>% 
  group_by(from.state,feature.assay, feature.direction) %>%
  summarise(feature = paste(unique(feature), collapse=',')) %>%
  spread(feature.assay, feature) %>%
  kable

writeLines('Top 5 differentially expressed miRNA signatures')
# Get top n differentially expresed entities
tmp %>% 
  filter(feature.assay %in% c('mirna')) %>%
  left_join(all.int  %>%
              dplyr::mutate(from.state = gsub('MESO5','MESENDO',from.state), from.state = gsub('DE','MESENDO',from.state),
                            to.state = gsub('MESO5','MESENDO',to.state), to.state = gsub('DE','MESENDO',to.state),
                            from.state = gsub('MESO15','MESO',from.state), from.state = gsub('MESO30','MESO',from.state),
                            to.state = gsub('MESO15','MESO',to.state), to.state = gsub('MESO30','MESO',to.state)) %>%
              dplyr::filter(from.state != 'EB', to.state != 'EB', !(from.state == 'MESO' & to.state == 'MESO'), 
                            !(from.state == 'MESENDO' & to.state == 'MESENDO'))) %>%
  filter(!is.na(target.assay)) %>%
  ddply(.(feature, feature.assay, feature.direction, from.state),
        .fun = function(x){
          x = x[which.max(abs(x$target.lfc)), ]
        }) %>% 
  group_by(from.state, feature.assay, feature.direction) %>% 
  top_n(5, feature.lfc) %>% 
  group_by(from.state,feature.assay, feature.direction) %>%
  summarise(feature = paste(unique(feature), collapse=','),
            target = paste(unique(target), collapse=',')) %>%
  kable

writeLines('Top 5 differentially expressed methylation signatures')
tmp %>% 
  filter(feature.assay %in% c('methyl')) %>%
  left_join(all.int  %>%
              dplyr::mutate(from.state = gsub('MESO5','MESENDO',from.state), from.state = gsub('DE','MESENDO',from.state),
                            to.state = gsub('MESO5','MESENDO',to.state), to.state = gsub('DE','MESENDO',to.state),
                            from.state = gsub('MESO15','MESO',from.state), from.state = gsub('MESO30','MESO',from.state),
                            to.state = gsub('MESO15','MESO',to.state), to.state = gsub('MESO30','MESO',to.state)) %>%
              dplyr::filter(from.state != 'EB', to.state != 'EB', !(from.state == 'MESO' & to.state == 'MESO'), 
                            !(from.state == 'MESENDO' & to.state == 'MESENDO'))) %>%
  filter(!is.na(target.assay)) %>%
  ddply(.(feature, feature.assay, feature.direction, from.state),
        .fun = function(x){
          x = x[which.max(abs(x$target.lfc)), ]
        }) %>% 
  group_by(from.state, feature.assay, feature.direction) %>% 
  top_n(5, feature.lfc) %>% 
  group_by(from.state,feature.assay, feature.direction) %>%
  summarise(feature = paste(unique(feature), collapse=','),
            target = paste(unique(target), collapse=',')) %>%
  kable

writeLines('Top 10 differentially spliced signatures and their targets')
# Get top n differentially expresed entities
tmp %>% 
  filter(feature.assay %in% c('splicing')) %>%
  left_join(splicing.annot %>% 
              dplyr::select(one_of('Minor-Isoform', 'Symbol')) %>% 
              plyr::rename(c('Minor-Isoform' = 'feature', 'Symbol' = 'target')) %>%
              unique) %>%
  dplyr::group_by(feature, feature.assay, feature.direction, from.state) %>% 
  summarise(median.lfc = median(logFC), target = paste(sort(unique(target)), collapse = ',')) %>% 
  group_by(from.state, feature.assay, feature.direction) %>% 
  top_n(10, median.lfc) %>% 
  group_by(from.state,feature.assay, feature.direction) %>%
  summarise(feature = paste(unique(feature), collapse=','),
            target = paste(unique(target), collapse=',')) %>%
  kable
```

### Source Code
[Source R Markdown](`r thisFile`)