---
title: Markers/signatures of pluripotency and germline stages
author: "Thanneer Perumal"
date: '`r date()`'
output: pdf_document
---
```{r knit2synapse, eval = FALSE, include=FALSE}
library(synapseClient)
library(knitr)
library(knit2synapse)

synapseLogin()

knit2synapse::knitToFolderEntity(file = 'Section2_SignatureExtraction.Rmd',
                                 parentId = "syn7162983",
                                 entityName = "2 Pluripotency and Germline specific Signatures")
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")

# Load libraries
library(CovariateAnalysis)
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
library(matrixStats)

library(fpc)
library(WGCNA)
library(flashClust)

library(synapseClient)
library(knitr)
library(githubr)

library(SNFtool)
library(igraph)
library(Vennerable)

synapseLogin()

library(circlize)
colfunc <- colorRampPalette(c("white", "red"))

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  fig.align = 'left',
  results = 'asis',
  out.width = '8in')
```

```{r synapseStore.params, include=FALSE}
parentId = "syn7162983"
ALL_USED_IDs = c()

thisFileName = 'Section2_SignatureExtraction.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='publication')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
```
### Data download from synapse
#### Get raw logCPM counts from synapse
```{r getRawCount, include=FALSE}
expr.id = c(mrna = 'syn5011095', mirna = 'syn5014454', methyl = 'syn4487642')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(expr.id))
expr = lapply(expr.id, downloadFile)
```

#### Get covariates from synapse
```{r covars, include=FALSE}
covariates.id = c(mrna = 'syn5011149', mirna = 'syn5014460', methyl = 'syn4487669')
covariates.table.id = c(mrna = 'syn3156503', mirna = 'syn3219876', methyl = 'syn3156828')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(covariates.id), as.character(covariates.table.id))
covariates = lapply(covariates.id, downloadFile)
covariates.table = lapply(covariates.table.id, function(x){
  synTableQuery(paste('select UID,C4_Cell_Line_ID from',x))@values
})

covariates = mapply(function(x,y){
  left_join(x,y)
},covariates, covariates.table)
```

#### Get all possible literature specific interactions from synapse
```{r all.int}
all.int = downloadFile('syn5643683')
```

#### Get differential expression results from synapse
```{r thresholds, echo=TRUE}
pval.th <- 0.01

mrna.lfc.th <- log2(2)
mirna.lfc.th <- log2(2)
methyl.changeBeta.th <- 0.35
splicing.changePSI.th <- 0.2
```

```{r diffexp, include=FALSE}
### Download differential expression results from synapse
diffexp.id = c(mrna = 'syn5013690', mirna = 'syn5014584', methyl = 'syn4527629', splicing = 'syn5049321')
ALL_USED_IDs = c(ALL_USED_IDs, as.character(diffexp.id))

# Diff states in a sensible, quasi time-based ordering
orderedDiffState = sort(c("SC", "MESO5", "DE", "MESO15", "MESO30", "ECTO", "EB"))
Comparisons = apply(combn(sort(orderedDiffState), 2), 2, function(x){paste(x, collapse = '_vs_')})

# Get mrna differential expression results from synapse
diffexp = list()
diffexp$mrna = downloadFile(diffexp.id['mrna']) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mrna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Get mirna differential expression results from synapse
diffexp$mirna = downloadFile(diffexp.id['mirna']) %>%
  dplyr::rename(feature = GeneSymbol) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(logFC) >= mirna.lfc.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC) 

# Get methyl differential expression results from synapse
diffexp$methyl = downloadFile(diffexp.id['methyl']) %>%
  dplyr::select(-one_of('feature')) %>%
  dplyr::rename(feature = methProbeIDs) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(changeBETA) >= methyl.changeBeta.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC, changeBETA) 

# Get splicing differential expression results from synapse
diffexp$splicing = downloadFile(diffexp.id['splicing']) %>%
  dplyr::select(-one_of('feature')) %>%
  plyr::rename(c('Minor-Isoform' = 'feature')) %>%
  tidyr::separate(Comparison, into = c('V1','Comparison','V3'), sep = '\\__') %>%
  dplyr::filter(adj.P.value <= pval.th & abs(changePSI) >= splicing.changePSI.th & Comparison %in% Comparisons) %>%
  dplyr::select(feature, Comparison, adj.P.value, logFC, changePSI) 

diffexp = rbindlist(diffexp, use.names = T, fill = T, idcol = 'Assay') %>%
  tidyr::separate(Comparison, c('from.state','to.state'), sep = '_vs_')

diffexp = rbindlist(list(diffexp,
                         diffexp %>%
                           dplyr::rename(to.state = from.state, from.state = to.state) %>%
                           dplyr::mutate(
                             logFC = -logFC, changeBETA = -changeBETA, changePSI = -changePSI)),
                    use.names = T, fill =T) %>%
  dplyr::mutate(Direction = logFC/abs(logFC))
```
Based on the previous analysis here we define four groups of differentiation stages: SC, DE, ECTO and MESODERM (which is a intersection of mesoderm at day 15 and 30).

### Indentify pluripotency and germline markers
Here we define the markers as genes, miRNAs and methylation probes that are differentially expressed between a given differentiation state and all the other differentiation states
```{r markers}
# Combine differential expression with literature specific interactions for miRNA and methylation
diffexp.int = diffexp %>%
  filter(Assay %in% c('mirna','methyl')) %>%
  dplyr::rename(feature.assay = Assay, feature.fdr = adj.P.value, feature.lfc = logFC, 
                feature.cb = changeBETA, feature.psi = changePSI, feature.direction = Direction) %>%
  left_join(all.int) %>%
  inner_join(diffexp %>%
               dplyr::rename(target = feature, target.assay = Assay, target.fdr = adj.P.value, target.lfc = logFC, 
                             target.cb = changeBETA, target.psi = changePSI, target.direction = Direction) %>%
               left_join(all.int)) %>%
  filter(feature.direction != target.direction) 

# Combine mrna, mirna, and dna methylation results
diffexp.all = diffexp %>%
  dplyr::rename(feature.assay = Assay, feature.direction = Direction) %>% 
  dplyr::select(feature, feature.assay, feature.direction, from.state, to.state) %>% 
  unique %>%
  dplyr::filter(from.state != 'EB', to.state != 'EB') %>%
  dplyr::filter(from.state != 'MESO5', to.state != 'MESO5')

# Extract specific signatures
specific.signatures = rbind(
  diffexp.all %>%
    filter(!(from.state %in% c('MESO15','MESO30'))) %>%
    group_by(feature, feature.assay, feature.direction, from.state) %>%
    summarise(ncount.to = length(unique(to.state)), to.state = paste(unique(to.state), collapse = ',')) %>%
    filter((ncount.to == 4 & feature.assay != 'methyl') | (ncount.to == 2 & feature.assay == 'methyl')),
  diffexp.all %>%
    filter(from.state %in% c('MESO15','MESO30')) %>%
    filter(!(to.state %in% c('MESO15', 'MESO30'))) %>%
    group_by(feature, feature.assay, feature.direction) %>%
    summarise(ncount.from = length(unique(from.state)), from.state = paste(unique(from.state), collapse = ','),
              ncount.to = length(unique(to.state)), to.state = paste(unique(to.state), collapse = ',')) %>%
  
      filter(ncount.from == 2 & ncount.to == 3) %>%
    dplyr::mutate(from.state = 'MESO')
  ) %>%
    data.frame

writeLines('Number of entities that are differentially expressed between any one differentiation cluster with respect to all the other are')
kable(specific.signatures %>%
        group_by(from.state, feature.assay, feature.direction) %>%
        summarise(ncount = length(unique(feature))) %>%
        spread(from.state, ncount))
```
Following are the visualisations of these signatures

#### Visualise pluripotency and germline markers as heatmaps
Visualise mRNA expression profiles

Up regulated mRNAs
```{r mrna.up, out.height= '10in'}
mrna = expr[['mrna']]
rownames(mrna) = mrna[,1]
mrna[,1] = NULL
mrna[is.na(mrna)] = 0
  
mrna.cov = covariates[['mrna']] %>%
  filter(!(Diffname_short %in% c('EB','MESO5')))
rownames(mrna.cov) = mrna.cov$UID

clust = specific.signatures %>%
  filter(feature.assay == 'mrna', feature.direction == 1) %>%
  dplyr::rename(hgnc_symbol = feature, cluster = from.state) %>%
  dplyr::select(hgnc_symbol, cluster)
rownames(clust) = clust$hgnc_symbol

tmp.mrna = mrna[rownames(mrna) %in% unique(clust$hgnc_symbol), mrna.cov$UID]
tmp.mrna = scale(tmp.mrna)
tmp.mrna = t(scale(t(tmp.mrna)))

mrna = mrna[rownames(mrna) %in% unique(clust$hgnc_symbol),rownames(mrna.cov)]
mrna = scale(mrna)
mrna = t(scale(t(mrna)))

mrna.cov = split(mrna.cov, mrna.cov$Diffname_short)
mrna = lapply(mrna.cov, function(x, mrna){
  mrna = mrna[,x$UID]
  clust = hclust(dist(t(mrna)))
  mrna = mrna[,clust$order]
}, mrna)
  
mrna.cov = do.call(rbind, mrna.cov[c('SC','DE','ECTO','MESO15','MESO30')])
colnames(mrna.cov) = gsub('Diffname_short','DifferentiationStage', colnames(mrna.cov))
mrna = do.call(cbind, mrna[c('SC','DE','ECTO','MESO15','MESO30')])
mrna = mrna[rownames(clust),]

# Top Annotation
ha = HeatmapAnnotation(mrna.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage',
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'DE' = 'green',
                                                             'ECTO' = 'yellow', 'MESO15' = 'brown', 
                                                             'MESO30' = 'purple')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('SC' = 'violet', 'DE' = 'green',
                                                 'ECTO' = 'yellow', 'MESO' = 'brown')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(mrna)[1],'mrna'))
h1 = Heatmap(mrna, name = 'Scaled logCPM', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-2,0,2), c("blue","black","yellow")),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h1)
```

Down regulated mRNAs
```{r mrna.down, out.height= '10in'}
mrna = expr[['mrna']]
rownames(mrna) = mrna[,1]
mrna[,1] = NULL
mrna[is.na(mrna)] = 0
  
mrna.cov = covariates[['mrna']] %>%
  filter(!(Diffname_short %in% c('EB','MESO5')))
rownames(mrna.cov) = mrna.cov$UID

clust = specific.signatures %>%
  filter(feature.assay == 'mrna', feature.direction == -1) %>%
  dplyr::rename(hgnc_symbol = feature, cluster = from.state) %>%
  dplyr::select(hgnc_symbol, cluster)
rownames(clust) = clust$hgnc_symbol

tmp.mrna = mrna[rownames(mrna) %in% unique(clust$hgnc_symbol), mrna.cov$UID]
tmp.mrna = scale(tmp.mrna)
tmp.mrna = t(scale(t(tmp.mrna)))

mrna = mrna[rownames(mrna) %in% unique(clust$hgnc_symbol),rownames(mrna.cov)]
mrna = scale(mrna)
mrna = t(scale(t(mrna)))

mrna.cov = split(mrna.cov, mrna.cov$Diffname_short)
mrna = lapply(mrna.cov, function(x, mrna){
  mrna = mrna[,x$UID]
  clust = hclust(dist(t(mrna)))
  mrna = mrna[,clust$order]
}, mrna)
  
mrna.cov = do.call(rbind, mrna.cov[c('SC','DE','ECTO','MESO15','MESO30')])
colnames(mrna.cov) = gsub('Diffname_short','DifferentiationStage', colnames(mrna.cov))
mrna = do.call(cbind, mrna[c('SC','DE','ECTO','MESO15','MESO30')])
mrna = mrna[rownames(clust),]

# Top Annotation
ha = HeatmapAnnotation(mrna.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage',
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'DE' = 'green',
                                                             'ECTO' = 'yellow', 'MESO15' = 'brown', 
                                                             'MESO30' = 'purple')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('SC' = 'violet', 'DE' = 'green',
                                                 'ECTO' = 'yellow', 'MESO' = 'brown')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(mrna)[1],'mrna'))
h1 = Heatmap(mrna, name = 'Scaled logCPM', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-2,0,2), c("blue","black","yellow")),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h1)
```

Visualise miRNA expression profiles
Upregulated miRNAs
```{r mirna.up, out.height= '10in'}
mirna = expr[['mirna']]
rownames(mirna) = mirna[,1]
mirna[,1] = NULL
mirna[is.na(mirna)] = 0
  
mirna.cov = covariates[['mirna']] %>%
  filter(!(Diffname_short %in% c('EB','MESO5')))
rownames(mirna.cov) = mirna.cov$UID

clust = specific.signatures %>%
  filter(feature.assay == 'mirna', feature.direction == 1) %>%
  dplyr::rename(hgnc_symbol = feature, cluster = from.state) %>%
  dplyr::select(hgnc_symbol, cluster)
rownames(clust) = clust$hgnc_symbol

tmp.mirna = mirna[rownames(mirna) %in% unique(clust$hgnc_symbol), mirna.cov$UID]
tmp.mirna = scale(tmp.mirna)
tmp.mirna = t(scale(t(tmp.mirna)))

mirna = mirna[rownames(mirna) %in% unique(clust$hgnc_symbol), mirna.cov$UID]
mirna = scale(mirna)
mirna = t(scale(t(mirna)))

mirna.cov = split(mirna.cov, mirna.cov$Diffname_short)
mirna = lapply(mirna.cov, function(x, mirna){
  mirna = mirna[,x$UID]
  clust = hclust(dist(t(mirna)))
  mirna = mirna[,clust$order]
}, mirna)
  
mirna.cov = do.call(rbind, mirna.cov[c('SC','DE','ECTO','MESO15','MESO30')])
colnames(mirna.cov) = gsub('Diffname_short','DifferentiationStage', colnames(mirna.cov))
mirna = do.call(cbind, mirna[c('SC','DE','ECTO','MESO15','MESO30')])
mirna = mirna[rownames(clust),]  

# Top Annotation
ha = HeatmapAnnotation(mirna.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow', 
                                                             'MESO15' = 'brown', 'MESO30' = 'purple')))

# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow', 'MESO' = 'brown')))

# Draw heatmap
writeLines(paste('Heatmap of',dim(mirna)[1],'mirna'))
h2 = Heatmap(mirna, name = 'Scaled logCPM', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-1, 0, 1), c('blue','black','yellow')),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h2)
```

Downregulated miRNAs
```{r mirna.down, out.height= '10in'}
mirna = expr[['mirna']]
rownames(mirna) = mirna[,1]
mirna[,1] = NULL
mirna[is.na(mirna)] = 0
  
mirna.cov = covariates[['mirna']] %>%
  filter(!(Diffname_short %in% c('EB','MESO5')))
rownames(mirna.cov) = mirna.cov$UID

clust = specific.signatures %>%
  filter(feature.assay == 'mirna', feature.direction == -1) %>%
  dplyr::rename(hgnc_symbol = feature, cluster = from.state) %>%
  dplyr::select(hgnc_symbol, cluster)
rownames(clust) = clust$hgnc_symbol

tmp.mirna = mirna[rownames(mirna) %in% unique(clust$hgnc_symbol), mirna.cov$UID]
tmp.mirna = scale(tmp.mirna)
tmp.mirna = t(scale(t(tmp.mirna)))

mirna = mirna[rownames(mirna) %in% unique(clust$hgnc_symbol), mirna.cov$UID]
mirna = scale(mirna)
mirna = t(scale(t(mirna)))

mirna.cov = split(mirna.cov, mirna.cov$Diffname_short)
mirna = lapply(mirna.cov, function(x, mirna){
  mirna = mirna[,x$UID]
  clust = hclust(dist(t(mirna)))
  mirna = mirna[,clust$order]
}, mirna)
  
mirna.cov = do.call(rbind, mirna.cov[c('SC','DE','ECTO','MESO15','MESO30')])
colnames(mirna.cov) = gsub('Diffname_short','DifferentiationStage', colnames(mirna.cov))
mirna = do.call(cbind, mirna[c('SC','DE','ECTO','MESO15','MESO30')])
mirna = mirna[rownames(clust),]  

# Top Annotation
ha = HeatmapAnnotation(mirna.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow', 
                                                             'MESO15' = 'brown', 'MESO30' = 'purple')))

# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow', 'MESO' = 'brown')))

# Draw heatmap
writeLines(paste('Heatmap of',dim(mirna)[1],'mirna'))
h2 = Heatmap(mirna, name = 'Scaled logCPM', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-2, 0, 2), c('blue','black','yellow')),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h2)
```

Visualise methylation expression profiles
Hypermethylated probes
```{r methyl.up, out.height= '10in'}
methyl = expr[['methyl']]
rownames(methyl) = methyl[,1]
methyl[,1] = NULL
methyl[is.na(methyl)] = 0
  
methyl.cov = covariates[['methyl']] %>%
  filter(!(Diffname_short %in% c('EB','MESO5')))
rownames(methyl.cov) = methyl.cov$UID

clust = specific.signatures %>%
  filter(feature.assay == 'methyl', feature.direction == 1) %>%
  dplyr::rename(hgnc_symbol = feature, cluster = from.state) %>%
  dplyr::select(hgnc_symbol, cluster)
rownames(clust) = clust$hgnc_symbol

tmp.methyl = methyl[rownames(methyl) %in% unique(clust$hgnc_symbol), methyl.cov$UID]
tmp.methyl = scale(tmp.methyl)
tmp.methyl = t(scale(t(tmp.methyl)))

methyl = methyl[rownames(methyl) %in% unique(clust$hgnc_symbol), methyl.cov$UID]
methyl = scale(methyl)
methyl = t(scale(t(methyl)))

methyl.cov = split(methyl.cov, methyl.cov$Diffname_short)
methyl = lapply(methyl.cov, function(x, methyl){
  methyl = methyl[,x$UID]
  clust = hclust(dist(t(methyl)))
  methyl = methyl[,clust$order]
}, methyl)
  
methyl.cov = do.call(rbind, methyl.cov[c('SC','DE','ECTO')])
colnames(methyl.cov) = gsub('Diffname_short','DifferentiationStage', colnames(methyl.cov))
methyl = do.call(cbind, methyl[c('SC','DE','ECTO')])
methyl = methyl[rownames(clust),]

# Top Annotation
ha = HeatmapAnnotation(methyl.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(methyl)[1],'methylation probes'))
h3 = Heatmap(methyl, name = 'Scaled Beta', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-1.5,0,1.5), c('blue','black','yellow')),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h3)
```

Hypomethylated probes
```{r methyl.down, out.height= '10in'}
methyl = expr[['methyl']]
rownames(methyl) = methyl[,1]
methyl[,1] = NULL
methyl[is.na(methyl)] = 0
  
methyl.cov = covariates[['methyl']] %>%
  filter(!(Diffname_short %in% c('EB','MESO5')))
rownames(methyl.cov) = methyl.cov$UID

clust = specific.signatures %>%
  filter(feature.assay == 'methyl', feature.direction == -1) %>%
  dplyr::rename(hgnc_symbol = feature, cluster = from.state) %>%
  dplyr::select(hgnc_symbol, cluster)
rownames(clust) = clust$hgnc_symbol

tmp.methyl = methyl[rownames(methyl) %in% unique(clust$hgnc_symbol), methyl.cov$UID]
tmp.methyl = scale(tmp.methyl)
tmp.methyl = t(scale(t(tmp.methyl)))

methyl = methyl[rownames(methyl) %in% unique(clust$hgnc_symbol), methyl.cov$UID]
methyl = scale(methyl)
methyl = t(scale(t(methyl)))

methyl.cov = split(methyl.cov, methyl.cov$Diffname_short)
methyl = lapply(methyl.cov, function(x, methyl){
  methyl = methyl[,x$UID]
  clust = hclust(dist(t(methyl)))
  methyl = methyl[,clust$order]
}, methyl)
  
methyl.cov = do.call(rbind, methyl.cov[c('SC','DE','ECTO')])
colnames(methyl.cov) = gsub('Diffname_short','DifferentiationStage', colnames(methyl.cov))
methyl = do.call(cbind, methyl[c('SC','DE','ECTO')])
methyl = methyl[rownames(clust),]

# Top Annotation
ha = HeatmapAnnotation(methyl.cov[,c('DifferentiationStage'), drop = F], name = 'DifferentiationStage', 
                       col = list('DifferentiationStage' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow')))
# Side Annotation
ha1 = HeatmapAnnotation(clust[,'cluster',drop=F], which = 'row',
                        col = list('cluster' = c('SC' = 'violet', 'DE' = 'green', 'ECTO' = 'yellow')))
  
# Draw heatmap
writeLines(paste('Heatmap of',dim(methyl)[1],'methylation probes'))
h3 = Heatmap(methyl, name = 'Scaled Beta', top_annotation = ha, split = clust$cluster,
        col = colorRamp2(c(-1.5,0,1.5), c('blue','black','yellow')),
        cluster_columns = F, show_row_names = F, show_column_names = F, show_row_dend = F) + ha1
draw(h3)
```

Compare the exisiting signatures with the signatures from Gene Ontology
```{r compare.GO, eval=FALSE}
# These signatures are 
DE.Sig = c('ONECUT1', 'OTX2', 'GATA6', 'NKX2-1', 'GATA4', 'BPTF', 'HHEX', 'SOX7', 'NKX2-5', 'PAX9', 'MMP14', 'HDAC1')
MESO.Sig = c('BTK', 'FOXC1', 'TEAD2', 'LEF1', 'T', 'SRF', 'FOXH1', 'KDM6A', 'IKZF1', 'SMAD1', 'TBX1', 'FOXC2', 'SNAI1',
             'ECSIT', 'HNF1A', 'HTT', 'HAND1', 'TCF15', 'WNT5A', 'ETV2', 'IRX3', 'EYA1', 'EYA2', 'KDM6B', 'ZFPM2', 'ZIC2',
             'CITED2', 'LHX2', 'TP63', 'HES7', 'OSR1', 'CHURC1', 'FOXF1', 'TBX6')
ECTO.Sig = c('ZBTB17', 'GRHL3', 'ZBTB7B', 'ERF')
SC.Sig = c('ZSCAN10', 'LDB1', 'ZFP36L2', 'PAX8', 'DPPA4', 'ZNF358', 'NFYA', 'KDM2B', 'PBX1', 'PRDM14', 'CNOT1', 'HES5',
           'CNOT2', 'CNOT3', 'EPAS1', 'PRO')

geneSets = dlply(specific.signatures, .(from.state, feature.assay, feature.direction), .fun = function(x){
  unique(x$feature)
})

venn.sc = Venn(c(list(SC.GO = SC.Sig), geneSets[intersect(grep('SC',names(geneSets)), grep('mrna',names(geneSets)))]))
plot(venn.sc, doWeights = F)

venn.de = Venn(c(list(DE.GO = DE.Sig), geneSets[intersect(grep('DE',names(geneSets)), grep('mrna',names(geneSets)))]))
plot(venn.de, doWeights = F)

venn.ecto = Venn(c(list(ECTO.GO = ECTO.Sig), geneSets[intersect(grep('ECTO',names(geneSets)), grep('mrna',names(geneSets)))]))
plot(venn.ecto, doWeights = F)

venn.meso = Venn(c(list(MESO.GO = MESO.Sig), geneSets[intersect(grep('MESO',names(geneSets)), grep('mrna',names(geneSets)))]))
plot(venn.meso, doWeights = F)
```

### Enrichment Analysis
```{r enrich}
## Function to compute Fishers enrichment analysis
fisherEnrichment <- function(genesInSignificantSet, # A character vector of differentially expressed or some significant genes to test
                             genesInGeneSet, # A character vector of genes in gene set like GO annotations, pathways etc...
                             genesInBackground # Background genes that are 
                             ){
  genesInSignificantSet = intersect(genesInSignificantSet, genesInBackground) # back ground filtering
  genesInNonSignificantSet = base::setdiff(genesInBackground, genesInSignificantSet)
  genesInGeneSet = intersect(genesInGeneSet, genesInBackground) # back ground filtering
  genesOutGeneSet = base::setdiff(genesInBackground,genesInGeneSet)

  pval = fisher.test(
    matrix(c(length(intersect(genesInGeneSet, genesInSignificantSet)),             
             length(intersect(genesInGeneSet, genesInNonSignificantSet)),
             length(intersect(genesOutGeneSet, genesInSignificantSet)),
             length(intersect(genesOutGeneSet, genesInNonSignificantSet))), 
           nrow=2, ncol=2),
    alternative="greater")

  OR = (length(intersect(genesInGeneSet, genesInSignificantSet)) * length(intersect(genesOutGeneSet, genesInNonSignificantSet))) / (length(intersect(genesInGeneSet, genesInNonSignificantSet)) * length(intersect(genesOutGeneSet, genesInSignificantSet)))
  return(data.frame(pval = as.numeric(pval$p.value),
                    ngenes = as.numeric(length(genesInGeneSet)),
                    noverlap = as.numeric(length(intersect(genesInGeneSet, genesInSignificantSet))),
                    OR = as.numeric(OR)))
  }

## Get gene sets for enrichment from synapse
geneset.id = 'syn4867851'
ALL_USED_IDs = c(ALL_USED_IDs, geneset.id)
load(synGet(geneset.id)@filePath)

# Choose gene set category for testing
gsets = c("GO_Biological_Process", "KEGG_2015", "Reactome", "WikiPathways_2015")
GeneSets = GeneSets[gsets]

## Get background genesets
genesInBackground = expr$mrna$GeneName
GeneSets = filterGeneSets(GeneSets, genesInBackground)

## Get list of gene signatures for enrichment
geneSetsToTest = specific.signatures %>%
  dplyr::mutate(feature.direction = factor(feature.direction, labels = c('DOWN','UP'))) %>%
  dlply(.(from.state,feature.direction,feature.assay), .fun = function(x){
    if (unique(x$feature.assay) == 'mirna' | unique(x$feature.assay) == 'methyl'){
      x %>%
        dplyr::select(-to.state) %>%
        left_join(diffexp.int %>% 
                    dplyr::mutate(feature.direction = factor(feature.direction, labels = c('DOWN','UP')))) %>%
        filter(target.assay == 'mrna') %>%
        dplyr::select(target) %>%
        unlist %>% unique
    } else {
      unique(x$feature)
    }
  })

geneSets = list()
for(diffstate in c('SC','DE','MESO','ECTO')){
  geneSets[[diffstate]] = list()
  for (direction in c('UP','DOWN')){
    if (direction == 'UP')
      geneSets[[diffstate]][[direction]] = do.call(c, geneSetsToTest[paste(diffstate,c('UP.mrna','DOWN.mirna','DOWN.methyl'),sep='.')]) %>%
        as.character %>% unique()
    else
      geneSets[[diffstate]][[direction]] = do.call(c, geneSetsToTest[paste(diffstate,c('DOWN.mrna','UP.mirna','UP.methyl'),sep='.')]) %>%
        as.character %>% unique()
  }
}

# Perform enrichment analysis
enrichResults = list()
for (diffState in names(geneSets)){
  for (direction in names(geneSets[[diffState]])){
    tmp = lapply(GeneSets,
                 function(x, comp, genesInBackground){
                   tmp = t(sapply(x, fisherEnrichment, comp, genesInBackground))
                   tmp = cbind(tmp, p.adjust(tmp$pval,'fdr'))
                   tmp = rownameToFirstColumn(tmp,'GeneSetName')
                   return(tmp)
                   }, geneSets[[diffState]][[direction]], genesInBackground)
    
    for (name1 in names(tmp))
      tmp[[name1]]$category = name1
    
    enrichResults[[diffState]][[direction]] = rbindlist(tmp)
    writeLines(paste0('Completed ',diffState,direction))  
  }
}

enrichResults = lapply(enrichResults, function(x){
  rbindlist(x, idcol = 'feature.direction')
}) %>%
  rbindlist(idcol = 'from.state')

enrichResults %>% 
  filter(pval <= 0.01, noverlap >= 5) %>% 
  dplyr::select(from.state, feature.direction, GeneSetName, OR) %>%
  dplyr::mutate(OR = as.numeric(OR)) %>%
  unite(signature.set, from.state, feature.direction) %>%
  dplyr::select(signature.set, GeneSetName, OR) %>%
  group_by(signature.set) %>%
  top_n(10, OR) %>%
  spread(signature.set, OR) %>%
  kable
```

### Store intermediate results in synapse
```{r synapse.store}
# Populate wiki with results
CODE <- Folder(name = '2 Pluripotency and Germline specific Signatures',parentId = 'syn7162983')
CODE <- synStore(CODE)

# Store differential expression results in synapse
specific.signatures %>%
  write.table(file = 'Signatures.tsv', sep = '\t', quote=F, row.names = F)
obj = File('Signatures.tsv', 
           name = 'Pluripotency and Germline Specific Signatures',
           parentId = CODE$properties$id)
obj = synStore(obj, 
               activityName = 'Signature Identification', 
               executed = thisFile, 
               used = ALL_USED_IDs)

# Store enrichment results in synapse
enrichResults %>%
  sapply(as.character) %>%
  write.table(file = 'EnrichmentResults.tsv', sep = '\t', quote=F, row.names = F)
obj = File('EnrichmentResults.tsv', 
           name = 'Pluripotency and Germline Specific Signature Gene Set Enrichment',
           parentId = CODE$properties$id)
obj = synStore(obj, 
               activityName = 'Signature Identification', 
               executed = thisFile, 
               used = ALL_USED_IDs)
```